// @pnp/office365-cli/dist/o365/teams/commands/message/message-reply-list.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const commands_1 = require("../../commands");
const GraphItemsListCommand_1 = require("../../../base/GraphItemsListCommand");
const Utils_1 = require("../../../../Utils");
const vorpal = require('../../../../vorpal-init');
class TeamsMessageReplyListCommand extends GraphItemsListCommand_1.GraphItemsListCommand {
    get name() {
        return `${commands_1.default.TEAMS_MESSAGE_REPLY_LIST}`;
    }
    get description() {
        return 'Retrieves replies to a message from a channel in a Microsoft Teams team';
    }
    commandAction(cmd, args, cb) {
        const endpoint = `${this.resource}/beta/teams/${args.options.teamId}/channels/${args.options.channelId}/messages/${args.options.messageId}/replies`;
        this
            .getAllItems(endpoint, cmd, true)
            .then(() => {
            if (args.options.output === 'json') {
                cmd.log(this.items);
            }
            else {
                cmd.log(this.items.map(m => {
                    return {
                        id: m.id,
                        body: m.body.content
                    };
                }));
            }
            if (this.verbose) {
                cmd.log(vorpal.chalk.green('DONE'));
            }
            cb();
        }, (err) => this.handleRejectedODataJsonPromise(err, cmd, cb));
    }
    options() {
        const options = [
            {
                option: '-i, --teamId <teamId>',
                description: 'The ID of the team where the channel is located'
            },
            {
                option: '-c, --channelId <channelId>',
                description: 'The ID of the channel that contains the message'
            },
            {
                option: '-m, --messageId <messageId>',
                description: 'The ID of the message to retrieve replies for'
            }
        ];
        const parentOptions = super.options();
        return options.concat(parentOptions);
    }
    validate() {
        return (args) => {
            if (!args.options.teamId) {
                return 'Required parameter teamId missing';
            }
            if (!args.options.channelId) {
                return 'Required parameter channelId missing';
            }
            if (!args.options.messageId) {
                return 'Required parameter messageId missing';
            }
            if (!Utils_1.default.isValidGuid(args.options.teamId)) {
                return `${args.options.teamId} is not a valid GUID`;
            }
            if (!Utils_1.default.isValidTeamsChannelId(args.options.channelId)) {
                return `${args.options.channelId} is not a valid Teams ChannelId`;
            }
            return true;
        };
    }
    commandHelp(args, log) {
        const chalk = vorpal.chalk;
        log(vorpal.find(this.name).helpInformation());
        log(`  Remarks:

    ${chalk.yellow('Attention:')} This command is based on an API that is currently
    in preview and is subject to change once the API reached general
    availability.

    You can only retrieve replies to a message from a Microsoft Teams team
    if you are a member of that team.

  Examples:
  
    Retrieve the replies from a specified message from a channel of the
    Microsoft Teams team
      ${this.name} --teamId 5f5d7b71-1161-44d8-bcc1-3da710eb4171 --channelId 19:88f7e66a8dfe42be92db19505ae912a8@thread.skype --messageId 1540747442203
`);
    }
}
module.exports = new TeamsMessageReplyListCommand();
