// @pnp/office365-cli/dist/o365/spo/commands/listitem/listitem-record-undeclare.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const config_1 = require("../../../../config");
const commands_1 = require("../../commands");
const request_1 = require("../../../../request");
const SpoCommand_1 = require("../../../base/SpoCommand");
const Utils_1 = require("../../../../Utils");
const ClientSvc_1 = require("../../ClientSvc");
const vorpal = require('../../../../vorpal-init');
class SpoListItemRecordUndeclareCommand extends SpoCommand_1.default {
    get name() {
        return commands_1.default.LISTITEM_RECORD_UNDECLARE;
    }
    get description() {
        return 'Undeclares list item as a record';
    }
    getTelemetryProperties(args) {
        const telemetryProps = super.getTelemetryProperties(args);
        telemetryProps.listId = typeof args.options.listId !== 'undefined';
        telemetryProps.listTitle = typeof args.options.listTitle !== 'undefined';
        return telemetryProps;
    }
    commandAction(cmd, args, cb) {
        const clientSvcCommons = new ClientSvc_1.ClientSvc(cmd, this.debug);
        const listIdArgument = args.options.listId || '';
        const listTitleArgument = args.options.listTitle || '';
        const listRestUrl = (args.options.listId ?
            `${args.options.webUrl}/_api/web/lists(guid'${encodeURIComponent(listIdArgument)}')`
            : `${args.options.webUrl}/_api/web/lists/getByTitle('${encodeURIComponent(listTitleArgument)}')`);
        let formDigestValue = '';
        let environmentListId = '';
        (() => {
            if (typeof args.options.listId !== 'undefined') {
                return Promise.resolve({ value: args.options.listId });
            }
            if (this.verbose) {
                cmd.log(`Getting list id...`);
            }
            const listRequestOptions = {
                url: `${listRestUrl}/id`,
                headers: {
                    'accept': 'application/json;odata=nometadata'
                },
                json: true
            };
            return request_1.default.get(listRequestOptions);
        })()
            .then((res) => {
            environmentListId = res.value;
            if (this.debug) {
                cmd.log(`getting request digest for request`);
            }
            return this.getRequestDigest(args.options.webUrl);
        })
            .then((res) => {
            formDigestValue = res.FormDigestValue;
            return clientSvcCommons.getCurrentWebIdentity(args.options.webUrl, formDigestValue);
        })
            .then((objectIdentity) => {
            if (this.verbose) {
                cmd.log(`Undeclare list item as a record in list ${args.options.listId || args.options.listTitle} in site ${args.options.webUrl}...`);
            }
            const requestOptions = {
                url: `${args.options.webUrl}/_vti_bin/client.svc/ProcessQuery`,
                headers: {
                    'Content-Type': 'text/xml',
                    'X-RequestDigest': formDigestValue,
                },
                body: `<Request AddExpandoFieldTypeSuffix="true" SchemaVersion="15.0.0.0" LibraryVersion="16.0.0.0" ApplicationName="${config_1.default.applicationName}" xmlns="http://schemas.microsoft.com/sharepoint/clientquery/2009"><Actions><StaticMethod TypeId="{ea8e1356-5910-4e69-bc05-d0c30ed657fc}" Name="UndeclareItemAsRecord" Id="53"><Parameters><Parameter ObjectPathId="49" /></Parameters></StaticMethod></Actions><ObjectPaths><Identity Id="49" Name="${objectIdentity.objectIdentity}:list:${environmentListId}:item:${args.options.id},1" /></ObjectPaths></Request>`
            };
            return request_1.default.post(requestOptions);
        })
            .then(() => {
            // REST post call doesn't return anything
            cb();
        }, (err) => this.handleRejectedPromise(err, cmd, cb));
    }
    ;
    options() {
        const options = [
            {
                option: '-u, --webUrl <webUrl>',
                description: 'The URL of the site where the list is located'
            },
            {
                option: '-i, --id <id>',
                description: 'ID of the list item to be undeclared as a record.'
            },
            {
                option: '-l, --listId [listId]',
                description: 'The ID of the list where the item is located. Specify listId or listTitle but not both'
            },
            {
                option: '-t, --listTitle [listTitle]',
                description: 'The title of the list where the item is located. Specify listId or listTitle but not both'
            }
        ];
        const parentOptions = super.options();
        return options.concat(parentOptions);
    }
    validate() {
        return (args) => {
            if (!args.options.webUrl) {
                return 'Required parameter webUrl missing';
            }
            if (!args.options.id) {
                return 'Required parameter id missing';
            }
            const id = parseInt(args.options.id);
            if (isNaN(id)) {
                return `${args.options.id} is not a valid list item ID`;
            }
            const isValidSharePointUrl = SpoCommand_1.default.isValidSharePointUrl(args.options.webUrl);
            if (isValidSharePointUrl !== true) {
                return isValidSharePointUrl;
            }
            if (!args.options.listId && !args.options.listTitle) {
                return `Specify listId or listTitle`;
            }
            if (args.options.listId && args.options.listTitle) {
                return `Specify listId or listTitle but not both`;
            }
            if (args.options.listId &&
                !Utils_1.default.isValidGuid(args.options.listId)) {
                return `${args.options.listId} in option listId is not a valid GUID`;
            }
            return true;
        };
    }
    commandHelp(args, log) {
        const chalk = vorpal.chalk;
        log(vorpal.find(this.name).helpInformation());
        log(`  Examples:

    Undeclare the list item as a record with ID ${chalk.grey(1)} from list with ID
    ${chalk.grey('0cd891ef-afce-4e55-b836-fce03286cccf')} located in site
    ${chalk.grey('https://contoso.sharepoint.com/sites/project-x')} 
      ${commands_1.default.LISTITEM_RECORD_UNDECLARE} --webUrl https://contoso.sharepoint.com/sites/project-x --listId 0cd891ef-afce-4e55-b836-fce03286cccf --id 1

    Undeclare the list item as a record with ID ${chalk.grey(1)} from list with title
    ${chalk.grey('List 1')} located in site ${chalk.grey('https://contoso.sharepoint.com/sites/project-x')} 
      ${commands_1.default.LISTITEM_RECORD_UNDECLARE} --webUrl https://contoso.sharepoint.com/sites/project-x --listTitle 'List 1' --id 1
     `);
    }
}
module.exports = new SpoListItemRecordUndeclareCommand();
