// @pnp/office365-cli/dist/o365/aad/commands/user/user-list.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const commands_1 = require("../../commands");
const GraphItemsListCommand_1 = require("../../../base/GraphItemsListCommand");
const vorpal = require('../../../../vorpal-init');
class AadUserListCommand extends GraphItemsListCommand_1.GraphItemsListCommand {
    get name() {
        return `${commands_1.default.USER_LIST}`;
    }
    get description() {
        return 'Lists users matching specified criteria';
    }
    allowUnknownOptions() {
        return true;
    }
    getTelemetryProperties(args) {
        const telemetryProps = super.getTelemetryProperties(args);
        telemetryProps.properties = args.options.properties;
        return telemetryProps;
    }
    commandAction(cmd, args, cb) {
        const properties = args.options.properties ?
            args.options.properties.split(',').map(p => p.trim()) :
            ['userPrincipalName', 'displayName'];
        const filter = this.getFilter(args.options);
        const url = `${this.resource}/v1.0/users?$select=${properties.join(',')}${(filter.length > 0 ? '&' + filter : '')}&$top=100`;
        this
            .getAllItems(url, cmd, true)
            .then(() => {
            cmd.log(this.items);
            if (this.verbose) {
                cmd.log(vorpal.chalk.green('DONE'));
            }
            cb();
        }, (err) => this.handleRejectedODataJsonPromise(err, cmd, cb));
    }
    getFilter(options) {
        const filters = {};
        const excludeOptions = [
            'properties',
            'debug',
            'verbose',
            'output'
        ];
        Object.keys(options).forEach(key => {
            if (excludeOptions.indexOf(key) === -1) {
                filters[key] = encodeURIComponent(options[key].replace(/'/g, `''`));
            }
        });
        let filter = Object.keys(filters).map(key => `startsWith(${key}, '${filters[key]}')`).join(' and ');
        if (filter.length > 0) {
            filter = '$filter=' + filter;
        }
        return filter;
    }
    options() {
        const options = [
            {
                option: '-p, --properties [properties]',
                description: 'Comma-separated list of properties to retrieve'
            }
        ];
        const parentOptions = super.options();
        return options.concat(parentOptions);
    }
    commandHelp(args, log) {
        const chalk = vorpal.chalk;
        log(vorpal.find(this.name).helpInformation());
        log(`  Remarks:

    Using the ${chalk.blue('--properties')} option, you can specify
    a comma-separated list of user properties to retrieve from the Microsoft
    Graph. If you don't specify any properties, the command will retrieve
    user's display name and account name.

    To filter the list of users, include additional options that match the user
    property that you want to filter with. For example
    ${chalk.blue('--displayName Patt')} will return all users whose displayName
    starts with ${chalk.grey('Patt')}. Multiple filters will be combined using
    the ${chalk.blue('and')} operator.

  Examples:

    List all users in the tenant
      ${this.name}

    List all users in the tenant. For each one return the display name and
    e-mail address
      ${this.name} --properties "displayName,mail"

    Show users whose display name starts with ${chalk.grey('Patt')}
      ${this.name} --displayName Patt

    Show all account managers whose display name starts with ${chalk.grey('Patt')}
      ${this.name} --displayName Patt --jobTitle 'Account manager'

  More information:

    Microsoft Graph User properties
      https://developer.microsoft.com/en-us/graph/docs/api-reference/v1.0/resources/user#properties
`);
    }
}
module.exports = new AadUserListCommand();
