// @pnp/office365-cli/dist/o365/teams/commands/app/app-list.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const commands_1 = require("../../commands");
const GraphItemsListCommand_1 = require("../../../base/GraphItemsListCommand");
const Utils_1 = require("../../../../Utils");
const vorpal = require('../../../../vorpal-init');
class TeamsAppListCommand extends GraphItemsListCommand_1.GraphItemsListCommand {
    get name() {
        return `${commands_1.default.TEAMS_APP_LIST}`;
    }
    get description() {
        return 'Lists apps from the Microsoft Teams app catalog or apps installed in the specified team';
    }
    getTelemetryProperties(args) {
        const telemetryProps = super.getTelemetryProperties(args);
        telemetryProps.all = args.options.all || false;
        telemetryProps.teamId = typeof args.options.teamId !== 'undefined';
        return telemetryProps;
    }
    commandAction(cmd, args, cb) {
        let endpoint = '';
        if (args.options.teamId) {
            endpoint = `${this.resource}/v1.0/teams/${encodeURIComponent(args.options.teamId)}/installedApps?$expand=teamsApp`;
            if (!args.options.all) {
                endpoint += `&$filter=teamsApp/distributionMethod eq 'organization'`;
            }
        }
        else {
            endpoint = `${this.resource}/v1.0/appCatalogs/teamsApps`;
            if (!args.options.all) {
                endpoint += `?$filter=distributionMethod eq 'organization'`;
            }
        }
        this
            .getAllItems(endpoint, cmd, true)
            .then(() => {
            if (args.options.output === 'json') {
                cmd.log(this.items);
            }
            else {
                if (args.options.teamId) {
                    cmd.log(this.items.map(i => {
                        return {
                            id: i.id,
                            displayName: i.teamsApp.displayName,
                            distributionMethod: i.teamsApp.distributionMethod
                        };
                    }));
                }
                else {
                    cmd.log(this.items.map(i => {
                        return {
                            id: i.id,
                            displayName: i.displayName,
                            distributionMethod: i.distributionMethod
                        };
                    }));
                }
            }
            if (this.verbose) {
                cmd.log(vorpal.chalk.green('DONE'));
            }
            cb();
        }, (err) => this.handleRejectedODataJsonPromise(err, cmd, cb));
    }
    options() {
        const options = [
            {
                option: '-a, --all',
                description: 'Specify, to get apps from your organization only'
            },
            {
                option: '-i, --teamId [teamId]',
                description: 'The ID of the team for which to list installed apps'
            }
        ];
        const parentOptions = super.options();
        return options.concat(parentOptions);
    }
    validate() {
        return (args) => {
            if (args.options.teamId && !Utils_1.default.isValidGuid(args.options.teamId)) {
                return `${args.options.teamId} is not a valid GUID`;
            }
            return true;
        };
    }
    commandHelp(args, log) {
        const chalk = vorpal.chalk;
        log(vorpal.find(this.name).helpInformation());
        log(`  Remarks:

    To list apps installed in the specified Microsoft Teams team, specify that
    team's ID using the ${chalk.grey('teamId')} option. If the ${chalk.grey('teamId')} option
    is not specified, the command will list apps available in the Teams app
    catalog.

  Examples:

    List all Microsoft Teams apps from your organization's app catalog only
      ${commands_1.default.TEAMS_APP_LIST}
         
    List all apps from the Microsoft Teams app catalog and the Microsoft Teams
    store
      ${commands_1.default.TEAMS_APP_LIST} --all

    List your organization's apps installed in the specified Microsoft Teams
    team
      ${commands_1.default.TEAMS_APP_LIST} --teamId 6f6fd3f7-9ba5-4488-bbe6-a789004d0d55
`);
    }
}
module.exports = new TeamsAppListCommand();
