// @pnp/office365-cli/dist/o365/teams/commands/app/app-publish.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const request_1 = require("../../../../request");
const fs = require("fs");
const path = require("path");
const commands_1 = require("../../commands");
const GraphCommand_1 = require("../../../base/GraphCommand");
const vorpal = require('../../../../vorpal-init');
class TeamsAppPublishCommand extends GraphCommand_1.default {
    get name() {
        return `${commands_1.default.TEAMS_APP_PUBLISH}`;
    }
    get description() {
        return 'Publishes Teams app to the organization\'s app catalog';
    }
    commandAction(cmd, args, cb) {
        const fullPath = path.resolve(args.options.filePath);
        if (this.verbose) {
            cmd.log(`Adding app '${fullPath}' to app catalog...`);
        }
        const requestOptions = {
            url: `${this.resource}/v1.0/appCatalogs/teamsApps`,
            headers: {
                "content-type": "application/zip",
                accept: 'application/json;odata.metadata=none'
            },
            body: fs.readFileSync(fullPath)
        };
        request_1.default
            .post(requestOptions)
            .then((res) => {
            if (res && res.id) {
                cmd.log(res.id);
            }
            if (this.verbose) {
                cmd.log(vorpal.chalk.green('DONE'));
            }
            cb();
        }, (res) => this.handleRejectedODataJsonPromise(res, cmd, cb));
    }
    options() {
        const options = [
            {
                option: '-p, --filePath <filePath>',
                description: 'Absolute or relative path to the Teams manifest zip file to add to the app catalog'
            }
        ];
        const parentOptions = super.options();
        return options.concat(parentOptions);
    }
    validate() {
        return (args) => {
            if (!args.options.filePath) {
                return 'Missing required option filePath';
            }
            const fullPath = path.resolve(args.options.filePath);
            if (!fs.existsSync(fullPath)) {
                return `File '${fullPath}' not found`;
            }
            if (fs.lstatSync(fullPath).isDirectory()) {
                return `Path '${fullPath}' points to a directory`;
            }
            return true;
        };
    }
    commandHelp(args, log) {
        const chalk = vorpal.chalk;
        log(vorpal.find(this.name).helpInformation());
        log(`  Remarks:

    You can only publish a Teams app as a global administrator.

  Examples:

    Add the ${chalk.grey('teams-manifest.zip')} file to the organization's app catalog
      ${this.name} --filePath ./teams-manifest.zip
`);
    }
}
module.exports = new TeamsAppPublishCommand();
