// @pnp/office365-cli/dist/o365/teams/commands/app/app-uninstall.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const request_1 = require("../../../../request");
const Utils_1 = require("../../../../Utils");
const commands_1 = require("../../commands");
const GraphCommand_1 = require("../../../base/GraphCommand");
const vorpal = require('../../../../vorpal-init');
class TeamsAppUninstallCommand extends GraphCommand_1.default {
    get name() {
        return `${commands_1.default.TEAMS_APP_UNINSTALL}`;
    }
    get description() {
        return 'Uninstalls an app from a Microsoft Team team';
    }
    commandAction(cmd, args, cb) {
        const uninstallApp = () => {
            const requestOptions = {
                url: `${this.resource}/v1.0/teams/${args.options.teamId}/installedApps/${args.options.appId}`,
                headers: {
                    accept: 'application/json;odata.metadata=none'
                }
            };
            request_1.default
                .delete(requestOptions)
                .then(() => {
                if (this.verbose) {
                    cmd.log(vorpal.chalk.green('DONE'));
                }
                cb();
            }, (res) => this.handleRejectedODataJsonPromise(res, cmd, cb));
        };
        if (args.options.confirm) {
            uninstallApp();
        }
        else {
            cmd.prompt({
                type: 'confirm',
                name: 'continue',
                default: false,
                message: `Are you sure you want to uninstall the app with id ${args.options.appId} from the Microsoft Teams team ${args.options.teamId}?`,
            }, (result) => {
                if (!result.continue) {
                    cb();
                }
                else {
                    uninstallApp();
                }
            });
        }
    }
    options() {
        const options = [
            {
                option: '--appId <appId>',
                description: 'The unique id of the app instance installed in the Microsoft Teams team'
            },
            {
                option: '--teamId <teamId>',
                description: 'The ID of the Microsoft Teams team from which to uninstall the app'
            },
            {
                option: '--confirm',
                description: 'Don\'t prompt for confirmation when uninstalling the app'
            }
        ];
        const parentOptions = super.options();
        return options.concat(parentOptions);
    }
    validate() {
        return (args) => {
            if (!args.options.appId) {
                return 'Required parameter appId missing';
            }
            if (!args.options.teamId) {
                return 'Required parameter teamId missing';
            }
            if (!Utils_1.default.isValidGuid(args.options.teamId)) {
                return `${args.options.teamId} is not a valid GUID`;
            }
            return true;
        };
    }
    commandHelp(args, log) {
        const chalk = vorpal.chalk;
        log(vorpal.find(this.name).helpInformation());
        log(`  Remarks:

    The ${chalk.grey(`appId`)} has to be the id the app instance installed in the Microsoft Teams
    team. Do not use the ID from the manifest of the zip app package or the id
    from the Microsoft Teams App Catalog.

  Examples:

    Uninstall an app from a Microsoft Teams team
      ${this.name} --appId YzUyN2E0NzAtYTg4Mi00ODFjLTk4MWMtZWU2ZWZhYmE4NWM3IyM0ZDFlYTA0Ny1mMTk2LTQ1MGQtYjJlOS0wZDI4NTViYTA1YTY= --teamId 2609af39-7775-4f94-a3dc-0dd67657e900
`);
    }
}
module.exports = new TeamsAppUninstallCommand();
