// @pnp/office365-cli/dist/o365/teams/commands/app/app-update.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const request_1 = require("../../../../request");
const fs = require("fs");
const path = require("path");
const Utils_1 = require("../../../../Utils");
const commands_1 = require("../../commands");
const GraphCommand_1 = require("../../../base/GraphCommand");
const vorpal = require('../../../../vorpal-init');
class TeamsAppUpdateCommand extends GraphCommand_1.default {
    get name() {
        return `${commands_1.default.TEAMS_APP_UPDATE}`;
    }
    get description() {
        return 'Updates Teams app in the organization\'s app catalog';
    }
    commandAction(cmd, args, cb) {
        const { id: appId, filePath } = args.options;
        const fullPath = path.resolve(filePath);
        if (this.verbose) {
            cmd.log(`Updating app with id '${appId}' and file '${fullPath}' in the app catalog...`);
        }
        const requestOptions = {
            url: `${this.resource}/v1.0/appCatalogs/teamsApps/${appId}`,
            headers: {
                "content-type": "application/zip"
            },
            body: fs.readFileSync(fullPath)
        };
        request_1.default
            .put(requestOptions)
            .then(() => {
            if (this.verbose) {
                cmd.log(vorpal.chalk.green('DONE'));
            }
            cb();
        }, (res) => this.handleRejectedODataJsonPromise(res, cmd, cb));
    }
    options() {
        const options = [
            {
                option: '-i, --id <id>',
                description: 'ID of the app to update'
            },
            {
                option: '-p, --filePath <filePath>',
                description: 'Absolute or relative path to the Teams manifest zip file to update in the app catalog'
            }
        ];
        const parentOptions = super.options();
        return options.concat(parentOptions);
    }
    validate() {
        return (args) => {
            if (!args.options.id) {
                return 'Missing required option id';
            }
            if (!Utils_1.default.isValidGuid(args.options.id)) {
                return `${args.options.id} is not a valid GUID`;
            }
            if (!args.options.filePath) {
                return 'Missing required option filePath';
            }
            const fullPath = path.resolve(args.options.filePath);
            if (!fs.existsSync(fullPath)) {
                return `File '${fullPath}' not found`;
            }
            if (fs.lstatSync(fullPath).isDirectory()) {
                return `Path '${fullPath}' points to a directory`;
            }
            return true;
        };
    }
    commandHelp(args, log) {
        const chalk = vorpal.chalk;
        log(vorpal.find(this.name).helpInformation());
        log(`  Remarks:

    You can only update a Teams app as a global administrator.

  Examples:

    Update the Teams app with ID ${chalk.grey('83cece1e-938d-44a1-8b86-918cf6151957')}
    from file ${chalk.grey('teams-manifest.zip')}
      ${this.name} --id 83cece1e-938d-44a1-8b86-918cf6151957 --filePath ./teams-manifest.zip
`);
    }
}
module.exports = new TeamsAppUpdateCommand();
