// @pnp/office365-cli/dist/o365/teams/commands/app/app-remove.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const request_1 = require("../../../../request");
const Utils_1 = require("../../../../Utils");
const commands_1 = require("../../commands");
const GraphCommand_1 = require("../../../base/GraphCommand");
const vorpal = require('../../../../vorpal-init');
class TeamsAppRemoveCommand extends GraphCommand_1.default {
    get name() {
        return `${commands_1.default.TEAMS_APP_REMOVE}`;
    }
    get description() {
        return 'Removes a Teams app from the organization\'s app catalog';
    }
    getTelemetryProperties(args) {
        const telemetryProps = super.getTelemetryProperties(args);
        telemetryProps.confirm = (!(!args.options.confirm)).toString();
        return telemetryProps;
    }
    commandAction(cmd, args, cb) {
        const { id: appId } = args.options;
        const removeApp = () => {
            const requestOptions = {
                url: `${this.resource}/v1.0/appCatalogs/teamsApps/${appId}`,
                headers: {
                    accept: 'application/json;odata.metadata=none'
                }
            };
            request_1.default
                .delete(requestOptions)
                .then(() => {
                if (this.verbose) {
                    cmd.log(vorpal.chalk.green('DONE'));
                }
                cb();
            }, (res) => this.handleRejectedODataJsonPromise(res, cmd, cb));
        };
        if (args.options.confirm) {
            removeApp();
        }
        else {
            cmd.prompt({
                type: 'confirm',
                name: 'continue',
                default: false,
                message: `Are you sure you want to remove the Teams app ${appId} from the app catalog?`,
            }, (result) => {
                if (!result.continue) {
                    cb();
                }
                else {
                    removeApp();
                }
            });
        }
    }
    options() {
        const options = [
            {
                option: '-i, --id <id>',
                description: 'ID of the Teams app to remove. Needs to be available in your organization\'s app catalog.'
            },
            {
                option: '--confirm',
                description: 'Don\'t prompt for confirming removing the app'
            }
        ];
        const parentOptions = super.options();
        return options.concat(parentOptions);
    }
    validate() {
        return (args) => {
            if (!args.options.id) {
                return 'Required parameter id missing';
            }
            if (!Utils_1.default.isValidGuid(args.options.id)) {
                return `${args.options.id} is not a valid GUID`;
            }
            return true;
        };
    }
    commandHelp(args, log) {
        const chalk = vorpal.chalk;
        log(vorpal.find(this.name).helpInformation());
        log(`  Remarks:

    You can only remove a Teams app as a global administrator.

  Examples:

    Remove the Teams app with ID ${chalk.grey('83cece1e-938d-44a1-8b86-918cf6151957')} from
    the organization's app catalog. Will prompt for confirmation before actually
    removing the app.
      ${this.name} --id 83cece1e-938d-44a1-8b86-918cf6151957

    Remove the Teams app with ID ${chalk.grey('83cece1e-938d-44a1-8b86-918cf6151957')} from
    the organization's app catalog. Don't prompt for confirmation.
      ${this.name} --id 83cece1e-938d-44a1-8b86-918cf6151957 --confirm
`);
    }
}
module.exports = new TeamsAppRemoveCommand();
