// @pnp/office365-cli/dist/o365/spo/commands/hubsite/hubsite-data-get.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const request_1 = require("../../../../request");
const commands_1 = require("../../commands");
const SpoCommand_1 = require("../../../base/SpoCommand");
const vorpal = require('../../../../vorpal-init');
class SpoHubSiteDataGetCommand extends SpoCommand_1.default {
    get name() {
        return `${commands_1.default.HUBSITE_DATA_GET}`;
    }
    get description() {
        return 'Get hub site data for the specified site';
    }
    getTelemetryProperties(args) {
        const telemetryProps = super.getTelemetryProperties(args);
        telemetryProps.forceRefresh = args.options.forceRefresh === true;
        return telemetryProps;
    }
    commandAction(cmd, args, cb) {
        if (this.verbose) {
            cmd.log('Retrieving hub site data...');
        }
        const forceRefresh = args.options.forceRefresh === true;
        const requestOptions = {
            url: `${args.options.webUrl}/_api/web/HubSiteData(${forceRefresh})`,
            headers: {
                accept: 'application/json;odata=nometadata'
            },
            json: true
        };
        request_1.default
            .get(requestOptions)
            .then((res) => {
            if (res['odata.null'] !== true) {
                cmd.log(JSON.parse(res.value));
            }
            else {
                if (this.verbose) {
                    cmd.log(`${args.options.webUrl} is not connected to a hub site and is not a hub site itself`);
                }
            }
            if (this.verbose) {
                cmd.log(vorpal.chalk.green('DONE'));
            }
            cb();
        }, (err) => this.handleRejectedODataJsonPromise(err, cmd, cb));
    }
    options() {
        const options = [
            {
                option: '-u, --webUrl <webUrl>',
                description: 'URL of the site for which to retrieve hub site data'
            },
            {
                option: '-f, --forceRefresh',
                description: `Set, to refresh the server cache with the latest updates`
            }
        ];
        const parentOptions = super.options();
        return options.concat(parentOptions);
    }
    validate() {
        return (args) => {
            if (!args.options.webUrl) {
                return 'Required parameter webUrl missing';
            }
            const isValidSharePointUrl = SpoCommand_1.default.isValidSharePointUrl(args.options.webUrl);
            if (isValidSharePointUrl !== true) {
                return isValidSharePointUrl;
            }
            return true;
        };
    }
    commandHelp(args, log) {
        const chalk = vorpal.chalk;
        log(vorpal.find(this.name).helpInformation());
        log(`  Remarks:

    ${chalk.yellow('Attention:')} This command is based on a SharePoint API that is currently
    in preview and is subject to change once the API reached general
    availability.

    By default, the hub site data is returned from the server's cache.
    To refresh the data with the latest updates, use the ${chalk.blue('-f, --forceRefresh')}
    option. Use this option, if you just made changes and need to see them right
    away.

    If the specified site is not connected to a hub site site and is not a hub
    site itself, no data will be retrieved.

  Examples:
  
    Get information about the hub site data for a site with URL ${chalk.grey('https://contoso.sharepoint.com/sites/project-x')}
      ${this.name} --webUrl https://contoso.sharepoint.com/sites/project-x

  More information:

    SharePoint hub sites new in Office 365
      https://techcommunity.microsoft.com/t5/SharePoint-Blog/SharePoint-hub-sites-new-in-Office-365/ba-p/109547
`);
    }
}
module.exports = new SpoHubSiteDataGetCommand();
