// @pnp/office365-cli/dist/o365/spo/commands/hubsite/hubsite-rights-revoke.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const config_1 = require("../../../../config");
const request_1 = require("../../../../request");
const commands_1 = require("../../commands");
const Command_1 = require("../../../../Command");
const SpoCommand_1 = require("../../../base/SpoCommand");
const Utils_1 = require("../../../../Utils");
const vorpal = require('../../../../vorpal-init');
class SpoHubSiteRightsRevokeCommand extends SpoCommand_1.default {
    get name() {
        return `${commands_1.default.HUBSITE_RIGHTS_REVOKE}`;
    }
    get description() {
        return 'Revokes rights to join sites to the specified hub site for one or more principals';
    }
    getTelemetryProperties(args) {
        const telemetryProps = super.getTelemetryProperties(args);
        telemetryProps.confirm = (!(!args.options.confirm)).toString();
        return telemetryProps;
    }
    commandAction(cmd, args, cb) {
        const revokeRights = () => {
            let spoAdminUrl = '';
            if (this.verbose) {
                cmd.log(`Revoking rights for ${args.options.principals} from ${args.options.url}...`);
            }
            this
                .getSpoAdminUrl(cmd, this.debug)
                .then((_spoAdminUrl) => {
                spoAdminUrl = _spoAdminUrl;
                return this.getRequestDigest(spoAdminUrl);
            })
                .then((res) => {
                const principals = args.options.principals
                    .split(',')
                    .map(p => `<Object Type="String">${Utils_1.default.escapeXml(p.trim())}</Object>`)
                    .join('');
                const requestOptions = {
                    url: `${spoAdminUrl}/_vti_bin/client.svc/ProcessQuery`,
                    headers: {
                        'X-RequestDigest': res.FormDigestValue
                    },
                    body: `<Request AddExpandoFieldTypeSuffix="true" SchemaVersion="15.0.0.0" LibraryVersion="16.0.0.0" ApplicationName="${config_1.default.applicationName}" xmlns="http://schemas.microsoft.com/sharepoint/clientquery/2009"><Actions><ObjectPath Id="10" ObjectPathId="9" /><Method Name="RevokeHubSiteRights" Id="11" ObjectPathId="9"><Parameters><Parameter Type="String">${Utils_1.default.escapeXml(args.options.url)}</Parameter><Parameter Type="Array">${principals}</Parameter></Parameters></Method></Actions><ObjectPaths><Constructor Id="9" TypeId="{268004ae-ef6b-4e9b-8425-127220d84719}" /></ObjectPaths></Request>`
                };
                return request_1.default.post(requestOptions);
            })
                .then((res) => {
                const json = JSON.parse(res);
                const response = json[0];
                if (response.ErrorInfo) {
                    cb(new Command_1.CommandError(response.ErrorInfo.ErrorMessage));
                    return;
                }
                else {
                    if (this.verbose) {
                        cmd.log(vorpal.chalk.green('DONE'));
                    }
                }
                cb();
            }, (err) => this.handleRejectedPromise(err, cmd, cb));
        };
        if (args.options.confirm) {
            revokeRights();
        }
        else {
            cmd.prompt({
                type: 'confirm',
                name: 'continue',
                default: false,
                message: `Are you sure you want to revoke rights to join sites to the hub site ${args.options.url} from the specified users?`,
            }, (result) => {
                if (!result.continue) {
                    cb();
                }
                else {
                    revokeRights();
                }
            });
        }
    }
    options() {
        const options = [
            {
                option: '-u, --url <url>',
                description: 'The URL of the hub site to revoke rights on'
            },
            {
                option: '-p, --principals <principals>',
                description: 'Comma-separated list of principals to revoke join rights. Principals can be users or mail-enabled security groups in the form of "alias" or "alias@<domain name>.com"'
            },
            {
                option: '--confirm',
                description: 'Don\'t prompt for confirming revoking rights'
            }
        ];
        const parentOptions = super.options();
        return options.concat(parentOptions);
    }
    validate() {
        return (args) => {
            if (!args.options.url) {
                return 'Required parameter url missing';
            }
            const isValidSharePointUrl = SpoCommand_1.default.isValidSharePointUrl(args.options.url);
            if (isValidSharePointUrl !== true) {
                return isValidSharePointUrl;
            }
            if (!args.options.principals) {
                return 'Required parameter principals missing';
            }
            return true;
        };
    }
    commandHelp(args, log) {
        const chalk = vorpal.chalk;
        log(vorpal.find(commands_1.default.HUBSITE_RIGHTS_REVOKE).helpInformation());
        log(`  ${chalk.yellow('Important:')} to use this command you have to have permissions to access
    the tenant admin site.

  Remarks:

    ${chalk.yellow('Attention:')} This command is based on a SharePoint API that is currently
    in preview and is subject to change once the API reached general
    availability.

  Examples:

    Revoke rights to join sites to the hub site with URL
    ${chalk.grey('https://contoso.sharepoint.com/sites/sales')} from user with alias ${chalk.grey('PattiF')}.
    Will prompt for confirmation before revoking the rights
      ${this.name} --url https://contoso.sharepoint.com/sites/sales --principals PattiF

    Revoke rights to join sites to the hub site with URL
    ${chalk.grey('https://contoso.sharepoint.com/sites/sales')} from user with aliases ${chalk.grey('PattiF')}
    and ${chalk.grey('AdeleV')} without prompting for confirmation
      ${this.name} --url https://contoso.sharepoint.com/sites/sales --principals "PattiF,AdeleV" --confirm

  More information:

    SharePoint hub sites new in Office 365
      https://techcommunity.microsoft.com/t5/SharePoint-Blog/SharePoint-hub-sites-new-in-Office-365/ba-p/109547
`);
    }
}
module.exports = new SpoHubSiteRightsRevokeCommand();
