// @pnp/office365-cli/dist/o365/spo/commands/hubsite/hubsite-list.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const request_1 = require("../../../../request");
const commands_1 = require("../../commands");
const SpoCommand_1 = require("../../../base/SpoCommand");
const vorpal = require('../../../../vorpal-init');
class SpoHubSiteListCommand extends SpoCommand_1.default {
    constructor() {
        super();
        this.batchSize = 30;
        this.batchSize = 30;
    }
    get name() {
        return `${commands_1.default.HUBSITE_LIST}`;
    }
    get description() {
        return 'Lists hub sites in the current tenant';
    }
    getTelemetryProperties(args) {
        const telemetryProps = super.getTelemetryProperties(args);
        telemetryProps.includeAssociatedSites = args.options.includeAssociatedSites === true;
        return telemetryProps;
    }
    commandAction(cmd, args, cb) {
        let hubSites;
        let spoAdminUrl = '';
        this
            .getSpoAdminUrl(cmd, this.debug)
            .then((_spoAdminUrl) => {
            spoAdminUrl = _spoAdminUrl;
            const requestOptions = {
                url: `${spoAdminUrl}/_api/hubsites`,
                headers: {
                    accept: 'application/json;odata=nometadata'
                },
                json: true
            };
            return request_1.default.get(requestOptions);
        })
            .then((res) => {
            hubSites = res.value;
            if (args.options.includeAssociatedSites !== true || args.options.output !== 'json') {
                return Promise.resolve();
            }
            else {
                if (this.debug) {
                    cmd.log('Retrieving associated sites...');
                    cmd.log('');
                }
            }
            const requestOptions = {
                url: `${spoAdminUrl}/_api/web/lists/GetByTitle('DO_NOT_DELETE_SPLIST_TENANTADMIN_AGGREGATED_SITECOLLECTIONS')/RenderListDataAsStream`,
                headers: {
                    accept: 'application/json;odata=nometadata'
                },
                json: true,
                body: {
                    parameters: {
                        ViewXml: "<View><Query><Where><And><And><IsNull><FieldRef Name=\"TimeDeleted\"/></IsNull><Neq><FieldRef Name=\"State\"/><Value Type='Integer'>0</Value></Neq></And><Neq><FieldRef Name=\"HubSiteId\"/><Value Type='Text'>{00000000-0000-0000-0000-000000000000}</Value></Neq></And></Where><OrderBy><FieldRef Name='Title' Ascending='true' /></OrderBy></Query><ViewFields><FieldRef Name=\"Title\"/><FieldRef Name=\"SiteUrl\"/><FieldRef Name=\"SiteId\"/><FieldRef Name=\"HubSiteId\"/></ViewFields><RowLimit Paged=\"TRUE\">" + this.batchSize + "</RowLimit></View>",
                        DatesInUtc: true
                    }
                }
            };
            if (this.debug) {
                cmd.log(`Will retrieve associated sites (including the hub sites) in batches of ${this.batchSize}`);
            }
            return this.getSites(requestOptions, requestOptions.url, cmd);
        })
            .then((res) => {
            if (res) {
                hubSites.forEach(h => {
                    const filteredSites = res.filter(f => {
                        // Only include sites of which the Site Id is not the same as the
                        // Hub Site ID (as this site is the actual hub site) and of which the
                        // Hub Site ID matches the ID of the Hub
                        return f.SiteId !== f.HubSiteId
                            && f.HubSiteId.toUpperCase() == `{${h.ID.toUpperCase()}}`;
                    });
                    h.AssociatedSites = filteredSites.map(a => {
                        return {
                            Title: a.Title,
                            SiteUrl: a.SiteUrl
                        };
                    });
                });
            }
            ;
            if (args.options.output === 'json') {
                cmd.log(hubSites);
            }
            else {
                cmd.log(hubSites.map(h => {
                    return {
                        ID: h.ID,
                        SiteUrl: h.SiteUrl,
                        Title: h.Title
                    };
                }));
            }
            if (this.verbose) {
                cmd.log(vorpal.chalk.green('DONE'));
            }
            cb();
        }, (err) => this.handleRejectedODataJsonPromise(err, cmd, cb));
    }
    getSites(reqOptions, nonPagedUrl, cmd, sites = [], batchNumber = 0) {
        return new Promise((resolve, reject) => {
            request_1.default
                .post(reqOptions)
                .then((res) => {
                batchNumber++;
                const retrievedSites = res.Row.length > 0 ? sites.concat(res.Row) : sites;
                if (this.debug) {
                    cmd.log(res);
                    cmd.log(`Retrieved ${res.Row.length} sites in batch ${batchNumber}`);
                }
                if (!!res.NextHref) {
                    reqOptions.url = nonPagedUrl + res.NextHref;
                    if (this.debug) {
                        cmd.log(`Url for next batch of sites: ${reqOptions.url}`);
                    }
                    this
                        .getSites(reqOptions, nonPagedUrl, cmd, retrievedSites, batchNumber)
                        .then((associatedSites) => {
                        resolve(associatedSites);
                    }, (err) => {
                        reject(err);
                    });
                }
                else {
                    if (this.debug) {
                        cmd.log(`Retrieved ${retrievedSites.length} sites in total`);
                    }
                    resolve(retrievedSites);
                }
            }, (err) => {
                reject(err);
            });
        });
    }
    options() {
        const options = [
            {
                option: '-i, --includeAssociatedSites',
                description: `Include the associated sites in the result (only in JSON output)`
            }
        ];
        const parentOptions = super.options();
        return options.concat(parentOptions);
    }
    commandHelp(args, log) {
        const chalk = vorpal.chalk;
        log(vorpal.find(this.name).helpInformation());
        log(`  Remarks:

    ${chalk.yellow('Attention:')} This command is based on a SharePoint API that is currently
    in preview and is subject to change once the API reached general
    availability.

    When using the text output type (default), the command lists only the
    values of the ${chalk.grey('ID')}, ${chalk.grey('SiteUrl')} and ${chalk.grey('Title')} properties of the hub site. When setting
    the output type to JSON, all available properties are included in
    the command output.

  Examples:
  
    List hub sites in the current tenant
      ${this.name}

    List hub sites, including their associated sites, in the current tenant. Associated site info is only shown in JSON output.
      ${this.name} --includeAssociatedSites --output json

  More information:

    SharePoint hub sites new in Office 365
      https://techcommunity.microsoft.com/t5/SharePoint-Blog/SharePoint-hub-sites-new-in-Office-365/ba-p/109547
`);
    }
}
module.exports = new SpoHubSiteListCommand();
