// @pnp/office365-cli/dist/o365/spo/commands/apppage/apppage-add.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const request_1 = require("../../../../request");
const commands_1 = require("../../commands");
const SpoCommand_1 = require("../../../base/SpoCommand");
const vorpal = require('../../../../vorpal-init');
class SpoAppPageAddCommand extends SpoCommand_1.default {
    get name() {
        return `${commands_1.default.APPPAGE_ADD}`;
    }
    getTelemetryProperties(args) {
        const telemetryProps = super.getTelemetryProperties(args);
        telemetryProps.addToQuickLaunch = args.options.addToQuickLaunch;
        return telemetryProps;
    }
    get description() {
        return 'Creates a single-part app page';
    }
    commandAction(cmd, args, cb) {
        const requestOptions = {
            url: `${args.options.webUrl}/_api/sitepages/Pages/CreateFullPageApp`,
            headers: {
                'content-type': 'application/json;odata=nometadata',
                accept: 'application/json;odata=nometadata'
            },
            json: true,
            body: {
                title: args.options.title,
                addToQuickLaunch: args.options.addToQuickLaunch ? true : false,
                webPartDataAsJson: args.options.webPartData
            }
        };
        request_1.default
            .post(requestOptions)
            .then((res) => {
            cmd.log(res);
            if (this.verbose) {
                cmd.log(vorpal.chalk.green('DONE'));
            }
            cb();
        }, (err) => this.handleRejectedODataJsonPromise(err, cmd, cb));
    }
    options() {
        const options = [
            {
                option: '-u, --webUrl <webUrl>',
                description: 'The URL of the site where the page should be created'
            },
            {
                option: '-t, --title <title>',
                description: 'The title of the page to be created'
            },
            {
                option: '-d, --webPartData <webPartData>',
                description: 'JSON string of the web part to put on the page'
            },
            {
                option: '--addToQuickLaunch',
                description: 'Set, to add the page to the quick launch'
            }
        ];
        const parentOptions = super.options();
        return options.concat(parentOptions);
    }
    validate() {
        return (args) => {
            if (!args.options.webUrl) {
                return 'Required parameter webUrl missing';
            }
            if (!args.options.title) {
                return 'Required parameter title missing';
            }
            if (!args.options.webPartData) {
                return 'Required parameter webPartData missing';
            }
            try {
                JSON.parse(args.options.webPartData);
            }
            catch (e) {
                return `Specified webPartData is not a valid JSON string. Error: ${e}`;
            }
            return true;
        };
    }
    commandHelp(args, log) {
        const chalk = vorpal.chalk;
        log(vorpal.find(this.name).helpInformation());
        log(`  Remarks:

    If you want to add the single-part app page to quick launch, use the
    ${chalk.blue('addToQuickLaunch')} flag.

  Examples:
  
    Create a single-part app page in a site with url
    https://contoso.sharepoint.com, webpart data is stored in the
    ${chalk.grey('$webPartData')} variable
      ${this.name} --title "Contoso" --webUrl "https://contoso.sharepoint.com" --webPartData $webPartData --addToQuickLaunch 
`);
    }
}
module.exports = new SpoAppPageAddCommand();
