// @pnp/office365-cli/dist/o365/spfx/commands/project/project-upgrade/rules/FN006001_CFG_PS_schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Rule_1 = require("./Rule");
class FN006001_CFG_PS_schema extends Rule_1.Rule {
    constructor(schema) {
        super();
        this.schema = schema;
    }
    get id() {
        return 'FN006001';
    }
    get title() {
        return 'package-solution.json schema';
    }
    get description() {
        return `Update package-solution.json schema URL`;
    }
    ;
    get resolution() {
        return `{
  "$schema": "${this.schema}"
}`;
    }
    ;
    get resolutionType() {
        return 'json';
    }
    ;
    get severity() {
        return 'Required';
    }
    ;
    get file() {
        return './config/package-solution.json';
    }
    ;
    visit(project, findings) {
        if (!project.packageSolutionJson) {
            return;
        }
        if (project.packageSolutionJson.$schema !== this.schema) {
            this.addFinding(findings);
        }
    }
}
exports.FN006001_CFG_PS_schema = FN006001_CFG_PS_schema;
