// @pnp/office365-cli/dist/o365/spfx/commands/project/project-upgrade/rules/FN006003_CFG_PS_isDomainIsolated.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Rule_1 = require("./Rule");
class FN006003_CFG_PS_isDomainIsolated extends Rule_1.Rule {
    constructor(isDomainIsolated) {
        super();
        this.isDomainIsolated = isDomainIsolated;
    }
    get id() {
        return 'FN006003';
    }
    get title() {
        return 'package-solution.json isDomainIsolated';
    }
    get description() {
        return `Update package-solution.json isDomainIsolated`;
    }
    ;
    get resolution() {
        return `{
  "solution": {
    "isDomainIsolated": ${this.isDomainIsolated}
  }
}`;
    }
    ;
    get resolutionType() {
        return 'json';
    }
    ;
    get severity() {
        return 'Required';
    }
    ;
    get file() {
        return './config/package-solution.json';
    }
    ;
    visit(project, findings) {
        if (!project.packageSolutionJson ||
            !project.packageSolutionJson.solution) {
            return;
        }
        if (project.packageSolutionJson.solution.isDomainIsolated !== this.isDomainIsolated) {
            this.addFinding(findings);
        }
    }
}
exports.FN006003_CFG_PS_isDomainIsolated = FN006003_CFG_PS_isDomainIsolated;
