// @pnp/office365-cli/dist/o365/spfx/commands/project/project-upgrade/rules/FN004001_CFG_CA_schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Rule_1 = require("./Rule");
class FN004001_CFG_CA_schema extends Rule_1.Rule {
    constructor(schema) {
        super();
        this.schema = schema;
    }
    get id() {
        return 'FN004001';
    }
    get title() {
        return 'copy-assets.json schema';
    }
    get description() {
        return `Update copy-assets.json schema URL`;
    }
    ;
    get resolution() {
        return `{
  "$schema": "${this.schema}"
}`;
    }
    ;
    get resolutionType() {
        return 'json';
    }
    ;
    get severity() {
        return 'Required';
    }
    ;
    get file() {
        return './config/copy-assets.json';
    }
    ;
    visit(project, findings) {
        if (!project.copyAssetsJson) {
            return;
        }
        if (project.copyAssetsJson.$schema !== this.schema) {
            this.addFinding(findings);
        }
    }
}
exports.FN004001_CFG_CA_schema = FN004001_CFG_CA_schema;
