// @pnp/office365-cli/dist/o365/spfx/commands/project/project-upgrade/rules/FN005001_CFG_DAS_schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Rule_1 = require("./Rule");
class FN005001_CFG_DAS_schema extends Rule_1.Rule {
    constructor(schema) {
        super();
        this.schema = schema;
    }
    get id() {
        return 'FN005001';
    }
    get title() {
        return 'deploy-azure-storage.json schema';
    }
    get description() {
        return `Update deploy-azure-storage.json schema URL`;
    }
    ;
    get resolution() {
        return `{
  "$schema": "${this.schema}"
}`;
    }
    ;
    get resolutionType() {
        return 'json';
    }
    ;
    get severity() {
        return 'Required';
    }
    ;
    get file() {
        return './config/deploy-azure-storage.json';
    }
    ;
    visit(project, findings) {
        if (!project.deployAzureStorageJson) {
            return;
        }
        if (project.deployAzureStorageJson.$schema !== this.schema) {
            this.addFinding(findings);
        }
    }
}
exports.FN005001_CFG_DAS_schema = FN005001_CFG_DAS_schema;
