// @pnp/office365-cli/dist/o365/spfx/commands/project/project-upgrade/rules/FN009001_CFG_WM_schema.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Rule_1 = require("./Rule");
class FN009001_CFG_WM_schema extends Rule_1.Rule {
    constructor(schema) {
        super();
        this.schema = schema;
    }
    get id() {
        return 'FN009001';
    }
    get title() {
        return 'write-manifests.json schema';
    }
    get description() {
        return `Update write-manifests.json schema URL`;
    }
    ;
    get resolution() {
        return `{
  "$schema": "${this.schema}"
}`;
    }
    ;
    get resolutionType() {
        return 'json';
    }
    ;
    get severity() {
        return 'Required';
    }
    ;
    get file() {
        return './config/write-manifests.json';
    }
    ;
    visit(project, findings) {
        if (!project.writeManifestsJson) {
            return;
        }
        if (project.writeManifestsJson.$schema !== this.schema) {
            this.addFinding(findings);
        }
    }
}
exports.FN009001_CFG_WM_schema = FN009001_CFG_WM_schema;
