// @pnp/office365-cli/dist/o365/spfx/commands/project/project-upgrade/rules/FN003005_CFG_localizedResource_pathLib.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Rule_1 = require("./Rule");
class FN003005_CFG_localizedResource_pathLib extends Rule_1.Rule {
    get id() {
        return 'FN003005';
    }
    get title() {
        return 'Update path of the localized resource';
    }
    get description() {
        return 'In the config.json file, update the path of the localized resource';
    }
    ;
    get resolution() {
        return '';
    }
    ;
    get resolutionType() {
        return 'json';
    }
    ;
    get severity() {
        return 'Required';
    }
    ;
    get file() {
        return './config/config.json';
    }
    ;
    visit(project, findings) {
        if (!project.configJson ||
            !project.configJson.localizedResources) {
            return;
        }
        const occurrences = [];
        Object.keys(project.configJson.localizedResources).forEach(k => {
            const path = project.configJson.localizedResources[k];
            if (path.indexOf('lib/') !== 0) {
                occurrences.push({
                    file: this.file,
                    resolution: JSON.stringify({ localizedResources: { k: `lib/${path}` } }, null, 2)
                });
            }
        });
        if (occurrences.length > 0) {
            this.addFindingWithOccurrences(occurrences, findings);
        }
    }
}
exports.FN003005_CFG_localizedResource_pathLib = FN003005_CFG_localizedResource_pathLib;
