// @pnp/office365-cli/dist/o365/spfx/commands/project/project-upgrade/rules/FN003004_CFG_entries.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Rule_1 = require("./Rule");
class FN003004_CFG_entries extends Rule_1.Rule {
    get id() {
        return 'FN003004';
    }
    get title() {
        return `config.json entries`;
    }
    get description() {
        return `Remove the "entries" property in ${this.file}`;
    }
    ;
    get resolution() {
        return '';
    }
    ;
    get resolutionType() {
        return 'json';
    }
    ;
    get severity() {
        return 'Required';
    }
    ;
    get file() {
        return './config/config.json';
    }
    ;
    visit(project, findings) {
        if (!project.configJson) {
            return;
        }
        const entries = project.configJson.entries;
        if (entries !== undefined) {
            this.addFindingWithOccurrences([{
                    file: this.file,
                    resolution: JSON.stringify({ entries: entries }, null, 2)
                }], findings);
        }
    }
}
exports.FN003004_CFG_entries = FN003004_CFG_entries;
