// @pnp/office365-cli/dist/o365/spfx/commands/project/project-upgrade/rules/FN012002_TSC_moduleResolution.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Rule_1 = require("./Rule");
class FN012002_TSC_moduleResolution extends Rule_1.Rule {
    constructor(moduleResolution) {
        super();
        this.moduleResolution = moduleResolution;
    }
    get id() {
        return 'FN012002';
    }
    get title() {
        return 'tsconfig.json moduleResolution';
    }
    get description() {
        return `Update moduleResolution in tsconfig.json`;
    }
    ;
    get resolution() {
        return `{
  "compilerOptions": {
    "moduleResolution": "${this.moduleResolution}"
  }
}`;
    }
    ;
    get resolutionType() {
        return 'json';
    }
    ;
    get severity() {
        return 'Required';
    }
    ;
    get file() {
        return './tsconfig.json';
    }
    ;
    visit(project, findings) {
        if (!project.tsConfigJson || !project.tsConfigJson.compilerOptions) {
            return;
        }
        if (project.tsConfigJson.compilerOptions.moduleResolution !== this.moduleResolution) {
            this.addFinding(findings);
        }
    }
}
exports.FN012002_TSC_moduleResolution = FN012002_TSC_moduleResolution;
