// @pnp/office365-cli/dist/o365/spfx/commands/project/project-upgrade/rules/FN012001_TSC_module.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Rule_1 = require("./Rule");
class FN012001_TSC_module extends Rule_1.Rule {
    constructor(module) {
        super();
        this.module = module;
    }
    get id() {
        return 'FN012001';
    }
    get title() {
        return 'tsconfig.json module';
    }
    get description() {
        return `Update module type in tsconfig.json`;
    }
    ;
    get resolution() {
        return `{
  "compilerOptions": {
    "module": "${this.module}"
  }
}`;
    }
    ;
    get resolutionType() {
        return 'json';
    }
    ;
    get severity() {
        return 'Required';
    }
    ;
    get file() {
        return './tsconfig.json';
    }
    ;
    visit(project, findings) {
        if (!project.tsConfigJson || !project.tsConfigJson.compilerOptions) {
            return;
        }
        if (project.tsConfigJson.compilerOptions.module !== this.module) {
            this.addFinding(findings);
        }
    }
}
exports.FN012001_TSC_module = FN012001_TSC_module;
