// @pnp/office365-cli/dist/o365/spfx/commands/project/project-upgrade/rules/FN012004_TSC_typeRoots_types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Rule_1 = require("./Rule");
class FN012004_TSC_typeRoots_types extends Rule_1.Rule {
    constructor() {
        super();
    }
    get id() {
        return 'FN012004';
    }
    get title() {
        return 'tsconfig.json typeRoots ./node_modules/@types';
    }
    get description() {
        return `Add ./node_modules/@types to typeRoots in tsconfig.json`;
    }
    ;
    get resolution() {
        return `{
  "compilerOptions": {
    "typeRoots": [
      "./node_modules/@types"
    ]
  }
}`;
    }
    ;
    get resolutionType() {
        return 'json';
    }
    ;
    get severity() {
        return 'Required';
    }
    ;
    get file() {
        return './tsconfig.json';
    }
    ;
    visit(project, findings) {
        if (!project.tsConfigJson || !project.tsConfigJson.compilerOptions) {
            return;
        }
        if (!project.tsConfigJson.compilerOptions.typeRoots ||
            project.tsConfigJson.compilerOptions.typeRoots.indexOf('./node_modules/@types') < 0) {
            this.addFinding(findings);
        }
    }
}
exports.FN012004_TSC_typeRoots_types = FN012004_TSC_typeRoots_types;
