// @pnp/office365-cli/dist/o365/spfx/commands/project/project-upgrade/rules/FN012013_TSC_exclude.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Rule_1 = require("./Rule");
class FN012013_TSC_exclude extends Rule_1.Rule {
    constructor(exclude) {
        super();
        this.exclude = exclude;
    }
    get id() {
        return 'FN012013';
    }
    get title() {
        return 'tsconfig.json exclude property';
    }
    get description() {
        return `Update tsconfig.json exclude property`;
    }
    ;
    get resolution() {
        return JSON.stringify({
            exclude: this.exclude
        }, null, 2);
    }
    ;
    get resolutionType() {
        return 'json';
    }
    ;
    get severity() {
        return 'Required';
    }
    ;
    get file() {
        return './tsconfig.json';
    }
    ;
    visit(project, findings) {
        if (!project.tsConfigJson) {
            return;
        }
        if (!project.tsConfigJson.exclude ||
            this.exclude.filter(e => project.tsConfigJson.exclude.indexOf(e) < 0).length > 0) {
            this.addFinding(findings);
        }
    }
}
exports.FN012013_TSC_exclude = FN012013_TSC_exclude;
