// @pnp/office365-cli/dist/o365/spfx/commands/project/project-upgrade/rules/FN012015_TSC_strictNullChecks.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Rule_1 = require("./Rule");
class FN012015_TSC_strictNullChecks extends Rule_1.Rule {
    constructor(strictNullChecks) {
        super();
        this.strictNullChecks = strictNullChecks;
    }
    get id() {
        return 'FN012015';
    }
    get title() {
        return 'tsconfig.json compiler options strictNullChecks';
    }
    get description() {
        return `Update tsconfig.json strictNullChecks value`;
    }
    ;
    get resolution() {
        return `{
  "compilerOptions": {
    "strictNullChecks": ${this.strictNullChecks}
  }
}`;
    }
    ;
    get resolutionType() {
        return 'json';
    }
    ;
    get severity() {
        return 'Required';
    }
    ;
    get file() {
        return './tsconfig.json';
    }
    ;
    visit(project, findings) {
        if (!project.tsConfigJson || !project.tsConfigJson.compilerOptions) {
            return;
        }
        if (project.tsConfigJson.compilerOptions.strictNullChecks !== this.strictNullChecks) {
            this.addFinding(findings);
        }
    }
}
exports.FN012015_TSC_strictNullChecks = FN012015_TSC_strictNullChecks;
