// @pnp/office365-cli/dist/o365/spfx/commands/project/project-upgrade/rules/FN010003_YORC_packageManager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Rule_1 = require("./Rule");
class FN010003_YORC_packageManager extends Rule_1.Rule {
    constructor(packageManager) {
        super();
        this.packageManager = packageManager;
    }
    get id() {
        return 'FN010003';
    }
    get title() {
        return '.yo-rc.json packageManager';
    }
    get description() {
        return `Update packageManager in .yo-rc.json`;
    }
    ;
    get resolution() {
        return `{
  "@microsoft/generator-sharepoint": {
    "packageManager": "${this.packageManager}"
  }
}`;
    }
    ;
    get resolutionType() {
        return 'json';
    }
    ;
    get severity() {
        return 'Recommended';
    }
    ;
    get file() {
        return './.yo-rc.json';
    }
    ;
    visit(project, findings) {
        if (!project.yoRcJson) {
            return;
        }
        if (project.yoRcJson["@microsoft/generator-sharepoint"].packageManager !== this.packageManager) {
            this.addFinding(findings);
        }
    }
}
exports.FN010003_YORC_packageManager = FN010003_YORC_packageManager;
