// @pnp/office365-cli/dist/o365/spfx/commands/project/project-upgrade/rules/FN010002_YORC_isCreatingSolution.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Rule_1 = require("./Rule");
class FN010002_YORC_isCreatingSolution extends Rule_1.Rule {
    constructor(value) {
        super();
        this.value = value;
    }
    get id() {
        return 'FN010002';
    }
    get title() {
        return '.yo-rc.json isCreatingSolution';
    }
    get description() {
        return `Update isCreatingSolution in .yo-rc.json`;
    }
    ;
    get resolution() {
        return `{
  "@microsoft/generator-sharepoint": {
    "isCreatingSolution": ${this.value.toString()}
  }
}`;
    }
    ;
    get resolutionType() {
        return 'json';
    }
    ;
    get severity() {
        return 'Recommended';
    }
    ;
    get file() {
        return './.yo-rc.json';
    }
    ;
    visit(project, findings) {
        if (!project.yoRcJson) {
            return;
        }
        if (project.yoRcJson["@microsoft/generator-sharepoint"].isCreatingSolution !== this.value) {
            this.addFinding(findings);
        }
    }
}
exports.FN010002_YORC_isCreatingSolution = FN010002_YORC_isCreatingSolution;
