// @pnp/office365-cli/dist/o365/spfx/commands/project/project-upgrade/rules/FN012012_TSC_include.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Rule_1 = require("./Rule");
class FN012012_TSC_include extends Rule_1.Rule {
    constructor(include) {
        super();
        this.include = include;
    }
    get id() {
        return 'FN012012';
    }
    get title() {
        return 'tsconfig.json include property';
    }
    get description() {
        return `Update tsconfig.json include property`;
    }
    ;
    get resolution() {
        return JSON.stringify({
            include: this.include
        }, null, 2);
    }
    ;
    get resolutionType() {
        return 'json';
    }
    ;
    get severity() {
        return 'Required';
    }
    ;
    get file() {
        return './tsconfig.json';
    }
    ;
    visit(project, findings) {
        if (!project.tsConfigJson) {
            return;
        }
        if (!project.tsConfigJson.include ||
            this.include.filter(i => project.tsConfigJson.include.indexOf(i) < 0).length > 0) {
            this.addFinding(findings);
        }
    }
}
exports.FN012012_TSC_include = FN012012_TSC_include;
