// @pnp/office365-cli/dist/o365/spfx/commands/project/project-upgrade/rules/FN012011_TSC_outDir.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Rule_1 = require("./Rule");
class FN012011_TSC_outDir extends Rule_1.Rule {
    constructor(outDir) {
        super();
        this.outDir = outDir;
    }
    get id() {
        return 'FN012011';
    }
    get title() {
        return 'tsconfig.json compiler options outDir';
    }
    get description() {
        return `Update tsconfig.json outDir value`;
    }
    ;
    get resolution() {
        return `{
  "compilerOptions": {
    "outDir": "${this.outDir}"
  }
}`;
    }
    ;
    get resolutionType() {
        return 'json';
    }
    ;
    get severity() {
        return 'Required';
    }
    ;
    get file() {
        return './tsconfig.json';
    }
    ;
    visit(project, findings) {
        if (!project.tsConfigJson || !project.tsConfigJson.compilerOptions) {
            return;
        }
        if (project.tsConfigJson.compilerOptions.outDir !== this.outDir) {
            this.addFinding(findings);
        }
    }
}
exports.FN012011_TSC_outDir = FN012011_TSC_outDir;
