// @pnp/office365-cli/dist/o365/spfx/commands/project/project-upgrade/rules/FN010001_YORC_version.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Rule_1 = require("./Rule");
class FN010001_YORC_version extends Rule_1.Rule {
    constructor(version) {
        super();
        this.version = version;
    }
    get id() {
        return 'FN010001';
    }
    get title() {
        return '.yo-rc.json version';
    }
    get description() {
        return `Update version in .yo-rc.json`;
    }
    ;
    get resolution() {
        return `{
  "@microsoft/generator-sharepoint": {
    "version": "${this.version}"
  }
}`;
    }
    ;
    get resolutionType() {
        return 'json';
    }
    ;
    get severity() {
        return 'Recommended';
    }
    ;
    get file() {
        return './.yo-rc.json';
    }
    ;
    visit(project, findings) {
        if (!project.yoRcJson) {
            return;
        }
        if (project.yoRcJson["@microsoft/generator-sharepoint"].version !== this.version) {
            this.addFinding(findings);
        }
    }
}
exports.FN010001_YORC_version = FN010001_YORC_version;
