// @pnp/office365-cli/dist/o365/spfx/commands/project/project-upgrade/rules/FN014004_CODE_settings_jsonSchemas_configJson_url.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Rule_1 = require("./Rule");
class FN014004_CODE_settings_jsonSchemas_configJson_url extends Rule_1.Rule {
    constructor(url) {
        super();
        this.url = url;
    }
    get id() {
        return 'FN014004';
    }
    get title() {
        return 'URL of the config.json JSON schema in .vscode/settings.json';
    }
    get description() {
        return `Update the URL of the config.json JSON schema in .vscode/settings.json`;
    }
    ;
    get resolution() {
        return `{
  "json.schemas": [
    {
      "fileMatch": [
        "/config/config.json"
      ],
      "url": "./node_modules/@microsoft/sp-build-core-tasks/lib/configJson/schemas/config-v1.schema.json"
    }
  ]
}`;
    }
    ;
    get resolutionType() {
        return 'json';
    }
    ;
    get severity() {
        return 'Required';
    }
    ;
    get file() {
        return '.vscode/settings.json';
    }
    ;
    visit(project, findings) {
        if (!project.vsCode ||
            !project.vsCode.settingsJson ||
            !project.vsCode.settingsJson["json.schemas"]) {
            return;
        }
        const schemas = project.vsCode.settingsJson["json.schemas"];
        for (let i = 0; i < schemas.length; i++) {
            const schema = schemas[i];
            if (schema.fileMatch.indexOf('/config/config.json') === -1) {
                continue;
            }
            if (schema.url !== this.url) {
                this.addFinding(findings);
            }
            return;
        }
    }
}
exports.FN014004_CODE_settings_jsonSchemas_configJson_url = FN014004_CODE_settings_jsonSchemas_configJson_url;
