// @pnp/office365-cli/dist/o365/spfx/commands/project/project-upgrade/rules/FN012007_TSC_lib_es5.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Rule_1 = require("./Rule");
class FN012007_TSC_lib_es5 extends Rule_1.Rule {
    constructor() {
        super();
    }
    get id() {
        return 'FN012007';
    }
    get title() {
        return 'tsconfig.json es5 lib';
    }
    get description() {
        return `Add es5 lib in tsconfig.json`;
    }
    ;
    get resolution() {
        return `{
  "compilerOptions": {
    "lib": [
      "es5"
    ]
  }
}`;
    }
    ;
    get resolutionType() {
        return 'json';
    }
    ;
    get severity() {
        return 'Required';
    }
    ;
    get file() {
        return './tsconfig.json';
    }
    ;
    visit(project, findings) {
        if (!project.tsConfigJson || !project.tsConfigJson.compilerOptions) {
            return;
        }
        if (!project.tsConfigJson.compilerOptions.lib ||
            project.tsConfigJson.compilerOptions.lib.indexOf('es5') < 0) {
            this.addFinding(findings);
        }
    }
}
exports.FN012007_TSC_lib_es5 = FN012007_TSC_lib_es5;
