// @pnp/office365-cli/dist/o365/spfx/commands/project/project-upgrade/rules/FN010005_YORC_environment.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Rule_1 = require("./Rule");
class FN010005_YORC_environment extends Rule_1.Rule {
    constructor(environment) {
        super();
        this.environment = environment;
    }
    get id() {
        return 'FN010005';
    }
    get title() {
        return '.yo-rc.json environment';
    }
    get description() {
        return `Update environment in .yo-rc.json`;
    }
    ;
    get resolution() {
        return `{
  "@microsoft/generator-sharepoint": {
    "environment": "${this.environment}"
  }
}`;
    }
    ;
    get resolutionType() {
        return 'json';
    }
    ;
    get severity() {
        return 'Recommended';
    }
    ;
    get file() {
        return './.yo-rc.json';
    }
    ;
    visit(project, findings) {
        if (!project.yoRcJson) {
            return;
        }
        if (project.yoRcJson["@microsoft/generator-sharepoint"].environment !== this.environment) {
            this.addFinding(findings);
        }
    }
}
exports.FN010005_YORC_environment = FN010005_YORC_environment;
