// @pnp/office365-cli/dist/o365/spfx/commands/project/project-upgrade/rules/FN010007_YORC_isDomainIsolated.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Rule_1 = require("./Rule");
class FN010007_YORC_isDomainIsolated extends Rule_1.Rule {
    constructor(value) {
        super();
        this.value = value;
    }
    get id() {
        return 'FN010007';
    }
    get title() {
        return '.yo-rc.json isDomainIsolated';
    }
    get description() {
        return `Update isDomainIsolated in .yo-rc.json`;
    }
    ;
    get resolution() {
        return `{
  "@microsoft/generator-sharepoint": {
    "isDomainIsolated": ${this.value.toString()}
  }
}`;
    }
    ;
    get resolutionType() {
        return 'json';
    }
    ;
    get severity() {
        return 'Recommended';
    }
    ;
    get file() {
        return './.yo-rc.json';
    }
    ;
    visit(project, findings) {
        if (!project.yoRcJson) {
            return;
        }
        if (project.yoRcJson["@microsoft/generator-sharepoint"].isDomainIsolated !== this.value) {
            this.addFinding(findings);
        }
    }
}
exports.FN010007_YORC_isDomainIsolated = FN010007_YORC_isDomainIsolated;
