// @pnp/office365-cli/dist/o365/spfx/commands/project/project-upgrade/rules/FN018003_TEAMS_tab20x20_png.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Rule_1 = require("./Rule");
const path = require("path");
const fs = require("fs");
class FN018003_TEAMS_tab20x20_png extends Rule_1.Rule {
    constructor() {
        super();
    }
    get id() {
        return 'FN018003';
    }
    get title() {
        return 'Web part Microsoft Teams tab small icon';
    }
    get description() {
        return 'Create Microsoft Teams tab small icon for the web part';
    }
    get resolution() {
        return '';
    }
    ;
    get resolutionType() {
        return 'cmd';
    }
    get file() {
        return '';
    }
    ;
    get severity() {
        return 'Optional';
    }
    visit(project, findings) {
        if (!project.manifests ||
            project.manifests.length < 1) {
            return;
        }
        const webPartManifests = project.manifests.filter(m => m.componentType === 'WebPart');
        if (webPartManifests.length < 1) {
            return;
        }
        const occurrences = [];
        webPartManifests.forEach(manifest => {
            const webPartFolderName = path.basename(path.dirname(manifest.path));
            const teamsFolderName = `teams_${webPartFolderName}`;
            const teamsFolderPath = path.join(project.path, teamsFolderName);
            const iconPath = path.join(teamsFolderPath, 'tab20x20.png');
            if (!fs.existsSync(iconPath)) {
                occurrences.push({
                    file: path.relative(project.path, iconPath),
                    resolution: `copy_cmd ${path.join(__dirname, '..', 'assets', 'tab20x20.png')}DestinationParam${iconPath}`
                });
            }
        });
        if (occurrences.length > 0) {
            this.addFindingWithOccurrences(occurrences, findings);
        }
    }
}
exports.FN018003_TEAMS_tab20x20_png = FN018003_TEAMS_tab20x20_png;
