// @pnp/office365-cli/dist/o365/spfx/commands/project/project-upgrade/rules/FN012005_TSC_typeRoots_microsoft.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Rule_1 = require("./Rule");
class FN012005_TSC_typeRoots_microsoft extends Rule_1.Rule {
    constructor() {
        super();
    }
    get id() {
        return 'FN012005';
    }
    get title() {
        return 'tsconfig.json typeRoots ./node_modules/@microsoft';
    }
    get description() {
        return `Add ./node_modules/@microsoft to typeRoots in tsconfig.json`;
    }
    ;
    get resolution() {
        return `{
  "compilerOptions": {
    "typeRoots": [
      "./node_modules/@microsoft"
    ]
  }
}`;
    }
    ;
    get resolutionType() {
        return 'json';
    }
    ;
    get severity() {
        return 'Required';
    }
    ;
    get file() {
        return './tsconfig.json';
    }
    ;
    visit(project, findings) {
        if (!project.tsConfigJson || !project.tsConfigJson.compilerOptions) {
            return;
        }
        if (!project.tsConfigJson.compilerOptions.typeRoots ||
            project.tsConfigJson.compilerOptions.typeRoots.indexOf('./node_modules/@microsoft') < 0) {
            this.addFinding(findings);
        }
    }
}
exports.FN012005_TSC_typeRoots_microsoft = FN012005_TSC_typeRoots_microsoft;
