// @pnp/office365-cli/dist/o365/spfx/commands/project/project-upgrade/rules/FN014001_CODE_settings_jsonSchemas.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Rule_1 = require("./Rule");
class FN014001_CODE_settings_jsonSchemas extends Rule_1.Rule {
    constructor(add) {
        super();
        this.add = add;
    }
    get id() {
        return 'FN014001';
    }
    get title() {
        return 'json.schemas in .vscode/settings.json';
    }
    get description() {
        return `${(this.add ? 'Add' : 'Remove')} json.schemas in .vscode/settings.json`;
    }
    ;
    get resolution() {
        return `{
  "json.schemas": []
}`;
    }
    ;
    get resolutionType() {
        return 'json';
    }
    ;
    get severity() {
        return 'Required';
    }
    ;
    get file() {
        return '.vscode/settings.json';
    }
    ;
    visit(project, findings) {
        if (!project.vsCode || !project.vsCode.settingsJson) {
            return;
        }
        if (this.add) {
            if (!project.vsCode.settingsJson["json.schemas"]) {
                this.addFinding(findings);
            }
        }
        else {
            if (project.vsCode.settingsJson["json.schemas"]) {
                this.addFinding(findings);
            }
        }
    }
}
exports.FN014001_CODE_settings_jsonSchemas = FN014001_CODE_settings_jsonSchemas;
