// @pnp/office365-cli/dist/o365/spfx/commands/project/project-upgrade/rules/FN012017_TSC_extends.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Rule_1 = require("./Rule");
class FN012017_TSC_extends extends Rule_1.Rule {
    // extends is a reserved word so _extends is used instead
    constructor(_extends) {
        super();
        this._extends = _extends;
    }
    get id() {
        return 'FN012017';
    }
    get title() {
        return 'tsconfig.json extends property';
    }
    get description() {
        return `Update tsconfig.json extends property`;
    }
    ;
    get resolution() {
        return `{
  "extends": "${this._extends}"
}`;
    }
    ;
    get resolutionType() {
        return 'json';
    }
    ;
    get severity() {
        return 'Required';
    }
    ;
    get file() {
        return './tsconfig.json';
    }
    ;
    visit(project, findings) {
        if (!project.tsConfigJson) {
            return;
        }
        if (!project.tsConfigJson.extends || project.tsConfigJson.extends !== this._extends) {
            this.addFinding(findings);
        }
    }
}
exports.FN012017_TSC_extends = FN012017_TSC_extends;
