// @pnp/office365-cli/dist/o365/spfx/commands/project/project-upgrade/rules/FN010004_YORC_componentType.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Rule_1 = require("./Rule");
class FN010004_YORC_componentType extends Rule_1.Rule {
    constructor(componentType) {
        super();
        this.componentType = componentType;
    }
    get id() {
        return 'FN010004';
    }
    get title() {
        return '.yo-rc.json componentType';
    }
    get description() {
        return `Update componentType in .yo-rc.json`;
    }
    ;
    get resolution() {
        return `{
  "@microsoft/generator-sharepoint": {
    "componentType": "${this.componentType}"
  }
}`;
    }
    ;
    get resolutionType() {
        return 'json';
    }
    ;
    get severity() {
        return 'Recommended';
    }
    ;
    get file() {
        return './.yo-rc.json';
    }
    ;
    visit(project, findings) {
        if (!project.yoRcJson) {
            return;
        }
        if (project.yoRcJson["@microsoft/generator-sharepoint"].componentType !== this.componentType) {
            this.addFinding(findings);
        }
    }
}
exports.FN010004_YORC_componentType = FN010004_YORC_componentType;
