// @pnp/office365-cli/dist/o365/spfx/commands/project/project-upgrade/rules/FN010006_YORC_framework.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Rule_1 = require("./Rule");
class FN010006_YORC_framework extends Rule_1.Rule {
    constructor(framework, add) {
        super();
        this.framework = framework;
        this.add = add;
    }
    get id() {
        return 'FN010006';
    }
    get title() {
        return '.yo-rc.json framework';
    }
    get description() {
        return `${this.add ? 'Update' : 'Remove'} framework in .yo-rc.json`;
    }
    ;
    get resolution() {
        return `{
  "@microsoft/generator-sharepoint": {
    "framework": "${this.framework}"
  }
}`;
    }
    ;
    get resolutionType() {
        return 'json';
    }
    ;
    get severity() {
        return 'Recommended';
    }
    ;
    get file() {
        return './.yo-rc.json';
    }
    ;
    visit(project, findings) {
        if (!project.yoRcJson) {
            return;
        }
        if (this.add) {
            if (project.yoRcJson["@microsoft/generator-sharepoint"].framework !== this.framework) {
                this.addFinding(findings);
            }
        }
        else {
            if (project.yoRcJson["@microsoft/generator-sharepoint"].framework) {
                this.addFinding(findings);
            }
        }
    }
}
exports.FN010006_YORC_framework = FN010006_YORC_framework;
