// @pnp/office365-cli/dist/o365/spfx/commands/project/project-upgrade/rules/FN012006_TSC_types_es6_collections.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Rule_1 = require("./Rule");
class FN012006_TSC_types_es6_collections extends Rule_1.Rule {
    constructor(add) {
        super();
        this.add = add;
    }
    get id() {
        return 'FN012006';
    }
    get title() {
        return 'tsconfig.json es6-collections types';
    }
    get description() {
        return `${(this.add ? 'Add' : 'Remove')} es6-collections type in tsconfig.json`;
    }
    ;
    get resolution() {
        return `{
  "compilerOptions": {
    "types": [
      "es6-collections"
    ]
  }
}`;
    }
    ;
    get resolutionType() {
        return 'json';
    }
    ;
    get severity() {
        return 'Required';
    }
    ;
    get file() {
        return './tsconfig.json';
    }
    ;
    visit(project, findings) {
        if (!project.tsConfigJson || !project.tsConfigJson.compilerOptions) {
            return;
        }
        if (this.add) {
            if (!project.tsConfigJson.compilerOptions.types ||
                project.tsConfigJson.compilerOptions.types.indexOf('es6-collections') < 0) {
                this.addFinding(findings);
            }
        }
        else {
            if (project.tsConfigJson.compilerOptions.types &&
                project.tsConfigJson.compilerOptions.types.indexOf('es6-collections') > -1) {
                this.addFinding(findings);
            }
        }
    }
}
exports.FN012006_TSC_types_es6_collections = FN012006_TSC_types_es6_collections;
