// @pnp/office365-cli/dist/o365/spfx/commands/project/project-upgrade/rules/FN022002_SCSS_add_fabric_react.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const ScssRule_1 = require("./ScssRule");
const __1 = require("../");
class FN022002_SCSS_add_fabric_react extends ScssRule_1.ScssRule {
    constructor(importValue) {
        super();
        this.importValue = importValue;
    }
    get id() {
        return 'FN022002';
    }
    get title() {
        return `Scss file import`;
    }
    get description() {
        return `Add scss file import`;
    }
    ;
    get resolution() {
        return `@import '${this.importValue}'`;
    }
    ;
    get resolutionType() {
        return 'scss';
    }
    ;
    get severity() {
        return 'Optional';
    }
    ;
    get file() {
        return '';
    }
    ;
    visit(project, findings) {
        if (__1.Utils.isReactProject(project) === false) {
            return;
        }
        if (!project.scssFiles ||
            project.scssFiles.length === 0) {
            return;
        }
        const occurrences = [];
        project.scssFiles.forEach((file) => {
            if (file.source.indexOf(this.importValue) === -1) {
                this.addOccurrence(this.resolution, file.path, project.path, occurrences);
            }
        });
        if (occurrences.length > 0) {
            this.addFindingWithOccurrences(occurrences, findings);
        }
    }
}
exports.FN022002_SCSS_add_fabric_react = FN022002_SCSS_add_fabric_react;
