// @pnp/office365-cli/dist/o365/spfx/commands/project/project-upgrade/rules/PackageRule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Rule_1 = require("./Rule");
class PackageRule extends Rule_1.Rule {
    constructor(propertyName, add, propertyValue) {
        super();
        this.propertyName = propertyName;
        this.add = add;
        this.propertyValue = propertyValue;
    }
    get title() {
        return this.propertyName;
    }
    get description() {
        return `${this.add ? 'Add' : 'Remove'} package.json property`;
    }
    ;
    get resolution() {
        return `{
  "${this.propertyName}": "${this.propertyValue}"
}`;
    }
    ;
    get resolutionType() {
        return 'json';
    }
    ;
    get severity() {
        return 'Required';
    }
    ;
    get file() {
        return './package.json';
    }
    ;
    visit(project, findings) {
        if (!project.packageJson) {
            return;
        }
        if (!project.packageJson ||
            (this.add && !project.packageJson[this.propertyName]) ||
            (!this.add && project.packageJson[this.propertyName])) {
            this.addFinding(findings);
        }
    }
}
exports.PackageRule = PackageRule;
