// @pnp/office365-cli/dist/o365/aad/commands/oauth2grant/oauth2grant-remove.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const commands_1 = require("../../commands");
const request_1 = require("../../../../request");
const AadCommand_1 = require("../../../base/AadCommand");
const vorpal = require('../../../../vorpal-init');
class AadOAuth2GrantRemoveCommand extends AadCommand_1.default {
    get name() {
        return commands_1.default.OAUTH2GRANT_REMOVE;
    }
    get description() {
        return 'Remove specified service principal OAuth2 permissions';
    }
    commandAction(cmd, args, cb) {
        if (this.verbose) {
            cmd.log(`Removing OAuth2 permissions...`);
        }
        const requestOptions = {
            url: `${this.resource}/myorganization/oauth2PermissionGrants/${encodeURIComponent(args.options.grantId)}?api-version=1.6`,
            json: true
        };
        request_1.default
            .delete(requestOptions)
            .then(() => {
            if (this.verbose) {
                cmd.log(vorpal.chalk.green('DONE'));
            }
            cb();
        }, (rawRes) => this.handleRejectedODataJsonPromise(rawRes, cmd, cb));
    }
    options() {
        const options = [
            {
                option: '-i, --grantId <grantId>',
                description: 'objectId of OAuth2 permission grant to remove'
            }
        ];
        const parentOptions = super.options();
        return options.concat(parentOptions);
    }
    validate() {
        return (args) => {
            if (!args.options.grantId) {
                return 'Required option grantId missing';
            }
            return true;
        };
    }
    commandHelp(args, log) {
        const chalk = vorpal.chalk;
        log(vorpal.find(commands_1.default.OAUTH2GRANT_REMOVE).helpInformation());
        log(`  Remarks:
  
    Before you can remove service principal's OAuth2 permissions, you need to
    get the ${chalk.grey('objectId')} of the permissions grant to remove. You can retrieve it
    using the ${chalk.blue(commands_1.default.OAUTH2GRANT_LIST)} command.

    If the ${chalk.grey('objectId')} listed when using the ${chalk.blue(commands_1.default.OAUTH2GRANT_LIST)} command has a 
    minus sign ('-') prefix, you may receive an error indicating --grantId is
    missing. To resolve this issue simply  escape the leading '-',
    eg. ${chalk.blue(commands_1.default.OAUTH2GRANT_REMOVE)} --grantId \\-Zc1JRY8REeLxmXz5KtixAYU3Q6noCBPlhwGiX7pxmU     

  Examples:
  
    Remove the OAuth2 permission grant with ID ${chalk.grey('YgA60KYa4UOPSdc-lpxYEnQkr8KVLDpCsOXkiV8i-ek')}
      ${commands_1.default.OAUTH2GRANT_REMOVE} --grantId YgA60KYa4UOPSdc-lpxYEnQkr8KVLDpCsOXkiV8i-ek

  More information:
  
    Application and service principal objects in Azure Active Directory (Azure AD)
      https://docs.microsoft.com/en-us/azure/active-directory/develop/active-directory-application-objects
`);
    }
}
module.exports = new AadOAuth2GrantRemoveCommand();
