// @pnp/office365-cli/dist/o365/aad/commands/oauth2grant/oauth2grant-set.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const commands_1 = require("../../commands");
const request_1 = require("../../../../request");
const AadCommand_1 = require("../../../base/AadCommand");
const vorpal = require('../../../../vorpal-init');
class AadOAuth2GrantSetCommand extends AadCommand_1.default {
    get name() {
        return commands_1.default.OAUTH2GRANT_SET;
    }
    get description() {
        return 'Update OAuth2 permissions for the service principal';
    }
    commandAction(cmd, args, cb) {
        if (this.verbose) {
            cmd.log(`Updating OAuth2 permissions...`);
        }
        const requestOptions = {
            url: `${this.resource}/myorganization/oauth2PermissionGrants/${encodeURIComponent(args.options.grantId)}?api-version=1.6`,
            headers: {
                'content-type': 'application/json'
            },
            json: true,
            body: {
                "scope": args.options.scope
            }
        };
        request_1.default
            .patch(requestOptions)
            .then(() => {
            if (this.verbose) {
                cmd.log(vorpal.chalk.green('DONE'));
            }
            cb();
        }, (rawRes) => this.handleRejectedODataJsonPromise(rawRes, cmd, cb));
    }
    options() {
        const options = [
            {
                option: '-i, --grantId <grantId>',
                description: 'objectId of OAuth2 permission grant to update'
            },
            {
                option: '-s, --scope <scope>',
                description: 'Permissions to grant'
            }
        ];
        const parentOptions = super.options();
        return options.concat(parentOptions);
    }
    validate() {
        return (args) => {
            if (!args.options.grantId) {
                return 'Required option grantId missing';
            }
            if (!args.options.scope) {
                return 'Required option scope missing';
            }
            return true;
        };
    }
    commandHelp(args, log) {
        const chalk = vorpal.chalk;
        log(vorpal.find(commands_1.default.OAUTH2GRANT_SET).helpInformation());
        log(`  Remarks:
  
    Before you can update service principal's OAuth2 permissions, you need to
    get the ${chalk.grey('objectId')} of the permissions grant to update. You can retrieve it
    using the ${chalk.blue(commands_1.default.OAUTH2GRANT_LIST)} command.

    If the ${chalk.grey('objectId')} listed when using the ${chalk.blue(commands_1.default.OAUTH2GRANT_LIST)} command has a 
    minus sign ('-') prefix, you may receive an error indicating --grantId is
    missing. To resolve this issue simply escape the leading '-',
    eg. ${chalk.blue(commands_1.default.OAUTH2GRANT_SET)} --grantId \\-Zc1JRY8REeLxmXz5KtixAYU3Q6noCBPlhwGiX7pxmU --scope 'Calendars.Read'     
       
  Examples:
  
    Update the existing OAuth2 permission grant with ID ${chalk.grey('YgA60KYa4UOPSdc-lpxYEnQkr8KVLDpCsOXkiV8i-ek')}
    to the ${chalk.grey('Calendars.Read Mail.Read')} permissions
      ${commands_1.default.OAUTH2GRANT_SET} --grantId YgA60KYa4UOPSdc-lpxYEnQkr8KVLDpCsOXkiV8i-ek --scope "Calendars.Read Mail.Read"

  More information:
  
    Application and service principal objects in Azure Active Directory (Azure AD)
      https://docs.microsoft.com/en-us/azure/active-directory/develop/active-directory-application-objects
`);
    }
}
module.exports = new AadOAuth2GrantSetCommand();
