// @pnp/office365-cli/dist/o365/aad/commands/o365group/o365group-teamify.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const config_1 = require("../../../../config");
const commands_1 = require("../../commands");
const Utils_1 = require("../../../../Utils");
const request_1 = require("../../../../request");
const GraphCommand_1 = require("../../../base/GraphCommand");
const vorpal = require('../../../../vorpal-init');
class AadO365GroupTeamifyCommand extends GraphCommand_1.default {
    get name() {
        return `${commands_1.default.O365GROUP_TEAMIFY}`;
    }
    get description() {
        return 'Creates a new Microsoft Teams team under existing Office 365 group';
    }
    commandAction(cmd, args, cb) {
        const body = {
            "template@odata.bind": "https://graph.microsoft.com/beta/teamsTemplates('standard')",
            "group@odata.bind": `https://graph.microsoft.com/v1.0/groups('${encodeURIComponent(args.options.groupId)}')`
        };
        const requestOptions = {
            url: `${this.resource}/beta/teams`,
            headers: {
                accept: 'application/json;odata.metadata=none'
            },
            body: body,
            json: true
        };
        request_1.default
            .post(requestOptions)
            .then(() => {
            if (this.verbose) {
                cmd.log(vorpal.chalk.green('DONE'));
            }
            cb();
        }, (err) => this.handleRejectedODataJsonPromise(err, cmd, cb));
    }
    options() {
        const options = [
            {
                option: '-i, --groupId <groupId>',
                description: 'The ID of the Office 365 Group to connect to Microsoft Teams'
            }
        ];
        const parentOptions = super.options();
        return options.concat(parentOptions);
    }
    validate() {
        return (args) => {
            if (!args.options.groupId) {
                return 'Required parameter teamId missing';
            }
            if (!Utils_1.default.isValidGuid(args.options.groupId)) {
                return `${args.options.groupId} is not a valid GUID`;
            }
            return true;
        };
    }
    commandHelp(args, log) {
        const chalk = vorpal.chalk;
        log(vorpal.find(this.name).helpInformation());
        log(`  Remarks:

    ${chalk.yellow('Attention:')} This command is based on an API that is currently in preview
    and is subject to change once the API reached general availability.

  Examples:
  
    Creates a new Microsoft Teams team under existing Office 365 group
      ${chalk.grey(config_1.default.delimiter)} ${this.name} --groupId e3f60f99-0bad-481f-9e9f-ff0f572fbd03
`);
    }
}
module.exports = new AadO365GroupTeamifyCommand();
