// @pnp/office365-cli/dist/o365/aad/commands/o365group/o365group-user-remove.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const commands_1 = require("../../commands");
const commands_2 = require("../../../teams/commands");
const Utils_1 = require("../../../../Utils");
const request_1 = require("../../../../request");
const GraphCommand_1 = require("../../../base/GraphCommand");
const vorpal = require('../../../../vorpal-init');
class AadO365GroupUserRemoveCommand extends GraphCommand_1.default {
    get name() {
        return `${commands_1.default.O365GROUP_USER_REMOVE}`;
    }
    get description() {
        return 'Removes the specified user from specified Office 365 Group or Microsoft Teams team';
    }
    alias() {
        return [commands_2.default.TEAMS_USER_REMOVE];
    }
    getTelemetryProperties(args) {
        const telemetryProps = super.getTelemetryProperties(args);
        telemetryProps.confirm = (!(!args.options.confirm)).toString();
        telemetryProps.teamId = typeof args.options.teamId !== 'undefined';
        telemetryProps.groupId = typeof args.options.groupId !== 'undefined';
        return telemetryProps;
    }
    commandAction(cmd, args, cb) {
        let userId = '';
        const groupId = (typeof args.options.groupId !== 'undefined') ? args.options.groupId : args.options.teamId;
        const removeUser = () => {
            const requestOptions = {
                url: `${this.resource}/v1.0/users/${encodeURIComponent(args.options.userName)}/id`,
                headers: {
                    accept: 'application/json;odata.metadata=none'
                },
                json: true
            };
            request_1.default
                .get(requestOptions)
                .then((res) => {
                userId = res.value;
                const requestOptions = {
                    url: `${this.resource}/v1.0/groups/${groupId}/owners?$select=id,displayName,userPrincipalName,userType`,
                    headers: {
                        accept: 'application/json;odata.metadata=none'
                    },
                    json: true
                };
                return request_1.default.get(requestOptions);
            })
                .then((res) => {
                const userIsOwner = (res.value.filter((i) => i.userPrincipalName === args.options.userName).length > 0);
                const endpoint = `${this.resource}/v1.0/groups/${groupId}/${userIsOwner ? 'owners' : 'members'}/${userId}/$ref`;
                const requestOptions = {
                    url: endpoint,
                    headers: {
                        'accept': 'application/json;odata.metadata=none'
                    },
                };
                return request_1.default.delete(requestOptions);
            })
                .then(() => {
                if (this.verbose) {
                    cmd.log(vorpal.chalk.green('DONE'));
                }
                cb();
            }, (err) => this.handleRejectedODataJsonPromise(err, cmd, cb));
        };
        if (args.options.confirm) {
            removeUser();
        }
        else {
            cmd.prompt({
                type: 'confirm',
                name: 'continue',
                default: false,
                message: `Are you sure you want to remove ${args.options.userName} from the ${(typeof args.options.groupId !== 'undefined' ? 'group' : 'team')} ${groupId}?`,
            }, (result) => {
                if (!result.continue) {
                    cb();
                }
                else {
                    removeUser();
                }
            });
        }
    }
    options() {
        const options = [
            {
                option: "-i, --groupId [groupId]",
                description: "The ID of the Office 365 Group from which to remove the user"
            },
            {
                option: "--teamId [teamId]",
                description: "The ID of the Microsoft Teams team from which to remove the user"
            },
            {
                option: '-n, --userName <userName>',
                description: 'User\'s UPN (user principal name), eg. johndoe@example.com'
            },
            {
                option: '--confirm',
                description: 'Don\'t prompt for confirming removing user from the specified Office 365 Group or Microsoft Teams team'
            }
        ];
        const parentOptions = super.options();
        return options.concat(parentOptions);
    }
    validate() {
        return (args) => {
            if (!args.options.groupId && !args.options.teamId) {
                return 'Please provide one of the following parameters: groupId or teamId';
            }
            if (args.options.groupId && args.options.teamId) {
                return 'You cannot provide both a groupId and teamId parameter, please provide only one';
            }
            if (args.options.teamId && !Utils_1.default.isValidGuid(args.options.teamId)) {
                return `${args.options.teamId} is not a valid GUID`;
            }
            if (args.options.groupId && !Utils_1.default.isValidGuid(args.options.groupId)) {
                return `${args.options.groupId} is not a valid GUID`;
            }
            if (!args.options.userName) {
                return 'Required parameter userName missing';
            }
            return true;
        };
    }
    commandHelp(args, log) {
        log(vorpal.find(this.name).helpInformation());
        log(`  Remarks:

    You can remove users from a Office 365 Group or Microsoft Teams team if you
    are owner of that group or team.

  Examples:

    Removes user from the specified Office 365 Group
      ${this.name} --groupId '00000000-0000-0000-0000-000000000000' --userName 'anne.matthews@contoso.onmicrosoft.com'

    Removes user from the specified Office 365 Group without confirmation
      ${this.name} --groupId '00000000-0000-0000-0000-000000000000' --userName 'anne.matthews@contoso.onmicrosoft.com' --confirm

    Removes user from the specified Microsoft Teams team
      ${this.alias()[0]} --teamId '00000000-0000-0000-0000-000000000000' --userName 'anne.matthews@contoso.onmicrosoft.com'
`);
    }
}
module.exports = new AadO365GroupUserRemoveCommand();
