// @pnp/office365-cli/dist/o365/aad/commands/o365group/o365group-get.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const commands_1 = require("../../commands");
const request_1 = require("../../../../request");
const Utils_1 = require("../../../../Utils");
const GraphCommand_1 = require("../../../base/GraphCommand");
const vorpal = require('../../../../vorpal-init');
class AadO365GroupGetCommand extends GraphCommand_1.default {
    get name() {
        return `${commands_1.default.O365GROUP_GET}`;
    }
    get description() {
        return 'Gets information about the specified Office 365 Group';
    }
    commandAction(cmd, args, cb) {
        let group;
        const requestOptions = {
            url: `${this.resource}/v1.0/groups/${args.options.id}`,
            headers: {
                accept: 'application/json;odata.metadata=none'
            },
            json: true
        };
        request_1.default
            .get(requestOptions)
            .then((res) => {
            group = res;
            if (args.options.includeSiteUrl) {
                const requestOptions = {
                    url: `${this.resource}/v1.0/groups/${group.id}/drive?$select=webUrl`,
                    headers: {
                        accept: 'application/json;odata.metadata=none'
                    },
                    json: true
                };
                return request_1.default.get(requestOptions);
            }
            else {
                return Promise.resolve(undefined);
            }
        })
            .then((res) => {
            if (res) {
                group.siteUrl = res.webUrl ? res.webUrl.substr(0, res.webUrl.lastIndexOf('/')) : '';
            }
            cmd.log(group);
            if (this.verbose) {
                cmd.log(vorpal.chalk.green('DONE'));
            }
            cb();
        }, (err) => this.handleRejectedODataJsonPromise(err, cmd, cb));
    }
    options() {
        const options = [
            {
                option: '-i, --id <id>',
                description: 'The ID of the Office 365 Group to retrieve information for'
            },
            {
                option: '--includeSiteUrl',
                description: 'Set to retrieve the site URL for the group'
            }
        ];
        const parentOptions = super.options();
        return options.concat(parentOptions);
    }
    validate() {
        return (args) => {
            if (!args.options.id) {
                return 'Required parameter id missing';
            }
            if (!Utils_1.default.isValidGuid(args.options.id)) {
                return `${args.options.id} is not a valid GUID`;
            }
            return true;
        };
    }
    commandHelp(args, log) {
        const chalk = vorpal.chalk;
        log(vorpal.find(this.name).helpInformation());
        log(`  Examples:
  
    Get information about the Office 365 Group with id ${chalk.grey(`1caf7dcd-7e83-4c3a-94f7-932a1299c844`)}
      ${this.name} --id 1caf7dcd-7e83-4c3a-94f7-932a1299c844

    Get information about the Office 365 Group with id ${chalk.grey(`1caf7dcd-7e83-4c3a-94f7-932a1299c844`)}
    and also retrieve the URL of the corresponding SharePoint site
      ${this.name} --id 1caf7dcd-7e83-4c3a-94f7-932a1299c844 --includeSiteUrl
`);
    }
}
module.exports = new AadO365GroupGetCommand();
