// @pnp/office365-cli/dist/o365/commands/login.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Auth_1 = require("../../Auth");
const commands_1 = require("./commands");
const Command_1 = require("../../Command");
const Auth_2 = require("../../Auth");
const fs = require("fs");
const vorpal = require('../../vorpal-init');
class LoginCommand extends Command_1.default {
    get name() {
        return `${commands_1.default.LOGIN}`;
    }
    get description() {
        return 'Log in to Office 365';
    }
    getTelemetryProperties(args) {
        const telemetryProps = super.getTelemetryProperties(args);
        telemetryProps.authType = args.options.authType || 'deviceCode';
        return telemetryProps;
    }
    commandAction(cmd, args, cb) {
        const chalk = vorpal.chalk;
        // disconnect before re-connecting
        if (this.debug) {
            cmd.log(`Logging out from Office 365...`);
        }
        const logout = () => {
            Auth_1.default.service.logout();
            if (this.verbose) {
                cmd.log(chalk.green('DONE'));
            }
        };
        const login = () => {
            if (this.verbose) {
                cmd.log(`Signing in to Office 365...`);
            }
            switch (args.options.authType) {
                case 'password':
                    Auth_1.default.service.authType = Auth_2.AuthType.Password;
                    Auth_1.default.service.userName = args.options.userName;
                    Auth_1.default.service.password = args.options.password;
                    break;
                case 'certificate':
                    Auth_1.default.service.authType = Auth_2.AuthType.Certificate;
                    Auth_1.default.service.certificate = fs.readFileSync(args.options.certificateFile, 'base64');
                    Auth_1.default.service.thumbprint = args.options.thumbprint;
                    Auth_1.default.service.password = args.options.password;
                    break;
                case 'identity':
                    Auth_1.default.service.authType = Auth_2.AuthType.Identity;
                    Auth_1.default.service.userName = args.options.userName;
            }
            Auth_1.default
                .ensureAccessToken(Auth_1.default.defaultResource, cmd, this.debug)
                .then(() => {
                if (this.verbose) {
                    cmd.log(chalk.green('DONE'));
                }
                Auth_1.default.service.connected = true;
                cb();
            }, (rej) => {
                if (this.debug) {
                    cmd.log('Error:');
                    cmd.log(rej);
                    cmd.log('');
                }
                if (rej !== 'Polling_Request_Cancelled') {
                    cb(new Command_1.CommandError(rej));
                    return;
                }
                cb();
            });
        };
        Auth_1.default
            .clearConnectionInfo()
            .then(() => {
            logout();
            login();
        }, (error) => {
            if (this.debug) {
                cmd.log(new Command_1.CommandError(error));
            }
            logout();
            login();
        });
    }
    action() {
        const cmd = this;
        return function (args, cb) {
            Auth_1.default
                .restoreAuth()
                .then(() => {
                args = cmd.processArgs(args);
                cmd.initAction(args, this);
                cmd.commandAction(this, args, cb);
            }, (error) => {
                cb(new Command_1.CommandError(error));
            });
        };
    }
    cancel() {
        return () => {
            Auth_1.default.cancel();
        };
    }
    options() {
        const options = [
            {
                option: '-t, --authType [authType]',
                description: 'The type of authentication to use. Allowed values certificate|deviceCode|password|identity. Default deviceCode',
                autocomplete: ['certificate', 'deviceCode', 'password', 'identity']
            },
            {
                option: '-u, --userName [userName]',
                description: 'Name of the user to authenticate. Required when authType is set to password'
            },
            {
                option: '-p, --password [password]',
                description: 'Password for the user. Required when authType is set to password'
            },
            {
                option: '-c, --certificateFile [certificateFile]',
                description: 'Path to the file with certificate private key. Required when authType is set to certificate'
            },
            {
                option: '--thumbprint [thumbprint]',
                description: 'Certificate thumbprint. Required when authType is set to certificate'
            }
        ];
        const parentOptions = super.options();
        return options.concat(parentOptions);
    }
    validate() {
        return (args) => {
            if (args.options.authType === 'password') {
                if (!args.options.userName) {
                    return 'Required option userName missing';
                }
                if (!args.options.password) {
                    return 'Required option password missing';
                }
            }
            if (args.options.authType === 'certificate') {
                if (!args.options.certificateFile) {
                    return 'Required option certificateFile missing';
                }
                if (!fs.existsSync(args.options.certificateFile)) {
                    return `File '${args.options.certificateFile}' does not exist`;
                }
                if (!args.options.thumbprint) {
                    return 'Required option thumbprint missing';
                }
            }
            return true;
        };
    }
    commandHelp(args, log) {
        const chalk = vorpal.chalk;
        log(vorpal.find(commands_1.default.LOGIN).helpInformation());
        log(`  Remarks:
    
    Using the ${chalk.blue(commands_1.default.LOGIN)} command you log in to Office 365.

    By default, the ${chalk.blue(commands_1.default.LOGIN)} command uses device code OAuth flow
    to log in to the Microsoft Graph. Alternatively, you can
    authenticate using a user name and password or certificate, which are
    convenient for CI/CD scenarios, but which come with their own limitations.
    See the Office 365 CLI manual for more information.
    
    When logging in to Office 365, the ${chalk.blue(commands_1.default.LOGIN)} command stores
    in memory the access token and the refresh token. Both tokens are cleared
    from memory after exiting the CLI or by calling the ${chalk.blue(commands_1.default.LOGOUT)}
    command.

    When logging in to Office 365 using the user name and password, next to the
    access and refresh token, the Office 365 CLI will store the user credentials
    so that it can automatically re-authenticate if necessary. Similarly to the
    tokens, the credentials are removed by re-authenticating using the device
    code or by calling the ${chalk.blue(commands_1.default.LOGOUT)} command.

    When logging in to the Office 365 using a certificate, the Office 365 CLI
    will store the contents of the certificate so that it can automatically
    re-authenticate if necessary. The contents of the certificate are removed
    by re-authenticating using the device code or by calling
    the ${chalk.blue(commands_1.default.LOGOUT)} command.

    To log in to Office 365 using a certificate, you will typically create
    a custom Azure AD application. To use this application with
    the Office 365 CLI, you will set the ${chalk.grey('OFFICE365CLI_AADAPPID')}
    environment variable to the application's ID and the ${chalk.grey('OFFICE365CLI_TENANT')}
    environment variable to the ID of the Azure AD tenant, where you created
    the Azure AD application.

  Examples:
  
    Log in to Office 365 using the device code
      ${commands_1.default.LOGIN}

    Log in to Office 365 using the device code in debug mode including detailed
    debug information in the console output
      ${commands_1.default.LOGIN} --debug

    Log in to Office 365 using a user name and password
      ${commands_1.default.LOGIN} --authType password --userName user@contoso.com --password pass@word1

    Log in to Office 365 using a PEM certificate
      ${commands_1.default.LOGIN} --authType certificate --certificateFile /Users/user/dev/localhost.pem --thumbprint 47C4885736C624E90491F32B98855AA8A7562AF1

    Log in to Office 365 using a personal information exchange (.pfx) file
      o365 login --authType certificate --certificateFile /Users/user/dev/localhost.pfx --thumbprint 47C4885736C624E90491F32B98855AA8A7562AF1 --password 'pass@word1'

`);
    }
}
module.exports = new LoginCommand();
