// @pnp/office365-cli/dist/o365/commands/logout.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Auth_1 = require("../../Auth");
const commands_1 = require("./commands");
const Command_1 = require("../../Command");
const vorpal = require('../../vorpal-init');
class LogoutCommand extends Command_1.default {
    get name() {
        return commands_1.default.LOGOUT;
    }
    get description() {
        return 'Log out from Office 365';
    }
    commandAction(cmd, args, cb) {
        const chalk = vorpal.chalk;
        if (this.verbose) {
            cmd.log('Logging out from Office 365...');
        }
        const logout = () => {
            Auth_1.default.service.logout();
            if (this.verbose) {
                cmd.log(chalk.green('DONE'));
            }
            cb();
        };
        Auth_1.default
            .clearConnectionInfo()
            .then(() => {
            logout();
        }, (error) => {
            if (this.debug) {
                cmd.log(new Command_1.CommandError(error));
            }
            logout();
        });
    }
    action() {
        const cmd = this;
        return function (args, cb) {
            Auth_1.default
                .restoreAuth()
                .then(() => {
                args = cmd.processArgs(args);
                cmd.initAction(args, this);
                cmd.commandAction(this, args, cb);
            }, (error) => {
                cb(new Command_1.CommandError(error));
            });
        };
    }
    commandHelp(args, log) {
        const chalk = vorpal.chalk;
        log(vorpal.find(commands_1.default.LOGOUT).helpInformation());
        log(`  Remarks:

    The ${chalk.blue(commands_1.default.LOGOUT)} command logs out from Office 365
    and removes any access and refresh tokens from memory.

  Examples:
  
    Log out from Office 365
      ${commands_1.default.LOGOUT}

    Log out from Office 365 in debug mode including detailed debug
    information in the console output
      ${commands_1.default.LOGOUT} --debug
`);
    }
}
module.exports = new LogoutCommand();
