// @pnp/office365-cli/dist/o365/teams/commands/channel/channel-list.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const commands_1 = require("../../commands");
const GraphItemsListCommand_1 = require("../../../base/GraphItemsListCommand");
const Utils_1 = require("../../../../Utils");
const vorpal = require('../../../../vorpal-init');
class TeamsChannelListCommand extends GraphItemsListCommand_1.GraphItemsListCommand {
    get name() {
        return `${commands_1.default.TEAMS_CHANNEL_LIST}`;
    }
    get description() {
        return 'Lists channels in the specified Microsoft Teams team';
    }
    commandAction(cmd, args, cb) {
        const endpoint = `${this.resource}/v1.0/teams/${args.options.teamId}/channels`;
        this
            .getAllItems(endpoint, cmd, true)
            .then(() => {
            if (args.options.output === 'json') {
                cmd.log(this.items);
            }
            else {
                cmd.log(this.items.map(m => {
                    return {
                        id: m.id,
                        displayName: m.displayName
                    };
                }));
            }
            if (this.verbose) {
                cmd.log(vorpal.chalk.green('DONE'));
            }
            cb();
        }, (err) => this.handleRejectedODataJsonPromise(err, cmd, cb));
    }
    options() {
        const options = [
            {
                option: '-i, --teamId <teamId>',
                description: 'The ID of the team to list the channels of'
            }
        ];
        const parentOptions = super.options();
        return options.concat(parentOptions);
    }
    validate() {
        return (args) => {
            if (!args.options.teamId) {
                return 'Required parameter teamId missing';
            }
            if (!Utils_1.default.isValidGuid(args.options.teamId)) {
                return `${args.options.teamId} is not a valid GUID`;
            }
            return true;
        };
    }
    commandHelp(args, log) {
        log(vorpal.find(this.name).helpInformation());
        log(`  Examples:
  
    List the channels in a specified Microsoft Teams team
      ${this.name} --teamId 00000000-0000-0000-0000-000000000000
`);
    }
}
module.exports = new TeamsChannelListCommand();
