// @pnp/office365-cli/dist/o365/spo/commands/feature/feature-disable.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const SpoCommand_1 = require("../../../base/SpoCommand");
const commands_1 = require("../../commands");
const request_1 = require("../../../../request");
const vorpal = require('../../../../vorpal-init');
class SpoFeatureDisableCommand extends SpoCommand_1.default {
    get name() {
        return commands_1.default.FEATURE_DISABLE;
    }
    get description() {
        return 'Disables feature for the specified site or web';
    }
    getTelemetryProperties(args) {
        const telemetryProps = super.getTelemetryProperties(args);
        telemetryProps.scope = args.options.scope || 'web';
        telemetryProps.force = args.options.force || false;
        return telemetryProps;
    }
    commandAction(cmd, args, cb) {
        let scope = args.options.scope;
        let force = args.options.force;
        if (!scope) {
            scope = "web";
        }
        if (!force) {
            force = false;
        }
        if (this.verbose) {
            cmd.log(`Disabling feature '${args.options.featureId}' on scope '${scope}' for url '${args.options.url}' (force='${force}')...`);
        }
        const url = `${args.options.url}/_api/${scope}/features/remove(featureId=guid'${args.options.featureId}',force=${force})`;
        const requestOptions = {
            url: url,
            headers: {
                accept: 'application/json;odata=nometadata'
            }
        };
        request_1.default
            .post(requestOptions)
            .then((res) => {
            if (this.verbose) {
                cmd.log(vorpal.chalk.green('DONE'));
            }
            cb();
        }, (err) => this.handleRejectedODataJsonPromise(err, cmd, cb));
    }
    types() {
        return {
            string: ['scope', 's']
        };
    }
    options() {
        const options = [
            {
                option: '-u, --url <url>',
                description: 'The URL of the site or web for which to disable the feature'
            },
            {
                option: '-f, --featureId <id>',
                description: 'The ID of the feature to disable'
            },
            {
                option: '-s, --scope [scope]',
                description: 'Scope of the Feature to disable. Allowed values `Site|Web`. Default `Web`',
                autocomplete: ['Site', 'Web']
            },
            {
                option: '--force',
                description: 'Specifies whether to continue if an error occurs when disabling the feature'
            }
        ];
        const parentOptions = super.options();
        return options.concat(parentOptions);
    }
    validate() {
        return (args) => {
            if (!args.options.url) {
                return 'Required parameter url missing';
            }
            if (!args.options.featureId) {
                return 'Required parameter featureId missing';
            }
            if (args.options.scope) {
                if (['site', 'web'].indexOf(args.options.scope.toLowerCase()) < 0) {
                    return `${args.options.scope} is not a valid Feature scope. Allowed values are Site|Web`;
                }
            }
            return true;
        };
    }
    commandHelp(args, log) {
        const chalk = vorpal.chalk;
        log(vorpal.find(commands_1.default.FEATURE_DISABLE).helpInformation());
        log(`  Remarks:

    If the specified ${chalk.grey('url')} doesn't refer to an existing site collection,
    you will get a ${chalk.grey('"404 FILE NOT FOUND"')} error.
      
  Examples:
  
    Disable site feature
      ${this.name} --url https://contoso.sharepoint.com/sites/sales --featureId 915c240e-a6cc-49b8-8b2c-0bff8b553ed3 --scope Site

    Disable web feature (with force to ignore errors)
      ${this.name} --url https://contoso.sharepoint.com/sites/sales --featureId 00bfea71-5932-4f9c-ad71-1557e5751100 --scope Web --force
    `);
    }
}
module.exports = new SpoFeatureDisableCommand();
