// @pnp/office365-cli/dist/o365/spo/commands/file/file-remove.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const commands_1 = require("../../commands");
const request_1 = require("../../../../request");
const SpoCommand_1 = require("../../../base/SpoCommand");
const Utils_1 = require("../../../../Utils");
const vorpal = require('../../../../vorpal-init');
class SpoFileRemoveCommand extends SpoCommand_1.default {
    get name() {
        return commands_1.default.FILE_REMOVE;
    }
    get description() {
        return 'Removes the specified file';
    }
    getTelemetryProperties(args) {
        const telemetryProps = super.getTelemetryProperties(args);
        telemetryProps.id = (!(!args.options.id)).toString();
        telemetryProps.url = (!(!args.options.url)).toString();
        telemetryProps.recycle = (!(!args.options.recycle)).toString();
        telemetryProps.confirm = (!(!args.options.confirm)).toString();
        return telemetryProps;
    }
    commandAction(cmd, args, cb) {
        const removeFile = () => {
            if (this.verbose) {
                cmd.log(`Removing file in site at ${args.options.webUrl}...`);
            }
            let requestUrl = '';
            if (args.options.id) {
                requestUrl = `${args.options.webUrl}/_api/web/GetFileById(guid'${encodeURIComponent(args.options.id)}')`;
            }
            else {
                // concatenate trailing '/' if not provided
                // so if the provided url is for the root site, the substr bellow will get the right value
                let serverRelativeSiteUrl = args.options.webUrl;
                if (!serverRelativeSiteUrl.endsWith('/')) {
                    serverRelativeSiteUrl = `${serverRelativeSiteUrl}/`;
                }
                serverRelativeSiteUrl = serverRelativeSiteUrl.substr(serverRelativeSiteUrl.indexOf('/', 8));
                let fileUrl = args.options.url;
                if (!fileUrl.startsWith(serverRelativeSiteUrl)) {
                    fileUrl = `${serverRelativeSiteUrl}${fileUrl}`;
                }
                requestUrl = `${args.options.webUrl}/_api/web/GetFileByServerRelativeUrl('${encodeURIComponent(fileUrl)}')`;
            }
            if (args.options.recycle) {
                requestUrl += `/recycle()`;
            }
            const requestOptions = {
                url: requestUrl,
                method: 'POST',
                headers: {
                    'X-HTTP-Method': 'DELETE',
                    'If-Match': '*',
                    'accept': 'application/json;odata=nometadata'
                },
                json: true
            };
            request_1.default
                .post(requestOptions)
                .then(() => {
                // REST post call doesn't return anything
                cb();
            }, (err) => this.handleRejectedODataJsonPromise(err, cmd, cb));
        };
        if (args.options.confirm) {
            removeFile();
        }
        else {
            cmd.prompt({
                type: 'confirm',
                name: 'continue',
                default: false,
                message: `Are you sure you want to ${args.options.recycle ? "recycle" : "remove"} the file ${args.options.id || args.options.url} located in site ${args.options.webUrl}?`,
            }, (result) => {
                if (!result.continue) {
                    cb();
                }
                else {
                    removeFile();
                }
            });
        }
    }
    options() {
        const options = [
            {
                option: '-w, --webUrl <webUrl>',
                description: 'URL of the site where the file to remove is located'
            },
            {
                option: '-i, --id [id]',
                description: 'The ID of the file to remove. Specify either id or url but not both'
            },
            {
                option: '-u, --url [url]',
                description: 'The server- or site-relative url of the file to remove. Specify either id or url but not both'
            },
            {
                option: '--recycle',
                description: 'Recycle the file instead of actually deleting it'
            },
            {
                option: '--confirm',
                description: 'Don\'t prompt for confirming removing the file'
            }
        ];
        const parentOptions = super.options();
        return options.concat(parentOptions);
    }
    validate() {
        return (args) => {
            if (!args.options.webUrl) {
                return 'Required parameter webUrl missing';
            }
            const isValidSharePointUrl = SpoCommand_1.default.isValidSharePointUrl(args.options.webUrl);
            if (isValidSharePointUrl !== true) {
                return isValidSharePointUrl;
            }
            if (args.options.id &&
                !Utils_1.default.isValidGuid(args.options.id)) {
                return `${args.options.id} is not a valid GUID`;
            }
            if (args.options.id && args.options.url) {
                return 'Specify id or url, but not both';
            }
            if (!args.options.id && !args.options.url) {
                return 'Specify id or url';
            }
            return true;
        };
    }
    commandHelp(args, log) {
        const chalk = vorpal.chalk;
        log(vorpal.find(this.name).helpInformation());
        log(`  Examples:
  
    Remove the file with ID ${chalk.grey('0cd891ef-afce-4e55-b836-fce03286cccf')} located in site
    ${chalk.grey('https://contoso.sharepoint.com/sites/project-x')}
      ${commands_1.default.FILE_REMOVE} --webUrl https://contoso.sharepoint.com/sites/project-x --id 0cd891ef-afce-4e55-b836-fce03286cccf

    Remove the file with site-relative URL ${chalk.grey('SharedDocuments/Test.docx')} located in
    site ${chalk.grey('https://contoso.sharepoint.com/sites/project-x')}
      ${commands_1.default.FILE_REMOVE} --webUrl https://contoso.sharepoint.com/sites/project-x --url SharedDocuments/Test.docx

    Move the file with server-relative URL ${chalk.grey('/sites/project-x/SharedDocuments/Test.docx')}
    located in site ${chalk.grey('https://contoso.sharepoint.com/sites/project-x')}
    to the recycle bin
      ${commands_1.default.FILE_REMOVE} --webUrl https://contoso.sharepoint.com/sites/project-x --url /sites/project-x/SharedDocuments/Test.docx --recycle
      `);
    }
}
module.exports = new SpoFileRemoveCommand();
