// @pnp/office365-cli/dist/o365/spo/commands/homesite/homesite-get.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const request_1 = require("../../../../request");
const commands_1 = require("../../commands");
const SpoCommand_1 = require("../../../base/SpoCommand");
const vorpal = require('../../../../vorpal-init');
class SpoHomeSiteGetCommand extends SpoCommand_1.default {
    get name() {
        return commands_1.default.HOMESITE_GET;
    }
    get description() {
        return 'Gets information about the Home Site';
    }
    commandAction(cmd, args, cb) {
        this
            .getSpoUrl(cmd, this.debug)
            .then((spoUrl) => {
            const requestOptions = {
                url: `${spoUrl}/_api/SP.SPHSite/Details`,
                headers: {
                    accept: 'application/json;odata=nometadata'
                },
                json: true
            };
            return request_1.default.get(requestOptions);
        })
            .then((res) => {
            if (!res["odata.null"]) {
                cmd.log(res);
            }
            if (this.verbose) {
                cmd.log(vorpal.chalk.green('DONE'));
            }
            cb();
        }, (err) => this.handleRejectedODataJsonPromise(err, cmd, cb));
    }
    commandHelp(args, log) {
        log(vorpal.find(this.name).helpInformation());
        log(`  Examples:
  
    Get information about the Home Site
      ${commands_1.default.HOMESITE_GET}

  More information:

    SharePoint home sites: a landing for your organization on the intelligent
    intranet
      https://techcommunity.microsoft.com/t5/Microsoft-SharePoint-Blog/SharePoint-home-sites-a-landing-for-your-organization-on-the/ba-p/621933
`);
    }
}
module.exports = new SpoHomeSiteGetCommand();
