// @pnp/office365-cli/dist/o365/spo/commands/spo-get.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const commands_1 = require("../commands");
const SpoCommand_1 = require("../../base/SpoCommand");
const Auth_1 = require("../../../Auth");
const vorpal = require('../../../vorpal-init');
class SpoGetCommand extends SpoCommand_1.default {
    get name() {
        return `${commands_1.default.GET}`;
    }
    get description() {
        return 'Gets the context URL for the root SharePoint site collection and SharePoint tenant admin site';
    }
    commandAction(cmd, args, cb) {
        const spoContext = {
            SpoUrl: Auth_1.default.service.spoUrl ? Auth_1.default.service.spoUrl : ''
        };
        cmd.log(spoContext);
        cb();
    }
    options() {
        const options = [];
        const parentOptions = super.options();
        return options.concat(parentOptions);
    }
    validate() {
        return (args) => {
            return true;
        };
    }
    commandHelp(args, log) {
        log(vorpal.find(commands_1.default.GET).helpInformation());
        log(` Remarks:

    Office 365 CLI automatically discovers the URL of the root SharePoint site
    collection/SharePoint tenant admin site (whichever is needed to run
    the particular command). Using this command you can see which URLs the CLI has discovered.
      
  Examples:
  
    Get the context URL for the root SharePoint site collection and SharePoint tenant admin site
      ${commands_1.default.GET} --output json
`);
    }
}
module.exports = new SpoGetCommand();
