// @pnp/office365-cli/dist/o365/spo/commands/spo-set.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const commands_1 = require("../commands");
const Command_1 = require("../../../Command");
const SpoCommand_1 = require("../../base/SpoCommand");
const Auth_1 = require("../../../Auth");
const vorpal = require('../../../vorpal-init');
class SpoSetCommand extends SpoCommand_1.default {
    get name() {
        return `${commands_1.default.SET}`;
    }
    get description() {
        return 'Sets the URL of the root SharePoint site collection for use in SPO commands';
    }
    commandAction(cmd, args, cb) {
        Auth_1.default.service.spoUrl = args.options.url;
        Auth_1.default.storeConnectionInfo().then(() => {
            cb();
        }, err => {
            cb(new Command_1.CommandError(err));
        });
    }
    options() {
        const options = [
            {
                option: '-u, --url <url>',
                description: 'The URL of the root SharePoint site collection to use in SPO commands'
            }
        ];
        const parentOptions = super.options();
        return options.concat(parentOptions);
    }
    validate() {
        return (args) => {
            if (!args.options.url) {
                return 'Required parameter url missing';
            }
            const isValidSharePointUrl = SpoCommand_1.default.isValidSharePointUrl(args.options.url);
            if (isValidSharePointUrl !== true) {
                return isValidSharePointUrl;
            }
            return true;
        };
    }
    commandHelp(args, log) {
        log(vorpal.find(commands_1.default.SET).helpInformation());
        log(` Remarks:

    Office 365 CLI automatically discovers the URL of the root SharePoint site
    collection/SharePoint tenant admin site (whichever is needed to run
    the particular command). In specific cases, like when managing multi-geo
    Office 365 tenants, it could be desirable to make the CLI manage
    the specific geography. For such cases, you can use this command
    to explicitly specify the SPO URL that should be used when executing SPO
    commands.
      
  Examples:
  
    Set SPO URL to the specified URL
      ${commands_1.default.SET} --url https://contoso.sharepoint.com
`);
    }
}
module.exports = new SpoSetCommand();
