// @pnp/office365-cli/dist/o365/aad/commands/groupsetting/groupsetting-get.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const commands_1 = require("../../commands");
const request_1 = require("../../../../request");
const Utils_1 = require("../../../../Utils");
const GraphCommand_1 = require("../../../base/GraphCommand");
const vorpal = require('../../../../vorpal-init');
class AadGroupSettingGetCommand extends GraphCommand_1.default {
    get name() {
        return `${commands_1.default.GROUPSETTING_GET}`;
    }
    get description() {
        return 'Gets information about the particular group setting';
    }
    commandAction(cmd, args, cb) {
        const requestOptions = {
            url: `${this.resource}/v1.0/groupSettings/${args.options.id}`,
            headers: {
                accept: 'application/json;odata.metadata=none'
            },
            json: true
        };
        request_1.default
            .get(requestOptions)
            .then((res) => {
            cmd.log(res);
            if (this.verbose) {
                cmd.log(vorpal.chalk.green('DONE'));
            }
            cb();
        }, (err) => this.handleRejectedODataJsonPromise(err, cmd, cb));
    }
    options() {
        const options = [
            {
                option: '-i, --id <id>',
                description: 'The ID of the group setting to retrieve'
            }
        ];
        const parentOptions = super.options();
        return options.concat(parentOptions);
    }
    validate() {
        return (args) => {
            if (!args.options.id) {
                return 'Required parameter id missing';
            }
            if (!Utils_1.default.isValidGuid(args.options.id)) {
                return `${args.options.id} is not a valid GUID`;
            }
            return true;
        };
    }
    commandHelp(args, log) {
        const chalk = vorpal.chalk;
        log(vorpal.find(this.name).helpInformation());
        log(`  Examples:
  
    Get information about the group setting with id ${chalk.grey(`1caf7dcd-7e83-4c3a-94f7-932a1299c844`)}
      ${this.name} --id 1caf7dcd-7e83-4c3a-94f7-932a1299c844
`);
    }
}
module.exports = new AadGroupSettingGetCommand();
