// @pnp/office365-cli/dist/o365/aad/commands/groupsetting/groupsetting-list.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const commands_1 = require("../../commands");
const GraphItemsListCommand_1 = require("../../../base/GraphItemsListCommand");
const vorpal = require('../../../../vorpal-init');
class AadGroupSettingListCommand extends GraphItemsListCommand_1.GraphItemsListCommand {
    get name() {
        return `${commands_1.default.GROUPSETTING_LIST}`;
    }
    get description() {
        return 'Lists Azure AD group settings';
    }
    commandAction(cmd, args, cb) {
        this
            .getAllItems(`${this.resource}/v1.0/groupSettings`, cmd, true)
            .then(() => {
            if (args.options.output === 'json') {
                cmd.log(this.items);
            }
            else {
                cmd.log(this.items.map(i => {
                    return {
                        id: i.id,
                        displayName: i.displayName
                    };
                }));
            }
            if (this.verbose) {
                cmd.log(vorpal.chalk.green('DONE'));
            }
            cb();
        }, (err) => this.handleRejectedODataJsonPromise(err, cmd, cb));
    }
    commandHelp(args, log) {
        log(vorpal.find(this.name).helpInformation());
        log(`  Examples:
  
    List all group settings in the tenant
      ${this.name}
`);
    }
}
module.exports = new AadGroupSettingListCommand();
