// @pnp/office365-cli/dist/o365/aad/commands/groupsetting/groupsetting-add.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const commands_1 = require("../../commands");
const request_1 = require("../../../../request");
const Utils_1 = require("../../../../Utils");
const GraphCommand_1 = require("../../../base/GraphCommand");
const vorpal = require('../../../../vorpal-init');
class AadGroupSettingAddCommand extends GraphCommand_1.default {
    get name() {
        return `${commands_1.default.GROUPSETTING_ADD}`;
    }
    get description() {
        return 'Creates a group setting';
    }
    getTelemetryProperties(args) {
        const telemetryProps = super.getTelemetryProperties(args);
        telemetryProps.templateId = args.options.templateId;
        return telemetryProps;
    }
    allowUnknownOptions() {
        return true;
    }
    commandAction(cmd, args, cb) {
        if (this.verbose) {
            cmd.log(`Retrieving group setting template with id '${args.options.templateId}'...`);
        }
        const requestOptions = {
            url: `${this.resource}/v1.0/groupSettingTemplates/${args.options.templateId}`,
            headers: {
                accept: 'application/json;odata.metadata=none'
            },
            json: true
        };
        request_1.default
            .get(requestOptions)
            .then((groupSettingTemplate) => {
            const requestOptions = {
                url: `${this.resource}/v1.0/groupSettings`,
                headers: {
                    accept: 'application/json;odata.metadata=none',
                    'content-type': 'application/json'
                },
                body: {
                    templateId: args.options.templateId,
                    values: this.getGroupSettingValues(args.options, groupSettingTemplate)
                },
                json: true
            };
            return request_1.default.post(requestOptions);
        })
            .then((res) => {
            cmd.log(res);
            if (this.verbose) {
                cmd.log(vorpal.chalk.green('DONE'));
            }
            cb();
        }, (err) => this.handleRejectedODataJsonPromise(err, cmd, cb));
    }
    getGroupSettingValues(options, groupSettingTemplate) {
        const values = [];
        const excludeOptions = [
            'templateId',
            'debug',
            'verbose',
            'output'
        ];
        Object.keys(options).forEach(key => {
            if (excludeOptions.indexOf(key) === -1) {
                values.push({
                    name: key,
                    value: options[key]
                });
            }
        });
        groupSettingTemplate.values.forEach(v => {
            if (!values.find(e => e.name === v.name)) {
                values.push({
                    name: v.name,
                    value: v.defaultValue
                });
            }
        });
        return values;
    }
    options() {
        const options = [
            {
                option: '-i, --templateId <templateId>',
                description: 'The ID of the group setting template to use to create the group setting'
            }
        ];
        const parentOptions = super.options();
        return options.concat(parentOptions);
    }
    validate() {
        return (args) => {
            if (!args.options.templateId) {
                return 'Required option templateId missing';
            }
            if (!Utils_1.default.isValidGuid(args.options.templateId)) {
                return `${args.options.templateId} is not a valid GUID`;
            }
            return true;
        };
    }
    commandHelp(args, log) {
        const chalk = vorpal.chalk;
        log(vorpal.find(this.name).helpInformation());
        log(`  Remarks:

    To create a group setting, you have to specify the ID of the group setting
    template that should be used to create the setting. You can retrieve the ID
    of the template using the ${chalk.blue(commands_1.default.GROUPSETTINGTEMPLATE_LIST)}
    command.

    To specify values for the different properties specified in the group
    setting template, include additional options that match the property in the
    group setting template.
    For example ${chalk.blue("--ClassificationList 'HBI, MBI, LBI, GDPR'")} will set
    the list of classifications to use on modern SharePoint sites.

    Each group setting template specifies default value for each property. If
    you don't specify a value for the particular property yourself, the default
    value from the group setting template will be used. To find out which
    properties are available for the particular group setting template, use the
    ${chalk.blue(commands_1.default.GROUPSETTINGTEMPLATE_GET)} command.

    If the specified ${chalk.blue('templateId')} doesn't reference a valid group setting
    template, you will get a ${chalk.grey("Resource 'xyz' does not exist or one of its ")}
    ${chalk.grey('queried reference-property objects are not present.')} error.

    If you try to add a group setting using a template, for which a setting
    already exists, you will get a ${chalk.grey('A conflicting object with one or more ')}
    ${chalk.grey('of the specified property values is present in the directory.')} error.

  Examples:
  
    Configure classification for modern SharePoint sites
      ${this.name} --templateId 62375ab9-6b52-47ed-826b-58e47e0e304b --UsageGuidelinesUrl https://contoso.sharepoint.com/sites/compliance --ClassificationList 'HBI, MBI, LBI, GDPR' --DefaultClassification MBI
`);
    }
}
module.exports = new AadGroupSettingAddCommand();
