// @pnp/office365-cli/dist/o365/spo/commands/list/list-contenttype-add.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const commands_1 = require("../../commands");
const request_1 = require("../../../../request");
const SpoCommand_1 = require("../../../base/SpoCommand");
const Utils_1 = require("../../../../Utils");
const vorpal = require('../../../../vorpal-init');
class SpoListContentTypeAddCommand extends SpoCommand_1.default {
    get name() {
        return commands_1.default.LIST_CONTENTTYPE_ADD;
    }
    get description() {
        return 'Adds content type to list';
    }
    types() {
        return {
            string: ['contentTypeId', 'c']
        };
    }
    getTelemetryProperties(args) {
        const telemetryProps = super.getTelemetryProperties(args);
        telemetryProps.listId = typeof args.options.listId !== 'undefined';
        telemetryProps.listTitle = typeof args.options.listTitle !== 'undefined';
        return telemetryProps;
    }
    commandAction(cmd, args, cb) {
        if (this.verbose) {
            const list = args.options.listId ? encodeURIComponent(args.options.listId) : encodeURIComponent(args.options.listTitle);
            cmd.log(`Adding content type ${args.options.contentTypeId} to list ${list} in site at ${args.options.webUrl}...`);
        }
        let requestUrl = '';
        if (args.options.listId) {
            requestUrl = `${args.options.webUrl}/_api/web/lists(guid'${encodeURIComponent(args.options.listId)}')/ContentTypes/AddAvailableContentType`;
        }
        else {
            requestUrl = `${args.options.webUrl}/_api/web/lists/GetByTitle('${encodeURIComponent(args.options.listTitle)}')/ContentTypes/AddAvailableContentType`;
        }
        const requestBody = this.mapRequestBody(args.options);
        const requestOptions = {
            url: requestUrl,
            headers: {
                'accept': 'application/json;odata=nometadata'
            },
            body: requestBody,
            json: true
        };
        request_1.default
            .post(requestOptions)
            .then((res) => {
            cmd.log(res);
            cb();
        }, (err) => this.handleRejectedODataJsonPromise(err, cmd, cb));
    }
    mapRequestBody(options) {
        return {
            contentTypeId: options.contentTypeId
        };
    }
    options() {
        const options = [
            {
                option: '-u, --webUrl <webUrl>',
                description: 'URL of the site where the list is located'
            },
            {
                option: '-l, --listId [listId]',
                description: 'ID of the list to which to add the content type. Specify listId or listTitle but not both'
            },
            {
                option: '-t, --listTitle [listTitle]',
                description: 'Title of the list to which to add the content type. Specify listId or listTitle but not both'
            },
            {
                option: '-c, --contentTypeId <contentTypeId>',
                description: 'ID of the content type to add to the list'
            }
        ];
        const parentOptions = super.options();
        return options.concat(parentOptions);
    }
    validate() {
        return (args) => {
            if (!args.options.webUrl) {
                return 'Required parameter webUrl missing';
            }
            if (!args.options.contentTypeId) {
                return 'Required parameter contentTypeId missing';
            }
            const isValidSharePointUrl = SpoCommand_1.default.isValidSharePointUrl(args.options.webUrl);
            if (isValidSharePointUrl !== true) {
                return isValidSharePointUrl;
            }
            if (args.options.listId) {
                if (!Utils_1.default.isValidGuid(args.options.listId)) {
                    return `${args.options.listId} is not a valid GUID`;
                }
            }
            if (args.options.listId && args.options.listTitle) {
                return 'Specify listId or listTitle, but not both';
            }
            if (!args.options.listId && !args.options.listTitle) {
                return 'Specify listId or listTitle, one is required';
            }
            return true;
        };
    }
    commandHelp(args, log) {
        const chalk = vorpal.chalk;
        log(vorpal.find(this.name).helpInformation());
        log(`  Examples:
  
    Add existing content type with ID ${chalk.grey('0x0120')} to the list with ID
    ${chalk.grey('0cd891ef-afce-4e55-b836-fce03286cccf')} located in site
    ${chalk.grey('https://contoso.sharepoint.com/sites/project-x')}
      ${commands_1.default.LIST_CONTENTTYPE_ADD} --webUrl https://contoso.sharepoint.com/sites/project-x --listId 0cd891ef-afce-4e55-b836-fce03286cccf --contentTypeId 0x0120

    Add existing content type with ID ${chalk.grey('0x0120')} to the list with title ${chalk.grey('Documents')}
    located in site ${chalk.grey('https://contoso.sharepoint.com/sites/project-x')}
      ${commands_1.default.LIST_CONTENTTYPE_ADD} --webUrl https://contoso.sharepoint.com/sites/project-x --listTitle Documents --contentTypeId 0x0120
      `);
    }
}
module.exports = new SpoListContentTypeAddCommand();
