// @pnp/office365-cli/dist/o365/spo/commands/theme/theme-list.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const commands_1 = require("../../commands");
const request_1 = require("../../../../request");
const SpoCommand_1 = require("../../../base/SpoCommand");
const vorpal = require('../../../../vorpal-init');
class SpoThemeListCommand extends SpoCommand_1.default {
    get name() {
        return commands_1.default.THEME_LIST;
    }
    get description() {
        return 'Retrieves the list of custom themes';
    }
    commandAction(cmd, args, cb) {
        this
            .getSpoAdminUrl(cmd, this.debug)
            .then((spoAdminUrl) => {
            if (this.verbose) {
                cmd.log(`Retrieving themes from tenant store...`);
            }
            const requestOptions = {
                url: `${spoAdminUrl}/_api/thememanager/GetTenantThemingOptions`,
                headers: {
                    'accept': 'application/json;odata=nometadata'
                },
                json: true
            };
            return request_1.default.post(requestOptions);
        })
            .then((rawRes) => {
            const themePreviews = rawRes.themePreviews;
            if (themePreviews && themePreviews.length > 0) {
                if (args.options.output === 'json') {
                    cmd.log(themePreviews);
                }
                else {
                    cmd.log(themePreviews.map(a => {
                        return { Name: a.name };
                    }));
                }
            }
            else {
                if (this.verbose) {
                    cmd.log('No themes found');
                }
            }
            cb();
        }, (err) => this.handleRejectedODataJsonPromise(err, cmd, cb));
    }
    commandHelp(args, log) {
        const chalk = vorpal.chalk;
        log(vorpal.find(this.name).helpInformation());
        log(`  ${chalk.yellow('Important:')} to use this command you have to have permissions to access
    the tenant admin site.
    
  Examples:
  
    List available themes
      ${commands_1.default.THEME_LIST}

  More information:

    SharePoint site theming
      https://docs.microsoft.com/en-us/sharepoint/dev/declarative-customization/site-theming/sharepoint-site-theming-overview
      `);
    }
}
module.exports = new SpoThemeListCommand();
