// @pnp/office365-cli/dist/o365/yammer/commands/message/message-remove.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const commands_1 = require("../../commands");
const YammerCommand_1 = require("../../../base/YammerCommand");
const request_1 = require("../../../../request");
const vorpal = require('../../../../vorpal-init');
class YammerMessageRemoveCommand extends YammerCommand_1.default {
    get name() {
        return commands_1.default.YAMMER_MESSAGE_REMOVE;
    }
    get description() {
        return 'Removes a Yammer message';
    }
    getTelemetryProperties(args) {
        const telemetryProps = super.getTelemetryProperties(args);
        telemetryProps.confirm = (!(!args.options.confirm)).toString();
        return telemetryProps;
    }
    commandAction(cmd, args, cb) {
        const removeMessage = () => {
            const requestOptions = {
                url: `${this.resource}/v1/messages/${args.options.id}.json`,
                headers: {
                    accept: 'application/json;odata.metadata=none',
                    'content-type': 'application/json;odata=nometadata'
                },
                json: true
            };
            request_1.default
                .delete(requestOptions)
                .then((res) => {
                if (this.verbose) {
                    cmd.log(vorpal.chalk.green('DONE'));
                }
                cb();
            }, (err) => this.handleRejectedODataJsonPromise(err, cmd, cb));
        };
        if (args.options.confirm) {
            removeMessage();
        }
        else {
            cmd.prompt({
                type: 'confirm',
                name: 'continue',
                default: false,
                message: `Are you sure you want to remove the Yammer message ${args.options.id}?`,
            }, (result) => {
                if (!result.continue) {
                    cb();
                }
                else {
                    removeMessage();
                }
            });
        }
    }
    options() {
        const options = [
            {
                option: '--id <id>',
                description: 'The id of the Yammer message'
            },
            {
                option: '--confirm',
                description: 'Don\'t prompt for confirming removing the Yammer message'
            }
        ];
        const parentOptions = super.options();
        return options.concat(parentOptions);
    }
    validate() {
        return (args) => {
            if (!args.options.id) {
                return 'Required id value is missing';
            }
            if (typeof args.options.id !== 'number') {
                return `${args.options.id} is not a number`;
            }
            return true;
        };
    }
    commandHelp(args, log) {
        const chalk = vorpal.chalk;
        log(vorpal.find(this.name).helpInformation());
        log(`  Remarks:
  
    ${chalk.yellow('Attention:')} In order to use this command, you need to grant the Azure AD
    application used by the Office 365 CLI the permission to the Yammer API.
    To do this, execute the ${chalk.blue('consent --service yammer')} command.

    To remove a message, you must either:
      - have posted the message yourself 
      - be an administrator of the group the message was posted to or 
      - be an admin of the network the message is in.
    
  Examples:
  
    Removes the Yammer message with the id 1239871123
      ${this.name} --id 1239871123

    Removes the Yammer message with the id 1239871123 without prompting for
    confirmation.
      ${this.name} --id 1239871123 --confirm
`);
    }
}
module.exports = new YammerMessageRemoveCommand();
