// @pnp/office365-cli/dist/o365/spo/commands/app/app-get.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const commands_1 = require("../../commands");
const request_1 = require("../../../../request");
const Utils_1 = require("../../../../Utils");
const SpoAppBaseCommand_1 = require("./SpoAppBaseCommand");
const vorpal = require('../../../../vorpal-init');
class SpoAppGetCommand extends SpoAppBaseCommand_1.SpoAppBaseCommand {
    get name() {
        return commands_1.default.APP_GET;
    }
    get description() {
        return 'Gets information about the specific app from the specified app catalog';
    }
    getTelemetryProperties(args) {
        const telemetryProps = super.getTelemetryProperties(args);
        telemetryProps.id = (!(!args.options.id)).toString();
        telemetryProps.name = (!(!args.options.name)).toString();
        telemetryProps.appCatalogUrl = (!(!args.options.appCatalogUrl)).toString();
        telemetryProps.scope = args.options.scope || 'tenant';
        return telemetryProps;
    }
    commandAction(cmd, args, cb) {
        const scope = (args.options.scope) ? args.options.scope.toLowerCase() : 'tenant';
        let appCatalogSiteUrl = '';
        this
            .getSpoUrl(cmd, this.debug)
            .then((spoUrl) => {
            return this.getAppCatalogSiteUrl(cmd, spoUrl, args);
        })
            .then((appCatalogUrl) => {
            appCatalogSiteUrl = appCatalogUrl;
            if (args.options.id) {
                return Promise.resolve({ UniqueId: args.options.id });
            }
            if (this.verbose) {
                cmd.log(`Looking up app id for app named ${args.options.name}...`);
            }
            const requestOptions = {
                url: `${appCatalogSiteUrl}/_api/web/getfolderbyserverrelativeurl('AppCatalog')/files('${args.options.name}')?$select=UniqueId`,
                headers: {
                    accept: 'application/json;odata=nometadata'
                },
                json: true
            };
            return request_1.default.get(requestOptions);
        })
            .then((res) => {
            if (this.verbose) {
                cmd.log(`Retrieving information for app ${res}...`);
            }
            const requestOptions = {
                url: `${appCatalogSiteUrl}/_api/web/${scope}appcatalog/AvailableApps/GetById('${encodeURIComponent(res.UniqueId)}')`,
                headers: {
                    accept: 'application/json;odata=nometadata'
                },
                json: true
            };
            return request_1.default.get(requestOptions);
        })
            .then((res) => {
            cmd.log(res);
            cb();
        }, (rawRes) => this.handleRejectedODataJsonPromise(rawRes, cmd, cb));
    }
    options() {
        const options = [
            {
                option: '-i, --id [id]',
                description: 'ID of the app to retrieve information for. Specify the id or the name but not both'
            },
            {
                option: '-n, --name [name]',
                description: 'Name of the app to retrieve information for. Specify the id or the name but not both'
            },
            {
                option: '-u, --appCatalogUrl [appCatalogUrl]',
                description: 'URL of the tenant or site collection app catalog. It must be specified when the scope is \'sitecollection\''
            },
            {
                option: '-s, --scope [scope]',
                description: 'Scope of the app catalog: tenant|sitecollection. Default tenant',
                autocomplete: ['tenant', 'sitecollection']
            }
        ];
        const parentOptions = super.options();
        return options.concat(parentOptions);
    }
    validate() {
        return (args) => {
            // verify either 'tenant' or 'sitecollection' specified if scope provided
            if (args.options.scope) {
                const testScope = args.options.scope.toLowerCase();
                if (!(testScope === 'tenant' || testScope === 'sitecollection')) {
                    return `Scope must be either 'tenant' or 'sitecollection'`;
                }
                if (testScope === 'sitecollection' && !args.options.appCatalogUrl) {
                    return `You must specify appCatalogUrl when the scope is sitecollection`;
                }
            }
            if (!args.options.id && !args.options.name) {
                return 'Specify either the id or the name';
            }
            if (args.options.id && args.options.name) {
                return 'Specify either the id or the name but not both';
            }
            if (args.options.id && !Utils_1.default.isValidGuid(args.options.id)) {
                return `${args.options.id} is not a valid GUID`;
            }
            if (args.options.appCatalogUrl) {
                return SpoAppBaseCommand_1.SpoAppBaseCommand.isValidSharePointUrl(args.options.appCatalogUrl);
            }
            return true;
        };
    }
    commandHelp(args, log) {
        const chalk = vorpal.chalk;
        log(vorpal.find(commands_1.default.APP_GET).helpInformation());
        log(`  Remarks:
  
    When getting information about an app from the tenant app catalog,
    it's not necessary to specify the tenant app catalog URL. When the URL
    is not specified, the CLI will try to resolve the URL itself.
    Specifying the app catalog URL is required when you want to get information
    about an app from a site collection app catalog.

    When specifying site collection app catalog, you can specify the URL either
    with our without the ${chalk.grey('AppCatalog')} part, for example
    ${chalk.grey('https://contoso.sharepoint.com/sites/team-a/AppCatalog')} or
    ${chalk.grey('https://contoso.sharepoint.com/sites/team-a')}. CLI will accept both formats.
   
  Examples:
  
    Return details about the app with ID ${chalk.grey('b2307a39-e878-458b-bc90-03bc578531d6')}
    available in the tenant app catalog.
      ${commands_1.default.APP_GET} --id b2307a39-e878-458b-bc90-03bc578531d6

    Return details about the app with name ${chalk.grey('solution.sppkg')}
    available in the tenant app catalog. Will try to detect the app catalog URL
      ${commands_1.default.APP_GET} --name solution.sppkg

    Return details about the app with name ${chalk.grey('solution.sppkg')}
    available in the tenant app catalog using the specified app catalog URL
      ${commands_1.default.APP_GET} --name solution.sppkg --appCatalogUrl https://contoso.sharepoint.com/sites/apps

    Return details about the app with ID ${chalk.grey('b2307a39-e878-458b-bc90-03bc578531d6')}
    available in the site collection app catalog
    of site ${chalk.grey('https://contoso.sharepoint.com/sites/site1')}.
      ${commands_1.default.APP_GET} --id b2307a39-e878-458b-bc90-03bc578531d6 --scope sitecollection --appCatalogUrl https://contoso.sharepoint.com/sites/site1

  More information:
  
    Application Lifecycle Management (ALM) APIs
      https://docs.microsoft.com/en-us/sharepoint/dev/apis/alm-api-for-spfx-add-ins
`);
    }
}
module.exports = new SpoAppGetCommand();
