// @pnp/office365-cli/dist/o365/spo/commands/app/app-list.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const commands_1 = require("../../commands");
const request_1 = require("../../../../request");
const SpoAppBaseCommand_1 = require("./SpoAppBaseCommand");
const vorpal = require('../../../../vorpal-init');
class SpoAppListCommand extends SpoAppBaseCommand_1.SpoAppBaseCommand {
    get name() {
        return commands_1.default.APP_LIST;
    }
    get description() {
        return 'Lists apps from the specified app catalog';
    }
    getTelemetryProperties(args) {
        const telemetryProps = super.getTelemetryProperties(args);
        telemetryProps.appCatalogUrl = (!(!args.options.appCatalogUrl)).toString();
        telemetryProps.scope = args.options.scope || 'tenant';
        return telemetryProps;
    }
    commandAction(cmd, args, cb) {
        const scope = (args.options.scope) ? args.options.scope.toLowerCase() : 'tenant';
        let appCatalogSiteUrl = '';
        let spoUrl = '';
        this
            .getSpoUrl(cmd, this.debug)
            .then((_spoUrl) => {
            spoUrl = _spoUrl;
            return this.getAppCatalogSiteUrl(cmd, spoUrl, args);
        })
            .then((appCatalogUrl) => {
            appCatalogSiteUrl = appCatalogUrl;
            if (this.verbose) {
                cmd.log(`Retrieving apps...`);
            }
            const requestOptions = {
                url: `${appCatalogSiteUrl}/_api/web/${scope}appcatalog/AvailableApps`,
                headers: {
                    accept: 'application/json;odata=nometadata'
                },
                json: true
            };
            return request_1.default.get(requestOptions);
        })
            .then((apps) => {
            if (apps.value && apps.value.length > 0) {
                if (args.options.output === 'json') {
                    cmd.log(apps.value);
                }
                else {
                    cmd.log(apps.value.map(a => {
                        return {
                            Title: a.Title,
                            ID: a.ID,
                            Deployed: a.Deployed,
                            AppCatalogVersion: a.AppCatalogVersion
                        };
                    }));
                }
            }
            else {
                if (this.verbose) {
                    cmd.log('No apps found');
                }
            }
            cb();
        }, (rawRes) => this.handleRejectedODataJsonPromise(rawRes, cmd, cb));
    }
    options() {
        const options = [
            {
                option: '-s, --scope [scope]',
                description: 'Scope of the app catalog: tenant|sitecollection. Default tenant',
                autocomplete: ['tenant', 'sitecollection']
            },
            {
                option: '-u, --appCatalogUrl [appCatalogUrl]',
                description: 'URL of the tenant or site collection app catalog. It must be specified when the scope is \'sitecollection\''
            }
        ];
        const parentOptions = super.options();
        return options.concat(parentOptions);
    }
    validate() {
        return (args) => {
            // verify either 'tenant' or 'sitecollection' specified if scope provided
            if (args.options.scope) {
                const testScope = args.options.scope.toLowerCase();
                if (!(testScope === 'tenant' || testScope === 'sitecollection')) {
                    return `Scope must be either 'tenant' or 'sitecollection'`;
                }
                if (testScope === 'sitecollection' && !args.options.appCatalogUrl) {
                    return `You must specify appCatalogUrl when the scope is sitecollection`;
                }
                if (args.options.appCatalogUrl) {
                    return SpoAppBaseCommand_1.SpoAppBaseCommand.isValidSharePointUrl(args.options.appCatalogUrl);
                }
            }
            return true;
        };
    }
    commandHelp(args, log) {
        const chalk = vorpal.chalk;
        log(vorpal.find(commands_1.default.APP_LIST).helpInformation());
        log(`  Remarks:

    When listing information about apps available in the tenant app catalog,
    it's not necessary to specify the tenant app catalog URL. When the URL
    is not specified, the CLI will try to resolve the URL itself.
    Specifying the app catalog URL is required when you want to list information
    about apps in a site collection app catalog.

    When specifying site collection app catalog, you can specify the URL either
    with our without the ${chalk.grey('AppCatalog')} part, for example
    ${chalk.grey('https://contoso.sharepoint.com/sites/team-a/AppCatalog')} or
    ${chalk.grey('https://contoso.sharepoint.com/sites/team-a')}. CLI will accept both formats.

    When using the text output type (default), the command lists only the values
    of the ${chalk.grey('Title')}, ${chalk.grey('ID')}, ${chalk.grey('Deployed')} and ${chalk.grey('AppCatalogVersion')} properties of the app.
    When setting the output type to JSON, all available properties are included
    in the command output.
   
  Examples:
  
    Return the list of available apps from the tenant app catalog.
    Show the installed version in the site if applicable.
      ${commands_1.default.APP_LIST}

    Return the list of available apps from a site collection app catalog
    of site ${chalk.grey('https://contoso.sharepoint.com/sites/site1')}.
      ${commands_1.default.APP_LIST} --scope sitecollection --appCatalogUrl https://contoso.sharepoint.com/sites/site1

  More information:
  
    Application Lifecycle Management (ALM) APIs
      https://docs.microsoft.com/en-us/sharepoint/dev/apis/alm-api-for-spfx-add-ins
`);
    }
}
module.exports = new SpoAppListCommand();
