// @pnp/office365-cli/dist/o365/spo/commands/app/app-uninstall.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const commands_1 = require("../../commands");
const request_1 = require("../../../../request");
const SpoCommand_1 = require("../../../base/SpoCommand");
const Utils_1 = require("../../../../Utils");
const vorpal = require('../../../../vorpal-init');
class SpoAppUninstallCommand extends SpoCommand_1.default {
    get name() {
        return commands_1.default.APP_UNINSTALL;
    }
    get description() {
        return 'Uninstalls an app from the site';
    }
    getTelemetryProperties(args) {
        const telemetryProps = super.getTelemetryProperties(args);
        telemetryProps.confirm = (!(!args.options.confirm)).toString();
        telemetryProps.scope = args.options.scope || 'tenant';
        return telemetryProps;
    }
    commandAction(cmd, args, cb) {
        const uninstallApp = () => {
            const scope = (args.options.scope) ? args.options.scope.toLowerCase() : 'tenant';
            if (this.verbose) {
                cmd.log(`Uninstalling app '${args.options.id}' from the site '${args.options.siteUrl}'...`);
            }
            const requestOptions = {
                url: `${args.options.siteUrl}/_api/web/${scope}appcatalog/AvailableApps/GetById('${encodeURIComponent(args.options.id)}')/uninstall`,
                headers: {
                    accept: 'application/json;odata=nometadata'
                }
            };
            request_1.default
                .post(requestOptions)
                .then(() => {
                if (this.verbose) {
                    cmd.log(vorpal.chalk.green('DONE'));
                }
                cb();
            }, (rawRes) => this.handleRejectedODataPromise(rawRes, cmd, cb));
        };
        if (args.options.confirm) {
            uninstallApp();
        }
        else {
            cmd.prompt({
                type: 'confirm',
                name: 'continue',
                default: false,
                message: `Are you sure you want to uninstall the app ${args.options.id} from site ${args.options.siteUrl}?`,
            }, (result) => {
                if (!result.continue) {
                    cb();
                }
                else {
                    uninstallApp();
                }
            });
        }
    }
    options() {
        const options = [
            {
                option: '-i, --id <id>',
                description: 'ID of the app to uninstall'
            },
            {
                option: '-s, --siteUrl <siteUrl>',
                description: 'Absolute URL of the site to uninstall the app from'
            },
            {
                option: '--scope [scope]',
                description: 'Scope of the app catalog: tenant|sitecollection. Default tenant',
                autocomplete: ['tenant', 'sitecollection']
            },
            {
                option: '--confirm',
                description: 'Don\'t prompt for confirming uninstalling the app'
            }
        ];
        const parentOptions = super.options();
        return options.concat(parentOptions);
    }
    validate() {
        return (args) => {
            if (args.options.scope) {
                const testScope = args.options.scope.toLowerCase();
                if (!(testScope === 'tenant' || testScope === 'sitecollection')) {
                    return `Scope must be either 'tenant' or 'sitecollection' if specified`;
                }
            }
            if (!args.options.id) {
                return 'Required parameter id missing';
            }
            if (!Utils_1.default.isValidGuid(args.options.id)) {
                return `${args.options.id} is not a valid GUID`;
            }
            if (!args.options.siteUrl) {
                return 'Required parameter siteUrl missing';
            }
            return SpoCommand_1.default.isValidSharePointUrl(args.options.siteUrl);
        };
    }
    commandHelp(args, log) {
        const chalk = vorpal.chalk;
        log(vorpal.find(commands_1.default.APP_UNINSTALL).helpInformation());
        log(`  Remarks:
  
    If the app with the specified ID doesn't exist in the app catalog,
    the command will fail with an error.
   
  Examples:
  
    Uninstall the app with ID ${chalk.grey('b2307a39-e878-458b-bc90-03bc578531d6')}
    from the ${chalk.grey('https://contoso.sharepoint.com')} site.
      ${commands_1.default.APP_UNINSTALL} --id b2307a39-e878-458b-bc90-03bc578531d6 --siteUrl https://contoso.sharepoint.com

    Uninstall the app with ID ${chalk.grey('b2307a39-e878-458b-bc90-03bc578531d6')}
    from the ${chalk.grey('https://contoso.sharepoint.com')} site without prompting
    for confirmation.
      ${commands_1.default.APP_UNINSTALL} --id b2307a39-e878-458b-bc90-03bc578531d6 --siteUrl https://contoso.sharepoint.com --confirm

    Uninstall the app with ID ${chalk.grey('b2307a39-e878-458b-bc90-03bc578531d6')}
    from the ${chalk.grey('https://contoso.sharepoint.com')} site where the app is deployed
    to the site collection app catalog of ${chalk.grey('https://contoso.sharepoint.com')}.
      ${commands_1.default.APP_UNINSTALL} --id b2307a39-e878-458b-bc90-03bc578531d6 --siteUrl https://contoso.sharepoint.com --scope sitecollection

  More information:
  
    Application Lifecycle Management (ALM) APIs
      https://docs.microsoft.com/en-us/sharepoint/dev/apis/alm-api-for-spfx-add-ins
`);
    }
}
module.exports = new SpoAppUninstallCommand();
