// @pnp/office365-cli/dist/o365/spo/commands/app/SpoAppBaseCommand.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const SpoCommand_1 = require("../../../base/SpoCommand");
const request_1 = require("../../../../request");
class SpoAppBaseCommand extends SpoCommand_1.default {
    getAppCatalogSiteUrl(cmd, authSiteUrl, args) {
        return new Promise((resolve, reject) => {
            if (args.options.scope === 'sitecollection') {
                return resolve(args.options.appCatalogUrl.toLowerCase().replace('/appcatalog', ''));
            }
            if (args.options.appCatalogUrl) {
                return resolve(args.options.appCatalogUrl);
            }
            const requestOptions = {
                url: `${authSiteUrl}/_api/SP_TenantSettings_Current`,
                headers: {
                    accept: 'application/json;odata=nometadata'
                },
                json: true
            };
            request_1.default
                .get(requestOptions)
                .then((res) => {
                if (res.CorporateCatalogUrl) {
                    return resolve(res.CorporateCatalogUrl);
                }
                reject('Tenant app catalog is not configured.');
            })
                .catch((err) => {
                reject(err);
            });
        });
    }
}
exports.SpoAppBaseCommand = SpoAppBaseCommand;
