// @pnp/office365-cli/dist/o365/spo/commands/hidedefaultthemes/hidedefaultthemes-get.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const commands_1 = require("../../commands");
const request_1 = require("../../../../request");
const SpoCommand_1 = require("../../../base/SpoCommand");
const vorpal = require('../../../../vorpal-init');
class SpoHideDefaultThemesGetCommand extends SpoCommand_1.default {
    get name() {
        return commands_1.default.HIDEDEFAULTTHEMES_GET;
    }
    get description() {
        return 'Gets the current value of the HideDefaultThemes setting';
    }
    commandAction(cmd, args, cb) {
        this
            .getSpoAdminUrl(cmd, this.debug)
            .then((spoAdminUrl) => {
            if (this.verbose) {
                cmd.log(`Getting the current value of the HideDefaultThemes setting...`);
            }
            const requestOptions = {
                url: `${spoAdminUrl}/_api/thememanager/GetHideDefaultThemes`,
                headers: {
                    'accept': 'application/json;odata=nometadata'
                },
                json: true
            };
            return request_1.default.post(requestOptions);
        })
            .then((rawRes) => {
            cmd.log(rawRes.value);
            cb();
        }, (err) => this.handleRejectedODataJsonPromise(err, cmd, cb));
    }
    commandHelp(args, log) {
        const chalk = vorpal.chalk;
        log(vorpal.find(this.name).helpInformation());
        log(`  ${chalk.yellow('Important:')} to use this command you have to have permissions to access
    the tenant admin site.
      
  Examples:
  
    Get the current value of the HideDefaultThemes setting
      ${commands_1.default.HIDEDEFAULTTHEMES_GET}

  More information:

    SharePoint site theming
      https://docs.microsoft.com/en-us/sharepoint/dev/declarative-customization/site-theming/sharepoint-site-theming-overview
      `);
    }
}
module.exports = new SpoHideDefaultThemesGetCommand();
