// @pnp/office365-cli/dist/o365/spo/commands/page/page-section-get.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const commands_1 = require("../../commands");
const SpoCommand_1 = require("../../../base/SpoCommand");
const Page_1 = require("./Page");
const vorpal = require('../../../../vorpal-init');
class SpoPageSectionGetCommand extends SpoCommand_1.default {
    get name() {
        return `${commands_1.default.PAGE_SECTION_GET}`;
    }
    get description() {
        return 'Get information about the specified modern page section';
    }
    commandAction(cmd, args, cb) {
        Page_1.Page
            .getPage(args.options.name, args.options.webUrl, cmd, this.debug, this.verbose)
            .then((clientSidePage) => {
            const sections = clientSidePage.sections
                .filter(section => section.order === args.options.section);
            const isJSONOutput = args.options.output === 'json';
            if (sections.length) {
                cmd.log(Page_1.Page.getSectionInformation(sections[0], isJSONOutput));
            }
            if (this.verbose) {
                cmd.log(vorpal.chalk.green('DONE'));
            }
            cb();
        }, (err) => this.handleRejectedODataJsonPromise(err, cmd, cb));
    }
    options() {
        const options = [
            {
                option: '-u, --webUrl <webUrl>',
                description: 'URL of the site where the page to retrieve is located'
            },
            {
                option: '-n, --name <name>',
                description: 'Name of the page to get section information of'
            },
            {
                option: '-s, --section <sectionId>',
                description: 'ID of the section for which to retrieve information'
            }
        ];
        const parentOptions = super.options();
        return options.concat(parentOptions);
    }
    validate() {
        return (args) => {
            if (!args.options.name) {
                return 'Required parameter name missing';
            }
            if (!args.options.webUrl) {
                return 'Required parameter webUrl missing';
            }
            if (!args.options.section) {
                return 'Required parameter section missing';
            }
            else {
                if (isNaN(args.options.section)) {
                    return `${args.options.section} is not a number`;
                }
            }
            return SpoCommand_1.default.isValidSharePointUrl(args.options.webUrl);
        };
    }
    commandHelp(args, log) {
        const chalk = vorpal.chalk;
        log(vorpal.find(this.name).helpInformation());
        log(`  Remarks:

    If the specified ${chalk.grey('name')} doesn't refer to an existing modern 
    page, you will get a ${chalk.grey('File doesn\'t exists')} error.

  Examples:
  
    Get information about the specified section of the modern page
    named ${chalk.grey('home.aspx')}
      ${this.name} --webUrl https://contoso.sharepoint.com/sites/team-a --name home.aspx --section 1
`);
    }
}
module.exports = new SpoPageSectionGetCommand();
