// @pnp/office365-cli/dist/o365/spo/commands/sitedesign/sitedesign-rights-grant.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const request_1 = require("../../../../request");
const commands_1 = require("../../commands");
const SpoCommand_1 = require("../../../base/SpoCommand");
const Utils_1 = require("../../../../Utils");
const vorpal = require('../../../../vorpal-init');
class SpoSiteDesignRightsGrantCommand extends SpoCommand_1.default {
    get name() {
        return `${commands_1.default.SITEDESIGN_RIGHTS_GRANT}`;
    }
    get description() {
        return 'Grants access to a site design for one or more principals';
    }
    commandAction(cmd, args, cb) {
        let spoUrl = '';
        this
            .getSpoUrl(cmd, this.debug)
            .then((_spoUrl) => {
            spoUrl = _spoUrl;
            return this.getRequestDigest(spoUrl);
        })
            .then((res) => {
            const grantedRights = '1';
            const requestOptions = {
                url: `${spoUrl}/_api/Microsoft.Sharepoint.Utilities.WebTemplateExtensions.SiteScriptUtility.GrantSiteDesignRights`,
                headers: {
                    'X-RequestDigest': res.FormDigestValue,
                    'content-type': 'application/json;charset=utf-8',
                    accept: 'application/json;odata=nometadata'
                },
                body: {
                    id: args.options.id,
                    principalNames: args.options.principals.split(',').map(p => p.trim()),
                    grantedRights: grantedRights
                },
                json: true
            };
            return request_1.default.post(requestOptions);
        })
            .then(() => {
            if (this.verbose) {
                cmd.log(vorpal.chalk.green('DONE'));
            }
            cb();
        }, (err) => this.handleRejectedODataJsonPromise(err, cmd, cb));
    }
    options() {
        const options = [
            {
                option: '-i, --id <id>',
                description: 'The ID of the site design to grant rights on'
            },
            {
                option: '-p, --principals <principals>',
                description: 'Comma-separated list of principals to grant view rights. Principals can be users or mail-enabled security groups in the form of "alias" or "alias@<domain name>.com"'
            },
            {
                option: '-r, --rights <rights>',
                description: 'Rights to grant to principals. Available values View',
                autocomplete: ['View']
            }
        ];
        const parentOptions = super.options();
        return options.concat(parentOptions);
    }
    validate() {
        return (args) => {
            if (!args.options.id) {
                return 'Required parameter id missing';
            }
            if (!Utils_1.default.isValidGuid(args.options.id)) {
                return `${args.options.id} is not a valid GUID`;
            }
            if (!args.options.principals) {
                return 'Required parameter principals missing';
            }
            if (!args.options.rights) {
                return 'Required parameter rights missing';
            }
            if (args.options.rights !== 'View') {
                return `${args.options.rights} is not a valid rights value. Allowed values View`;
            }
            return true;
        };
    }
    commandHelp(args, log) {
        const chalk = vorpal.chalk;
        log(vorpal.find(this.name).helpInformation());
        log(`  Examples:

    Grant user with alias ${chalk.grey('PattiF')} view permission to the specified site design
      ${this.name} --id 9b142c22-037f-4a7f-9017-e9d8c0e34b98 --principals PattiF --rights View

    Grant users with aliases ${chalk.grey('PattiF')} and ${chalk.grey('AdeleV')} view permission to the specified site design
      ${this.name} --id 9b142c22-037f-4a7f-9017-e9d8c0e34b98 --principals "PattiF,AdeleV" --rights View

    Grant user with email ${chalk.grey('PattiF@contoso.com')} view permission to the specified site design
      ${this.name} --id 9b142c22-037f-4a7f-9017-e9d8c0e34b98 --principals PattiF@contoso.com --rights View

  More information:

    SharePoint site design and site script overview
      https://docs.microsoft.com/en-us/sharepoint/dev/declarative-customization/site-design-overview
`);
    }
}
module.exports = new SpoSiteDesignRightsGrantCommand();
