// @pnp/office365-cli/dist/o365/spo/commands/sitedesign/sitedesign-run-list.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const request_1 = require("../../../../request");
const commands_1 = require("../../commands");
const SpoCommand_1 = require("../../../base/SpoCommand");
const Utils_1 = require("../../../../Utils");
const vorpal = require('../../../../vorpal-init');
class SpoSiteDesignRunListCommand extends SpoCommand_1.default {
    get name() {
        return `${commands_1.default.SITEDESIGN_RUN_LIST}`;
    }
    get description() {
        return 'Lists information about site designs applied to the specified site';
    }
    getTelemetryProperties(args) {
        const telemetryProps = super.getTelemetryProperties(args);
        telemetryProps.siteDesignId = typeof args.options.siteDesignId !== 'undefined';
        return telemetryProps;
    }
    commandAction(cmd, args, cb) {
        const body = {};
        if (args.options.siteDesignId) {
            body.siteDesignId = args.options.siteDesignId;
        }
        const requestOptions = {
            url: `${args.options.webUrl}/_api/Microsoft.Sharepoint.Utilities.WebTemplateExtensions.SiteScriptUtility.GetSiteDesignRun`,
            headers: {
                accept: 'application/json;odata=nometadata',
                'content-type': 'application/json;odata=nometadata'
            },
            body: body,
            json: true
        };
        request_1.default.post(requestOptions)
            .then((res) => {
            if (args.options.output === 'json') {
                cmd.log(res.value);
            }
            else {
                cmd.log(res.value.map(d => {
                    return {
                        ID: d.ID,
                        SiteDesignID: d.SiteDesignID,
                        SiteDesignTitle: d.SiteDesignTitle,
                        StartTime: new Date(parseInt(d.StartTime)).toLocaleString()
                    };
                }));
            }
            if (this.verbose) {
                cmd.log(vorpal.chalk.green('DONE'));
            }
            cb();
        }, (err) => this.handleRejectedODataJsonPromise(err, cmd, cb));
    }
    options() {
        const options = [
            {
                option: '-u, --webUrl <webUrl>',
                description: 'The URL of the site for which to list applied site designs'
            },
            {
                option: '-i, --siteDesignId [siteDesignId]',
                description: 'The ID of the site design for which to display information'
            }
        ];
        const parentOptions = super.options();
        return options.concat(parentOptions);
    }
    validate() {
        return (args) => {
            if (!args.options.webUrl) {
                return 'Required parameter webUrl missing';
            }
            const isValidSharePointUrl = SpoCommand_1.default.isValidSharePointUrl(args.options.webUrl);
            if (isValidSharePointUrl !== true) {
                return isValidSharePointUrl;
            }
            if (args.options.siteDesignId) {
                if (!Utils_1.default.isValidGuid(args.options.siteDesignId)) {
                    return `${args.options.siteDesignId} is not a valid GUID`;
                }
            }
            return true;
        };
    }
    commandHelp(args, log) {
        log(vorpal.find(this.name).helpInformation());
        log(`  Examples:
  
    List site designs applied to the specified site
      ${this.name} --webUrl https://contoso.sharepoint.com/sites/team-a

    List information about the specified site design applied to the specified
    site
      ${this.name} --webUrl https://contoso.sharepoint.com/sites/team-a --siteDesignId 6ec3ca5b-d04b-4381-b169-61378556d76e

  More information:

    SharePoint site design and site script overview
      https://docs.microsoft.com/en-us/sharepoint/dev/declarative-customization/site-design-overview
`);
    }
}
module.exports = new SpoSiteDesignRunListCommand();
