// @pnp/office365-cli/dist/o365/teams/commands/guestsettings/guestsettings-list.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const commands_1 = require("../../commands");
const GraphCommand_1 = require("../../../base/GraphCommand");
const request_1 = require("../../../../request");
const Utils_1 = require("../../../../Utils");
const vorpal = require('../../../../vorpal-init');
class TeamsGuestSettingsListCommand extends GraphCommand_1.default {
    get name() {
        return `${commands_1.default.TEAMS_GUESTSETTINGS_LIST}`;
    }
    get description() {
        return 'Lists guest settings for a Microsoft Teams team';
    }
    commandAction(cmd, args, cb) {
        const requestOptions = {
            url: `${this.resource}/v1.0/teams/${encodeURIComponent(args.options.teamId)}?$select=guestSettings`,
            headers: {
                accept: 'application/json;odata.metadata=none'
            },
            json: true
        };
        request_1.default
            .get(requestOptions)
            .then((res) => {
            cmd.log(res.guestSettings);
            if (this.verbose) {
                cmd.log(vorpal.chalk.green('DONE'));
            }
            cb();
        }, (err) => this.handleRejectedODataJsonPromise(err, cmd, cb));
    }
    options() {
        const options = [
            {
                option: '-i, --teamId <teamId>',
                description: 'The ID of the team for which to get the guest settings'
            }
        ];
        const parentOptions = super.options();
        return options.concat(parentOptions);
    }
    validate() {
        return (args) => {
            if (!args.options.teamId) {
                return 'Required parameter teamId missing';
            }
            if (!Utils_1.default.isValidGuid(args.options.teamId)) {
                return `${args.options.teamId} is not a valid GUID`;
            }
            return true;
        };
    }
    commandHelp(args, log) {
        log(vorpal.find(this.name).helpInformation());
        log(`  Examples:
         
    Get guest settings for a Microsoft Teams team
      ${commands_1.default.TEAMS_GUESTSETTINGS_LIST} --teamId 2609af39-7775-4f94-a3dc-0dd67657e900
`);
    }
}
module.exports = new TeamsGuestSettingsListCommand();
