// @pnp/office365-cli/dist/o365/teams/commands/team/team-remove.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const commands_1 = require("../../commands");
const Utils_1 = require("../../../../Utils");
const request_1 = require("../../../../request");
const GraphCommand_1 = require("../../../base/GraphCommand");
const vorpal = require('../../../../vorpal-init');
class TeamsRemoveCommand extends GraphCommand_1.default {
    get name() {
        return `${commands_1.default.TEAMS_TEAM_REMOVE}`;
    }
    get description() {
        return 'Removes the specified Microsoft Teams team';
    }
    getTelemetryProperties(args) {
        const telemetryProps = super.getTelemetryProperties(args);
        telemetryProps.confirm = (!(!args.options.confirm)).toString();
        return telemetryProps;
    }
    commandAction(cmd, args, cb) {
        const removeTeam = () => {
            const requestOptions = {
                url: `${this.resource}/v1.0/groups/${encodeURIComponent(args.options.teamId)}`,
                headers: {
                    accept: 'application/json;odata.metadata=none'
                },
                json: true
            };
            request_1.default
                .delete(requestOptions)
                .then(() => {
                if (this.verbose) {
                    cmd.log(vorpal.chalk.green('DONE'));
                }
                cb();
            }, (err) => this.handleRejectedODataJsonPromise(err, cmd, cb));
        };
        if (args.options.confirm) {
            removeTeam();
        }
        else {
            cmd.prompt({
                type: 'confirm',
                name: 'continue',
                default: false,
                message: `Are you sure you want to remove the team ${args.options.teamId}?`,
            }, (result) => {
                if (!result.continue) {
                    cb();
                }
                else {
                    removeTeam();
                }
            });
        }
    }
    options() {
        const options = [
            {
                option: '-i, --teamId <teamId>',
                description: 'The ID of the Teams team to remove'
            },
            {
                option: '--confirm',
                description: 'Don\'t prompt for confirming removing the specified team'
            }
        ];
        const parentOptions = super.options();
        return options.concat(parentOptions);
    }
    validate() {
        return (args) => {
            if (!args.options.teamId) {
                return 'Required parameter teamId missing';
            }
            if (!Utils_1.default.isValidGuid(args.options.teamId)) {
                return `${args.options.teamId} is not a valid GUID`;
            }
            return true;
        };
    }
    commandHelp(args, log) {
        log(vorpal.find(this.name).helpInformation());
        log(`  Remarks:

    When deleted, Office 365 groups are moved to a temporary container and
    can be restored within 30 days. After that time, they are permanently
    deleted. This applies only to Office 365 groups.

  Examples:
  
    Removes the specified team 
      ${this.name} --teamId '00000000-0000-0000-0000-000000000000'

    Removes the specified team without confirmation
      ${this.name} --teamId '00000000-0000-0000-0000-000000000000' --confirm

  More information:

    directory resource type (deleted items)
      https://docs.microsoft.com/en-us/graph/api/resources/directory?view=graph-rest-1.0
  `);
    }
}
module.exports = new TeamsRemoveCommand();
