// @pnp/office365-cli/dist/o365/base/GraphItemsListCommand.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const GraphCommand_1 = require("./GraphCommand");
const request_1 = require("../../request");
class GraphItemsListCommand extends GraphCommand_1.default {
    constructor() {
        super();
        this.items = [];
    }
    getAllItems(url, cmd, firstRun) {
        return new Promise((resolve, reject) => {
            const requestOptions = {
                url: url,
                headers: {
                    accept: 'application/json;odata.metadata=none'
                },
                json: true
            };
            request_1.default
                .get(requestOptions)
                .then((res) => {
                if (firstRun) {
                    this.items = [];
                }
                this.items = this.items.concat(res.value);
                if (res['@odata.nextLink']) {
                    this
                        .getAllItems(res['@odata.nextLink'], cmd, false)
                        .then(() => {
                        resolve();
                    }, (err) => {
                        reject(err);
                    });
                }
                else {
                    resolve();
                }
            }, (err) => {
                reject(err);
            });
        });
    }
}
exports.GraphItemsListCommand = GraphItemsListCommand;
