// @pnp/office365-cli/dist/o365/spo/commands/tenant/tenant-appcatalogurl-get.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const request_1 = require("../../../../request");
const commands_1 = require("../../commands");
const SpoCommand_1 = require("../../../base/SpoCommand");
const vorpal = require('../../../../vorpal-init');
class SpoTenantAppCatalogUrlGetCommand extends SpoCommand_1.default {
    get name() {
        return commands_1.default.TENANT_APPCATALOGURL_GET;
    }
    get description() {
        return 'Gets the URL of the tenant app catalog';
    }
    commandAction(cmd, args, cb) {
        this
            .getSpoUrl(cmd, this.debug)
            .then((spoUrl) => {
            const requestOptions = {
                url: `${spoUrl}/_api/SP_TenantSettings_Current`,
                headers: {
                    accept: 'application/json;odata=nometadata'
                }
            };
            return request_1.default.get(requestOptions);
        })
            .then((res) => {
            const json = JSON.parse(res);
            if (json.CorporateCatalogUrl) {
                cmd.log(json.CorporateCatalogUrl);
            }
            else {
                if (this.verbose) {
                    cmd.log("Tenant app catalog is not configured.");
                }
            }
            cb();
        }, (err) => this.handleRejectedPromise(err, cmd, cb));
    }
    commandHelp(args, log) {
        log(vorpal.find(this.name).helpInformation());
        log(`  Examples:
  
    Get the URL of the tenant app catalog
      ${commands_1.default.TENANT_APPCATALOGURL_GET}
  `);
    }
}
module.exports = new SpoTenantAppCatalogUrlGetCommand();
