// @pnp/office365-cli/dist/o365/flow/commands/flow-disable.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const commands_1 = require("../commands");
const request_1 = require("../../../request");
const AzmgmtCommand_1 = require("../../base/AzmgmtCommand");
const os = require("os");
const vorpal = require('../../../vorpal-init');
class FlowDisableCommand extends AzmgmtCommand_1.default {
    get name() {
        return commands_1.default.FLOW_DISABLE;
    }
    get description() {
        return 'Disables specified Microsoft Flow';
    }
    commandAction(cmd, args, cb) {
        if (this.verbose) {
            cmd.log(`Disables Microsoft Flow ${args.options.name}...`);
        }
        const requestOptions = {
            url: `${this.resource}providers/Microsoft.ProcessSimple/${args.options.asAdmin ? 'scopes/admin/' : ''}environments/${encodeURIComponent(args.options.environment)}/flows/${encodeURIComponent(args.options.name)}/stop?api-version=2016-11-01`,
            headers: {
                accept: 'application/json'
            },
            json: true
        };
        request_1.default
            .post(requestOptions)
            .then(() => {
            if (this.verbose) {
                cmd.log(vorpal.chalk.green('DONE'));
            }
            cb();
        }, (rawRes) => this.handleRejectedODataJsonPromise(rawRes, cmd, cb));
    }
    options() {
        const options = [
            {
                option: '-n, --name <name>',
                description: 'The name of the Microsoft Flow to disable'
            },
            {
                option: '-e, --environment <environment>',
                description: 'The name of the environment for which to disable Flow'
            },
            {
                option: '--asAdmin',
                description: 'Set, to disable the Flow as admin'
            }
        ];
        const parentOptions = super.options();
        return options.concat(parentOptions);
    }
    validate() {
        return (args) => {
            if (!args.options.name) {
                return 'Required option name missing';
            }
            if (!args.options.environment) {
                return 'Required option environment missing';
            }
            return true;
        };
    }
    commandHelp(args, log) {
        const chalk = vorpal.chalk;
        log(vorpal.find(commands_1.default.FLOW_DISABLE).helpInformation());
        log(`  Remarks:

    ${chalk.yellow('Attention:')} This command is based on an API that is currently
    in preview and is subject to change once the API reached general
    availability.
  
    By default, the command will try to disable Microsoft Flows you own.
    If you want to disable Flow owned by another user, use the ${chalk.blue('asAdmin')}
    flag.

    If the environment with the name you specified doesn't exist, you will get
    the ${chalk.grey('Access to the environment \'xyz\' is denied.')} error.

    If the Microsoft Flow with the name you specified doesn't exist, you will
    get the ${chalk.grey(`The caller with object id \'abc\' does not have permission${os.EOL}` +
            '    for connection \'xyz\' under Api \'shared_logicflows\'.')} error.
    If you try to disable a non-existing flow as admin, you will get the
    ${chalk.grey('Could not find flow \'xyz\'.')} error.
   
  Examples:
  
    Disables Microsoft Flow owned by the currently signed-in user
      ${this.getCommandName()} --environment Default-d87a7535-dd31-4437-bfe1-95340acd55c5 --name 3989cb59-ce1a-4a5c-bb78-257c5c39381d

    Disables Microsoft Flow owned by another user
      ${this.getCommandName()} --environment Default-d87a7535-dd31-4437-bfe1-95340acd55c5 --name 3989cb59-ce1a-4a5c-bb78-257c5c39381d --asAdmin
`);
    }
}
module.exports = new FlowDisableCommand();
