// @pnp/office365-cli/dist/o365/flow/commands/flow-export.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const commands_1 = require("../commands");
const request_1 = require("../../../request");
const AzmgmtCommand_1 = require("../../base/AzmgmtCommand");
const Utils_1 = require("../../../Utils");
const os = require("os");
const path = require("path");
const fs = require("fs");
const vorpal = require('../../../vorpal-init');
class FlowExportCommand extends AzmgmtCommand_1.default {
    get name() {
        return commands_1.default.FLOW_EXPORT;
    }
    get description() {
        return 'Exports the specified Microsoft Flow as a file';
    }
    getTelemetryProperties(args) {
        const telemetryProps = super.getTelemetryProperties(args);
        telemetryProps.packageDisplayName = typeof args.options.packageDisplayName !== 'undefined';
        telemetryProps.packageDescription = typeof args.options.packageDescription !== 'undefined';
        telemetryProps.packageCreatedBy = typeof args.options.packageCreatedBy !== 'undefined';
        telemetryProps.packageSourceEnvironment = typeof args.options.packageSourceEnvironment !== 'undefined';
        telemetryProps.format = args.options.format;
        telemetryProps.path = typeof args.options.path !== 'undefined';
        return telemetryProps;
    }
    commandAction(cmd, args, cb) {
        let filenameFromApi = '';
        const formatArgument = args.options.format ? args.options.format.toLowerCase() : '';
        if (this.verbose) {
            cmd.log(`Retrieving package resources for Microsoft Flow ${args.options.id}...`);
        }
        (() => {
            if (formatArgument === 'json') {
                if (this.debug) {
                    cmd.log('format = json, skipping listing package resources step');
                }
                return Promise.resolve();
            }
            const requestOptions = {
                url: `${this.resource}providers/Microsoft.BusinessAppPlatform/environments/${encodeURIComponent(args.options.environment)}/listPackageResources?api-version=2016-11-01`,
                headers: {
                    accept: 'application/json'
                },
                body: {
                    "baseResourceIds": [
                        `/providers/Microsoft.Flow/flows/${args.options.id}`
                    ]
                },
                json: true
            };
            return request_1.default.post(requestOptions);
        })()
            .then((res) => {
            if (this.verbose) {
                cmd.log(`Initiating package export for Microsoft Flow ${args.options.id}...`);
            }
            const requestOptions = {
                url: `${this.resource}providers/${formatArgument === 'json' ?
                    `Microsoft.ProcessSimple/environments/${encodeURIComponent(args.options.environment)}/flows/${encodeURIComponent(args.options.id)}?api-version=2016-11-01`
                    : `Microsoft.BusinessAppPlatform/environments/${encodeURIComponent(args.options.environment)}/exportPackage?api-version=2016-11-01`}`,
                headers: {
                    accept: 'application/json'
                },
                json: true
            };
            if (formatArgument !== 'json') {
                requestOptions['body'] = {
                    "includedResourceIds": [
                        `/providers/Microsoft.Flow/flows/${args.options.id}`
                    ],
                    "details": {
                        "displayName": args.options.packageDisplayName,
                        "description": args.options.packageDescription,
                        "creator": args.options.packageCreatedBy,
                        "sourceEnvironment": args.options.packageSourceEnvironment
                    },
                    "resources": res.resources
                };
            }
            return formatArgument === 'json' ? request_1.default.get(requestOptions) : request_1.default.post(requestOptions);
        })
            .then((res) => {
            if (this.verbose) {
                cmd.log(`Getting file for Microsoft Flow ${args.options.id}...`);
            }
            if (res.errors && res.errors.length && res.errors.length > 0) {
                return Promise.reject(res.errors[0].message);
            }
            const downloadFileUrl = formatArgument === 'json' ? '' : res.packageLink.value;
            const filenameRegEx = /([^\/]+\.zip)/i;
            filenameFromApi = formatArgument === 'json' ? `${res.properties.displayName}.json` : (filenameRegEx.exec(downloadFileUrl) || ['output.zip'])[0];
            if (this.debug) {
                cmd.log(`Filename from PowerApps API: ${filenameFromApi}`);
                cmd.log('');
            }
            const requestOptions = {
                url: formatArgument === 'json' ?
                    `${this.resource}/providers/Microsoft.ProcessSimple/environments/${encodeURIComponent(args.options.environment)}/flows/${encodeURIComponent(args.options.id)}/exportToARMTemplate?api-version=2016-11-01`
                    : downloadFileUrl,
                encoding: null,
                headers: formatArgument === 'json' ? {
                    accept: 'application/json'
                } : {
                    'x-anonymous': true
                }
            };
            return formatArgument === 'json' ?
                request_1.default.post(requestOptions)
                : request_1.default.get(requestOptions);
        })
            .then((file) => {
            const path = args.options.path ? args.options.path : `./${filenameFromApi}`;
            fs.writeFileSync(path, file, 'binary');
            if (!args.options.path || this.verbose) {
                if (this.verbose) {
                    cmd.log(`File saved to path '${path}'`);
                }
                else {
                    cmd.log(path);
                }
            }
            cb();
        }, (rawRes) => this.handleRejectedODataJsonPromise(rawRes, cmd, cb));
    }
    options() {
        const options = [
            {
                option: '-i, --id <id>',
                description: 'The id of the Microsoft Flow to export'
            },
            {
                option: '-e, --environment <environment>',
                description: 'The name of the environment from which to export the Flow from'
            },
            {
                option: '-n, --packageDisplayName [packageDisplayName]',
                description: 'The display name to use in the exported package'
            },
            {
                option: '-d, --packageDescription [packageDescription]',
                description: 'The description to use in the exported package'
            },
            {
                option: '-c, --packageCreatedBy [packageCreatedBy]',
                description: 'The name of the person to be used as the creator of the exported package'
            },
            {
                option: '-s, --packageSourceEnvironment [packageSourceEnvironment]',
                description: 'The name of the source environment from which the exported package was taken'
            },
            {
                option: '-f, --format [format]',
                description: 'The format to export the Flow to json|zip. Default json'
            },
            {
                option: '-p, --path [path]',
                description: 'The path to save the exported package to'
            },
        ];
        const parentOptions = super.options();
        return options.concat(parentOptions);
    }
    validate() {
        return (args) => {
            const lowerCaseFormat = args.options.format ? args.options.format.toLowerCase() : '';
            if (!args.options.id) {
                return 'Required option id missing';
            }
            if (!args.options.environment) {
                return 'Required option environment missing';
            }
            if (!Utils_1.default.isValidGuid(args.options.id)) {
                return `${args.options.id} is not a valid GUID`;
            }
            if (args.options.format && (lowerCaseFormat !== 'json' && lowerCaseFormat !== 'zip')) {
                return 'Option format must be json or zip. Default is zip';
            }
            if (lowerCaseFormat === 'json') {
                if (args.options.packageCreatedBy) {
                    return 'packageCreatedBy cannot be specified with output of json';
                }
                if (args.options.packageDescription) {
                    return 'packageDescription cannot be specified with output of json';
                }
                if (args.options.packageDisplayName) {
                    return 'packageDisplayName cannot be specified with output of json';
                }
                if (args.options.packageSourceEnvironment) {
                    return 'packageSourceEnvironment cannot be specified with output of json';
                }
            }
            if (args.options.path && !fs.existsSync(path.dirname(args.options.path))) {
                return 'Specified path where to save the file does not exist';
            }
            return true;
        };
    }
    commandHelp(args, log) {
        const chalk = vorpal.chalk;
        log(vorpal.find(commands_1.default.FLOW_EXPORT).helpInformation());
        log(`  Remarks:

    ${chalk.yellow('Attention:')} This command is based on an API that is currently
    in preview and is subject to change once the API reached general
    availability.
  
    If the environment with the name you specified doesn't exist, you will get
    the ${chalk.grey('Access to the environment \'xyz\' is denied.')} error.

    If the Microsoft Flow with the id you specified doesn't exist, you will
    get the ${chalk.grey(`The caller with object id \'abc\' does not have permission${os.EOL}` +
            '    for connection \'xyz\' under Api \'shared_logicflows\'.')} error.
   
  Examples:
  
    Export the specified Microsoft Flow as a ZIP file
      ${this.getCommandName()} --environment Default-d87a7535-dd31-4437-bfe1-95340acd55c5 --id 3989cb59-ce1a-4a5c-bb78-257c5c39381d

    Export the specified Microsoft Flow as a JSON file
      ${this.getCommandName()} --environment Default-d87a7535-dd31-4437-bfe1-95340acd55c5 --id 3989cb59-ce1a-4a5c-bb78-257c5c39381d --format json

    Export the specified Microsoft Flow as a ZIP file with a package display name of 'My flow name'
      ${this.getCommandName()} --environment Default-d87a7535-dd31-4437-bfe1-95340acd55c5 --id 3989cb59-ce1a-4a5c-bb78-257c5c39381d --packageDisplayName 'My flow name'

    Export the specified Microsoft Flow as a ZIP file named 'MyFlow.zip' saved to the current directory
      ${this.getCommandName()} --environment Default-d87a7535-dd31-4437-bfe1-95340acd55c5 --id 3989cb59-ce1a-4a5c-bb78-257c5c39381d --path './MyFlow.zip'
`);
    }
}
module.exports = new FlowExportCommand();
