// @pnp/office365-cli/dist/o365/spo/commands/sitescript/sitescript-remove.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const request_1 = require("../../../../request");
const commands_1 = require("../../commands");
const SpoCommand_1 = require("../../../base/SpoCommand");
const Utils_1 = require("../../../../Utils");
const vorpal = require('../../../../vorpal-init');
class SpoSiteScriptRemoveCommand extends SpoCommand_1.default {
    get name() {
        return `${commands_1.default.SITESCRIPT_REMOVE}`;
    }
    get description() {
        return 'Removes the specified site script';
    }
    getTelemetryProperties(args) {
        const telemetryProps = super.getTelemetryProperties(args);
        telemetryProps.confirm = args.options.confirm || false;
        return telemetryProps;
    }
    commandAction(cmd, args, cb) {
        const removeSiteScript = () => {
            let spoUrl = '';
            this
                .getSpoUrl(cmd, this.debug)
                .then((_spoUrl) => {
                spoUrl = _spoUrl;
                return this.getRequestDigest(spoUrl);
            })
                .then((res) => {
                const requestOptions = {
                    url: `${spoUrl}/_api/Microsoft.Sharepoint.Utilities.WebTemplateExtensions.SiteScriptUtility.DeleteSiteScript`,
                    headers: {
                        'X-RequestDigest': res.FormDigestValue,
                        'content-type': 'application/json;charset=utf-8',
                        accept: 'application/json;odata=nometadata'
                    },
                    body: { id: args.options.id },
                    json: true
                };
                return request_1.default.post(requestOptions);
            })
                .then(() => {
                if (this.verbose) {
                    cmd.log(vorpal.chalk.green('DONE'));
                }
                cb();
            }, (err) => this.handleRejectedODataJsonPromise(err, cmd, cb));
        };
        if (args.options.confirm) {
            removeSiteScript();
        }
        else {
            cmd.prompt({
                type: 'confirm',
                name: 'continue',
                default: false,
                message: `Are you sure you want to remove the site script ${args.options.id}?`,
            }, (result) => {
                if (!result.continue) {
                    cb();
                }
                else {
                    removeSiteScript();
                }
            });
        }
    }
    options() {
        const options = [
            {
                option: '-i, --id <id>',
                description: 'Site script ID'
            },
            {
                option: '--confirm',
                description: 'Don\'t prompt for confirming removing the site script'
            }
        ];
        const parentOptions = super.options();
        return options.concat(parentOptions);
    }
    validate() {
        return (args) => {
            if (!args.options.id) {
                return 'Required parameter id missing';
            }
            if (!Utils_1.default.isValidGuid(args.options.id)) {
                return `${args.options.id} is not a valid GUID`;
            }
            return true;
        };
    }
    commandHelp(args, log) {
        const chalk = vorpal.chalk;
        log(vorpal.find(this.name).helpInformation());
        log(`  Remarks:

    If the specified ${chalk.grey('id')} doesn't refer to an existing site script, you will get
    a ${chalk.grey('File not found')} error.

  Examples:
  
    Remove site script with ID ${chalk.grey('2c1ba4c4-cd9b-4417-832f-92a34bc34b2a')}. Will prompt
    for confirmation before removing the script
      ${this.name} --id 2c1ba4c4-cd9b-4417-832f-92a34bc34b2a

    Remove site script with ID ${chalk.grey('2c1ba4c4-cd9b-4417-832f-92a34bc34b2a')} without prompting
    for confirmation
      ${this.name} --id 2c1ba4c4-cd9b-4417-832f-92a34bc34b2a --confirm

  More information:

    SharePoint site design and site script overview
      https://docs.microsoft.com/en-us/sharepoint/dev/declarative-customization/site-design-overview
`);
    }
}
module.exports = new SpoSiteScriptRemoveCommand();
