// @pnp/office365-cli/dist/o365/teams/commands/funsettings/funsettings-list.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const Utils_1 = require("../../../../Utils");
const commands_1 = require("../../commands");
const GraphCommand_1 = require("../../../base/GraphCommand");
const request_1 = require("../../../../request");
const vorpal = require('../../../../vorpal-init');
class TeamsFunSettingsListCommand extends GraphCommand_1.default {
    get name() {
        return `${commands_1.default.TEAMS_FUNSETTINGS_LIST}`;
    }
    get description() {
        return 'Lists fun settings for the specified Microsoft Teams team';
    }
    commandAction(cmd, args, cb) {
        const requestOptions = {
            url: `${this.resource}/v1.0/teams/${encodeURIComponent(args.options.teamId)}?$select=funSettings`,
            headers: {
                accept: 'application/json;odata.metadata=none'
            },
            json: true
        };
        request_1.default
            .get(requestOptions)
            .then((res) => {
            cmd.log(res.funSettings);
            if (this.verbose) {
                cmd.log(vorpal.chalk.green('DONE'));
            }
            cb();
        }, (err) => this.handleRejectedODataJsonPromise(err, cmd, cb));
    }
    ;
    options() {
        const options = [
            {
                option: '-i, --teamId <teamId>',
                description: 'The ID of the team for which to list fun settings'
            },
        ];
        const parentOptions = super.options();
        return options.concat(parentOptions);
    }
    validate() {
        return (args) => {
            if (!args.options.teamId) {
                return 'Required parameter teamId missing';
            }
            if (!Utils_1.default.isValidGuid(args.options.teamId)) {
                return `${args.options.teamId} is not a valid GUID`;
            }
            return true;
        };
    }
    commandHelp(args, log) {
        log(vorpal.find(this.name).helpInformation());
        log(`  Examples:

    List fun settings of a Microsoft Teams team
      ${this.name} --teamId 83cece1e-938d-44a1-8b86-918cf6151957
`);
    }
}
module.exports = new TeamsFunSettingsListCommand();
