// @pnp/office365-cli/dist/o365/spo/commands/storageentity/storageentity-list.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const request_1 = require("../../../../request");
const commands_1 = require("../../commands");
const SpoCommand_1 = require("../../../base/SpoCommand");
const vorpal = require('../../../../vorpal-init');
class SpoStorageEntityListCommand extends SpoCommand_1.default {
    get name() {
        return `${commands_1.default.STORAGEENTITY_LIST}`;
    }
    get description() {
        return 'Lists tenant properties stored on the specified SharePoint Online app catalog';
    }
    commandAction(cmd, args, cb) {
        if (this.verbose) {
            cmd.log(`Retrieving details for all tenant properties in ${args.options.appCatalogUrl}...`);
        }
        const requestOptions = {
            url: `${args.options.appCatalogUrl}/_api/web/AllProperties?$select=storageentitiesindex`,
            headers: {
                accept: 'application/json;odata=nometadata'
            },
            json: true
        };
        request_1.default
            .get(requestOptions)
            .then((web) => {
            try {
                if (!web.storageentitiesindex ||
                    web.storageentitiesindex.trim().length === 0) {
                    if (this.verbose) {
                        cmd.log('No tenant properties found');
                    }
                    cb();
                    return;
                }
                const properties = JSON.parse(web.storageentitiesindex);
                const keys = Object.keys(properties);
                if (keys.length === 0) {
                    if (this.verbose) {
                        cmd.log('No tenant properties found');
                    }
                }
                else {
                    cmd.log(keys.map((key) => {
                        const property = properties[key];
                        return {
                            Key: key,
                            Value: property.Value,
                            Description: property.Description,
                            Comment: property.Comment
                        };
                    }));
                }
                cb();
            }
            catch (e) {
                this.handleError(e, cmd, cb);
            }
        }, (err) => this.handleRejectedPromise(err, cmd, cb));
    }
    options() {
        const options = [{
                option: '-u, --appCatalogUrl <appCatalogUrl>',
                description: 'URL of the app catalog site'
            }];
        const parentOptions = super.options();
        return options.concat(parentOptions);
    }
    validate() {
        return (args) => {
            const result = SpoCommand_1.default.isValidSharePointUrl(args.options.appCatalogUrl);
            if (result === false) {
                return 'Missing required option appCatalogUrl';
            }
            else {
                return result;
            }
        };
    }
    commandHelp(args, log) {
        const chalk = vorpal.chalk;
        log(vorpal.find(commands_1.default.STORAGEENTITY_LIST).helpInformation());
        log(`  Remarks:

    Tenant properties are stored in the app catalog site. To list all tenant
    properties, you have to specify the absolute URL of the app catalog site.
    If you specify an incorrect URL, or the site at the given URL is not an
    app catalog site, no properties will be retrieved.

  Examples:
  
    List all tenant properties stored in the
    ${chalk.grey('https://contoso.sharepoint.com/sites/appcatalog')} app catalog site
      ${commands_1.default.STORAGEENTITY_LIST} --appCatalogUrl https://contoso.sharepoint.com/sites/appcatalog

  More information:

    SharePoint Framework Tenant Properties
      https://docs.microsoft.com/en-us/sharepoint/dev/spfx/tenant-properties
`);
    }
}
module.exports = new SpoStorageEntityListCommand();
