// @pnp/office365-cli/dist/o365/spo/commands/propertybag/propertybag-list.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const commands_1 = require("../../commands");
const SpoCommand_1 = require("../../../base/SpoCommand");
const propertybag_base_1 = require("./propertybag-base");
const ClientSvc_1 = require("../../ClientSvc");
const vorpal = require('../../../../vorpal-init');
class SpoPropertyBagListCommand extends propertybag_base_1.SpoPropertyBagBaseCommand {
    get name() {
        return `${commands_1.default.PROPERTYBAG_LIST}`;
    }
    get description() {
        return 'Gets property bag values';
    }
    getTelemetryProperties(args) {
        const telemetryProps = super.getTelemetryProperties(args);
        telemetryProps.folder = (!(!args.options.folder)).toString();
        return telemetryProps;
    }
    commandAction(cmd, args, cb) {
        const clientSvcCommons = new ClientSvc_1.ClientSvc(cmd, this.debug);
        this
            .getRequestDigest(args.options.webUrl)
            .then((contextResponse) => {
            this.formDigestValue = contextResponse.FormDigestValue;
            return clientSvcCommons.getCurrentWebIdentity(args.options.webUrl, this.formDigestValue);
        })
            .then((identityResp) => {
            const opts = args.options;
            if (opts.folder) {
                return this.getFolderPropertyBag(identityResp, opts.webUrl, opts.folder, cmd);
            }
            return this.getWebPropertyBag(identityResp, opts.webUrl, cmd);
        })
            .then((propertyBagData) => {
            cmd.log(this.formatOutput(propertyBagData));
            cb();
        }, (err) => this.handleRejectedPromise(err, cmd, cb));
    }
    options() {
        const options = [
            {
                option: '-u, --webUrl <webUrl>',
                description: 'The URL of the site from which the property bag value should be retrieved'
            },
            {
                option: '-f, --folder [folder]',
                description: 'Site-relative URL of the folder from which to retrieve property bag value. Case-sensitive',
            }
        ];
        const parentOptions = super.options();
        return options.concat(parentOptions);
    }
    validate() {
        return (args) => {
            if (SpoCommand_1.default.isValidSharePointUrl(args.options.webUrl) !== true) {
                return 'Missing required option url';
            }
            return true;
        };
    }
    commandHelp(args, log) {
        const chalk = vorpal.chalk;
        log(vorpal.find(commands_1.default.PROPERTYBAG_LIST).helpInformation());
        log(`  Examples:

    Return property bag values located in site ${chalk.grey('https://contoso.sharepoint.com/sites/test')}
      ${commands_1.default.PROPERTYBAG_LIST} --webUrl https://contoso.sharepoint.com/sites/test

    Return property bag values located in site root folder ${chalk.grey('https://contoso.sharepoint.com/sites/test')}
      ${commands_1.default.PROPERTYBAG_LIST} --webUrl https://contoso.sharepoint.com/sites/test -f /

    Return property bag values located in site document library ${chalk.grey('https://contoso.sharepoint.com/sites/test')}
      ${commands_1.default.PROPERTYBAG_LIST} --webUrl https://contoso.sharepoint.com/sites/test --folder '/Shared Documents'

    Return property bag values located in folder in site document library ${chalk.grey('https://contoso.sharepoint.com/sites/test')}
      ${commands_1.default.PROPERTYBAG_LIST} -w https://contoso.sharepoint.com/sites/test -f '/Shared Documents/MyFolder'

    Return property bag values located in site list ${chalk.grey('https://contoso.sharepoint.com/sites/test')}
      ${commands_1.default.PROPERTYBAG_LIST} --webUrl https://contoso.sharepoint.com/sites/test --folder /Lists/MyList
    `);
    }
    /**
     * The property bag data returned from the client.svc/ProcessQuery response
     * has to be formatted before displayed since the key, value objects
     * carry extra information.
     * @param propertyBag client.svc property bag javascript object
     */
    formatOutput(propertyBag) {
        let result = [];
        const keys = Object.keys(propertyBag);
        for (let i = 0; i < keys.length; i++) {
            if (keys[i] === '_ObjectType_') {
                // this is system data, do not include it
                continue;
            }
            const formattedProp = this.formatProperty(keys[i], propertyBag[keys[i]]);
            result.push(formattedProp);
        }
        return result;
    }
}
module.exports = new SpoPropertyBagListCommand();
