// @pnp/office365-cli/dist/o365/teams/commands/messagingsettings/messagingsettings-set.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const commands_1 = require("../../commands");
const Utils_1 = require("../../../../Utils");
const request_1 = require("../../../../request");
const GraphCommand_1 = require("../../../base/GraphCommand");
const vorpal = require('../../../../vorpal-init');
class TeamsMessageSettingsSetCommand extends GraphCommand_1.default {
    get name() {
        return `${commands_1.default.TEAMS_MESSAGINGSETTINGS_SET}`;
    }
    get description() {
        return 'Updates messaging settings of a Microsoft Teams team';
    }
    getTelemetryProperties(args) {
        const telemetryProps = super.getTelemetryProperties(args);
        TeamsMessageSettingsSetCommand.props.forEach((p) => {
            telemetryProps[p] = typeof args.options[p] !== 'undefined';
        });
        return telemetryProps;
    }
    commandAction(cmd, args, cb) {
        const body = {
            messagingSettings: {}
        };
        TeamsMessageSettingsSetCommand.props.forEach((p) => {
            if (typeof args.options[p] !== 'undefined') {
                body.messagingSettings[p] = args.options[p].toLowerCase() === 'true';
            }
        });
        const requestOptions = {
            url: `${this.resource}/v1.0/teams/${encodeURIComponent(args.options.teamId)}`,
            headers: {
                accept: 'application/json;odata.metadata=none'
            },
            body: body,
            json: true
        };
        request_1.default
            .patch(requestOptions)
            .then(() => {
            if (this.verbose) {
                cmd.log(vorpal.chalk.green('DONE'));
            }
            cb();
        }, (err) => this.handleRejectedODataJsonPromise(err, cmd, cb));
    }
    options() {
        const options = [
            {
                option: '-i, --teamId <teamId>',
                description: 'The ID of the Microsoft Teams team for which to update messaging settings'
            },
            {
                option: '--allowUserEditMessages [allowUserEditMessages]',
                description: 'Set to true to allow users to edit messages and to false to disallow it'
            },
            {
                option: '--allowUserDeleteMessages [allowUserDeleteMessages]',
                description: 'Set to true to allow users to delete messages and to false to disallow it'
            },
            {
                option: '--allowOwnerDeleteMessages [allowOwnerDeleteMessages]',
                description: 'Set to true to allow owner to delete messages and to false to disallow it'
            },
            {
                option: '--allowTeamMentions [allowTeamMentions]',
                description: 'Set to true to allow @team or @[team name] mentions and to false to disallow it'
            },
            {
                option: '--allowChannelMentions [allowChannelMentions]',
                description: 'Set to true to allow @channel or @[channel name] mentions and to false to disallow it'
            }
        ];
        const parentOptions = super.options();
        return options.concat(parentOptions);
    }
    validate() {
        return (args) => {
            if (!args.options.teamId) {
                return 'Required parameter teamId missing';
            }
            if (!Utils_1.default.isValidGuid(args.options.teamId)) {
                return `${args.options.teamId} is not a valid GUID`;
            }
            let hasDuplicate = false;
            let property = '';
            TeamsMessageSettingsSetCommand.props.forEach((prop) => {
                if (args.options[prop] instanceof Array) {
                    property = prop;
                    hasDuplicate = true;
                }
            });
            if (hasDuplicate) {
                return `Duplicate option ${property} specified. Specify only one`;
            }
            let isValid = true;
            let value = '';
            TeamsMessageSettingsSetCommand.props.every((p) => {
                property = p;
                value = args.options[p];
                isValid = typeof value === 'undefined' || Utils_1.default.isValidBoolean(value);
                return isValid;
            });
            if (!isValid) {
                return `Value ${value} for option ${property} is not a valid boolean`;
            }
            return true;
        };
    }
    commandHelp(args, log) {
        log(vorpal.find(this.name).helpInformation());
        log(`  Examples:
  
    Allow users to edit messages in channels
      ${this.name} --teamId '00000000-0000-0000-0000-000000000000' --allowUserEditMessages true

    Disallow users to delete messages in channels
      ${this.name} --teamId '00000000-0000-0000-0000-000000000000' --allowUserDeleteMessages false
`);
    }
}
TeamsMessageSettingsSetCommand.props = [
    'allowUserEditMessages',
    'allowUserDeleteMessages',
    'allowOwnerDeleteMessages',
    'allowTeamMentions',
    'allowChannelMentions'
];
module.exports = new TeamsMessageSettingsSetCommand();
