// @pnp/office365-cli/dist/o365/spo/commands/web/web-list.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const commands_1 = require("../../commands");
const request_1 = require("../../../../request");
const SpoCommand_1 = require("../../../base/SpoCommand");
const vorpal = require('../../../../vorpal-init');
class SpoWebListCommand extends SpoCommand_1.default {
    get name() {
        return commands_1.default.WEB_LIST;
    }
    get description() {
        return 'Lists subsites of the specified site';
    }
    commandAction(cmd, args, cb) {
        if (this.verbose) {
            cmd.log(`Retrieving all webs in site at ${args.options.webUrl}...`);
        }
        let requestUrl = `${args.options.webUrl}/_api/web/webs`;
        if (args.options.output !== 'json') {
            requestUrl += '?$select=Title,Id,URL';
        }
        const requestOptions = {
            url: requestUrl,
            headers: {
                'accept': 'application/json;odata=nometadata'
            },
            json: true
        };
        request_1.default
            .get(requestOptions)
            .then((webProperties) => {
            if (args.options.output === 'json') {
                cmd.log(webProperties);
            }
            else {
                cmd.log(webProperties.value.map(l => {
                    return {
                        Title: l.Title,
                        Url: l.Url,
                        Id: l.Id
                    };
                }));
            }
            cb();
        }, (err) => this.handleRejectedODataJsonPromise(err, cmd, cb));
    }
    options() {
        const options = [
            {
                option: '-u, --webUrl <webUrl>',
                description: 'URL of the parent site for which to retrieve the list of subsites'
            }
        ];
        const parentOptions = super.options();
        return options.concat(parentOptions);
    }
    validate() {
        return (args) => {
            if (!args.options.webUrl) {
                return 'Required parameter webUrl missing';
            }
            const isValidSharePointUrl = SpoCommand_1.default.isValidSharePointUrl(args.options.webUrl);
            if (isValidSharePointUrl !== true) {
                return isValidSharePointUrl;
            }
            return true;
        };
    }
    commandHelp(args, log) {
        const chalk = vorpal.chalk;
        log(vorpal.find(this.name).helpInformation());
        log(`  Examples:
  
    Return all subsites from site ${chalk.grey('https://contoso.sharepoint.com/sites/project-x')}
      ${commands_1.default.WEB_LIST} --webUrl https://contoso.sharepoint.com/sites/project-x
      `);
    }
}
module.exports = new SpoWebListCommand();
