// @pnp/office365-cli/dist/o365/flow/commands/environment/environment-list.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const commands_1 = require("../../commands");
const request_1 = require("../../../../request");
const AzmgmtCommand_1 = require("../../../base/AzmgmtCommand");
const vorpal = require('../../../../vorpal-init');
class FlowEnvironmentListCommand extends AzmgmtCommand_1.default {
    get name() {
        return commands_1.default.FLOW_ENVIRONMENT_LIST;
    }
    get description() {
        return 'Lists Microsoft Flow environments in the current tenant';
    }
    commandAction(cmd, args, cb) {
        if (this.verbose) {
            cmd.log(`Retrieving list of Microsoft Flow environments...`);
        }
        const requestOptions = {
            url: `${this.resource}providers/Microsoft.ProcessSimple/environments?api-version=2016-11-01`,
            headers: {
                accept: 'application/json'
            },
            json: true
        };
        request_1.default
            .get(requestOptions)
            .then((res) => {
            if (res.value && res.value.length > 0) {
                if (args.options.output === 'json') {
                    cmd.log(res.value);
                }
                else {
                    cmd.log(res.value.map(e => {
                        return {
                            name: e.name,
                            displayName: e.properties.displayName
                        };
                    }));
                }
            }
            cb();
        }, (rawRes) => this.handleRejectedODataJsonPromise(rawRes, cmd, cb));
    }
    commandHelp(args, log) {
        const chalk = vorpal.chalk;
        log(vorpal.find(commands_1.default.FLOW_ENVIRONMENT_LIST).helpInformation());
        log(`  Remarks:

    ${chalk.yellow('Attention:')} This command is based on an API that is currently
    in preview and is subject to change once the API reached general
    availability.
  
  Examples:
  
    List Microsoft Flow environments in the current tenant
      ${this.getCommandName()}
`);
    }
}
module.exports = new FlowEnvironmentListCommand();
