// @pnp/office365-cli/dist/o365/tenant/commands/id/id-get.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const commands_1 = require("../../commands");
const request_1 = require("../../../../request");
const Command_1 = require("../../../../Command");
const vorpal = require('../../../../vorpal-init');
class TenantIdGetCommand extends Command_1.default {
    get name() {
        return `${commands_1.default.TENANT_ID_GET}`;
    }
    get description() {
        return 'Gets Office 365 tenant ID for the specified domain';
    }
    commandAction(cmd, args, cb) {
        const requestOptions = {
            url: `https://login.windows.net/${args.options.domainName}/.well-known/openid-configuration`,
            headers: {
                'content-type': 'application/json',
                accept: 'application/json',
                'x-anonymous': true
            },
            json: true
        };
        request_1.default
            .get(requestOptions)
            .then((res) => {
            if (res.error) {
                cb(new Command_1.CommandError(res.error_description));
                return;
            }
            if (res.token_endpoint) {
                cmd.log(res.token_endpoint.split('/')[3]);
            }
            cb();
        }, (err) => this.handleRejectedODataJsonPromise(err, cmd, cb));
    }
    options() {
        const options = [
            {
                option: '-d, --domainName <domainName>',
                description: 'The domain name for which to retrieve the Office 365 tenant ID'
            }
        ];
        const parentOptions = super.options();
        return options.concat(parentOptions);
    }
    validate() {
        return (args) => {
            if (!args.options.domainName) {
                return 'Required parameter domainName missing';
            }
            return true;
        };
    }
    commandHelp(args, log) {
        log(vorpal.find(commands_1.default.TENANT_ID_GET).helpInformation());
        log(`Examples:
  
    Get Office 365 tenant ID for the specified domain
      ${commands_1.default.TENANT_ID_GET} --domainName contoso.com
`);
    }
}
module.exports = new TenantIdGetCommand();
