// @pnp/office365-cli/dist/o365/teams/commands/tab/tab-list.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const commands_1 = require("../../commands");
const Utils_1 = require("../../../../Utils");
const GraphItemsListCommand_1 = require("../../../base/GraphItemsListCommand");
const vorpal = require('../../../../vorpal-init');
class TeamsTabListCommand extends GraphItemsListCommand_1.GraphItemsListCommand {
    get name() {
        return `${commands_1.default.TEAMS_TAB_LIST}`;
    }
    get description() {
        return 'Lists tabs in the specified Microsoft Teams channel';
    }
    commandAction(cmd, args, cb) {
        const endpoint = `${this.resource}/v1.0/teams/${args.options.teamId}/channels/${encodeURIComponent(args.options.channelId)}/tabs?$expand=teamsApp`;
        this
            .getAllItems(endpoint, cmd, true)
            .then(() => {
            if (args.options.output === 'json') {
                cmd.log(this.items);
            }
            else {
                cmd.log(this.items.map((t) => {
                    return {
                        id: t.id,
                        displayName: t.displayName,
                        teamsAppTabId: t.teamsApp.id,
                    };
                }));
            }
            if (this.verbose) {
                cmd.log(vorpal.chalk.green('DONE'));
            }
            cb();
        }, (err) => this.handleRejectedODataJsonPromise(err, cmd, cb));
    }
    options() {
        const options = [
            {
                option: '-i, --teamId <teamId>',
                description: 'The ID of the Microsoft Teams team where the channel is located'
            },
            {
                option: '-c, --channelId <channelId>',
                description: 'The ID of the channel for which to list tabs'
            }
        ];
        const parentOptions = super.options();
        return options.concat(parentOptions);
    }
    validate() {
        return (args) => {
            if (!args.options.teamId) {
                return 'Required parameter teamId missing';
            }
            if (!args.options.channelId) {
                return 'Required parameter channelId missing';
            }
            if (!Utils_1.default.isValidGuid(args.options.teamId)) {
                return `${args.options.teamId} is not a valid GUID`;
            }
            if (!Utils_1.default.isValidTeamsChannelId(args.options.channelId)) {
                return `${args.options.channelId} is not a valid Teams ChannelId`;
            }
            return true;
        };
    }
    commandHelp(args, log) {
        log(vorpal.find(this.name).helpInformation());
        log(`  Remarks:

    You can only retrieve tabs for teams of which you are a member.

    Tabs 'Conversations' and 'Files' are present in every team and therefore not
    included in the list of available tabs.

  Examples:
  
    List all tabs in a Microsoft Teams channel
      ${this.name} --teamId 00000000-0000-0000-0000-000000000000 --channelId 19:00000000000000000000000000000000@thread.skype

    Include all the values from the tab configuration and associated teams app
      ${this.name} --teamId 00000000-0000-0000-0000-000000000000 --channelId 19:00000000000000000000000000000000@thread.skype --output json
`);
    }
}
module.exports = new TeamsTabListCommand();
